---
title: "uCHOOSE study"
date: "09/03/2019"
output: 
  html_document:
    code_folding: "hide"
    fig_width: 8
    fig_height: 8
    toc: true
    toc_depth: 4
    toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      eval = FALSE,
                      message = FALSE,
                      warning = FALSE,
                      fig.width = 12,
                      fig.height = 8)
```

# Aim
>
>Evaluate the effects of hormonal contraception options on two aspects of the adolescent female genital tract that may render them at increased risk of HIV: genital HIV target cell immune activation/inflammation, and the genital microbiome. 

Hypothesis: progestins alter FGT immune function and microbial environment, which influences genital HIV target cell susceptibility.

# Description of the contents of this file

In this re-analysis, we attempt to 

* clean up the RNAseq data, i.e. remove low-quality/outlier samples and filter out non-targeted genes. 
* run DESeq with treatment as a design-factor, where treatment has 4 levels - pretreatment, Injection, Pill and Ring.
* According-to-protocol, intent-to-treat and strict-per-protocol analyses included here.
* some preliminary analysis on inflammation data is also included.
* output figures/resutls are saved in **uChoose_Analysis_Oct2020/**
* corresponds to **uchoose_work-report_Oct2020.pptx** and other following versions of reports.

## To-Do

* Change the naming as follows -
Injection_NETEN: NET-EN, Pill: COC, Ring: CCVR, Pretreatment: Baseline
[Not changing these right now as we can simply put these annotations in the final figures it’ll be too much to change these names everywhere]
* Complex heatmap package
* Leading edge analysis
* BV/CST to look into variance of the data

# Study Design

Endocervical cytobrush samples were collected from 112 women randomly assigned to one of 3 different contraceptive treatments. Samples were collected at visit 1 and 2, 14 samples from visit 1 and 107 samples from visit 2. The contraceptive was administered every two months. Hence visit 1 is screening and visit 2 is at 16 weeks, i.e. data was collected right before the second dose was to be administered, and thus lowest peak for signature gene expression.

Of the 121 samples, 14 samples are from enrollment visit (visit 1) of the patient, categorized as NA, Naive and None. 107 samples collected were collected on visit 2: Pill - 34, Injection_NETEN - 44, Ring - 29 from visit 2.

* COC (combined oral contraceptive pill), 
* NET-EN (long-acting injectable norethisterone enanthate) 
* CCVR (etonorgesterol/ethinyl estradiol combined contraceptive vaginal ring)

***
# Work-Flow

Outline of the analysis:

* Load R packages 
* Why RNA Exome kit?
* Quality check of the raw reads
* Alignment of reads to the reference genome
* Obtain gene counts' files
* Generation of sample table
* QC filtering for low quality samples and non-targeted or lowly expressed genes.
* Differential expression analysis on filtered-data
* Inflammation analysis
* Intention-to-treat analysis

***
## Load Libraries

Load R packages required by different code chunks
```{r load_libraries, eval=FALSE}
# libraries
library(DESeq2)

# for merge
library(lessR)
library(tibble)

# for join
library(plyr)
# Use join for merging data frames by a column other than row names.

# for ggdraw
library(cowplot)

# for ggplot
library(ggplot2)

# for grid Arrange
library(gridExtra)
library(grid)
library(ggpubr)

# for melt
library(reshape2)
# for violin plot
library(ggforce)

# for RLE plot
library(EDASeq)

# for labels in violin plot
library(ggrepel)

# venn
library(VennDiagram)
library(gridExtra)

# venn
library(venn)

# for %>%
library(magrittr)

# for str_match
library(stringr)

# Heatmap function
source("/yerkes-cifs/runs/MSigDB/Heatmap_finalv2.R")
```

## RNA Exome library prep kit

The Illumina RNA exome library kit was used to prepare the libraries of the samples. This kit is helpful to sequence low-quality or degraded samples. Here's the explanation by Sydney on how she decided to go ahead with this particular kit and what was done.

"The reason we went with the RNA Exome kit was because it can handle samples with this amount of degradation. They assume high quality RNA for the RNA Exome kit has 77% of fragments longer than 200 base pairs, and can even go as low as 30% of samples above 200bp (this would be REALLY low quality). I’ve included a picture of the quality guidelines from the RNA Exome kit (on the left) next to an image of the bioanalyzer traces from some of the pre/post ECHO samples (on the right) so you can compare - you can see that although the samples are degraded and you don’t see those peaks for 28S and 18S rRNA, most of the fragments are longer than 200bp (there’s a ladder at the bottom of each sample trace and the 2nd tick is 200bp). 

N/A for the RIN score means that the bioanalyzer couldn’t give the sample a RIN score, so for those samples I looked at the bioanlayzer trace directly to make sure it was ok to use (by ensuring that at least 80% of the fragments were longer than 200bp). The units in the yield column are in nanograms. We did a pilot with the kit in January to make sure that the coverage across transcripts looked good and it did - I think we already had a call about the results of the pilot a while back - I’ve attached the powerpoint of the results from the pilot if you want look over the results again. 

The spreadsheet I sent was from the pre/post samples, not the case/controls, so they have already been processed using the RNA Exome kit. I sent the spreadsheet because it looks like we might want to rerun some of the samples along with the case/control samples that we just received 3 weeks ago, so Steve told me to send a list of samples with their remaining RNA quantities so that we can plan the batching for the case/control samples. I’ve just finished extracting the RNA from the case/control samples, and I’m currently QC’ing them. So far they look the same as the RNA from the pre/post samples, so the RNA Exome kit should work for the case/control samples as well." - Sydney

## Quality check using FastQC

FastQC was used to perform quality control on the samples. The results for lanes 1- 7 are in Unaligned_lane1/ and Unaligned_lanes_2-7/ folders respectively. The FastQC statistics were merged for all 121 samples and saved in **uCHOOSE_121samples_FastQCStats.txt**. Also, PICARD was run on the FastQC output files (*_fastqc.zip) to generate summary statistics.

**For each of these 121 samples, ~99% of the reads have a Q score >30.**

## Alignment of reads to the reference genome

STARv2.5 was used to align reads to the GRCh38 human genome. The alignments were run separately for lane1 and lanes2-7 in the Unaligned\*/ directories. The mapping statistics generated by STAR were merged for all 121 samples and saved in **uCHOOSE_121samples_mapping_stats.txt**. uCHOOSE_121samples_wSampleIDs_ReadsPerGene.txt contains the ReadsPerGene statistics for each samples, plus an additional column with SampleIDs.

Mapping statistics from the STAR-based alignment of reads to the reference human genome.
```{r MappingStats, message=FALSE, warning=FALSE, eval=FALSE}
# setwd("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/")
# Replaced spaces in header with '_', and removed percentage symbol from numbers

library(ggplot2)
library(reshape2)

statsReads <- read.table("uCHOOSE_121samples_wSampleIDs_ReadsPerGene.txt", header = TRUE, row.names = 1, check.names = FALSE, sep = "\t")

# DT table = interactive
#library(DT)
#datatable(statsReads, options = list(pageLength = 12))
#statsReads$sample <- NULL
#statsReads$sample <- row.names(statsReads)

# updated code
library(ggplot2)
library(reshape2)

statsReads <- read.table("uCHOOSE_121samples_wSampleIDs_ReadsPerGene.txt", header = TRUE, row.names = 1, check.names = FALSE, sep = "\t")
# DT table = interactive
#library(DT)
#datatable(statsReads, options = list(pageLength = 12))
#statsReads[order(statsReads$host),]
#statsReads$sample <- NULL
#statsReads$sample <- row.names(statsReads)
head(statsReads)
mappingTable <- statsReads[, c(1, 5, 6, 7, 8, 2)]
mappingTable <- melt(mappingTable, id.var="sample")
mappingTable$sample <- factor(mappingTable$sample, levels = statsReads$sample)

ggplot(mappingTable, aes(x = sample, y = value, fill = variable)) + 
  geom_bar(stat = "identity") + 
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5), axis.text.x = element_text(angle = 90, hjust = 1, size = 6)) +
  #scale_y_continuous(limits = c(0,100), breaks = seq(0,100, by = 20)) +
  #scale_x_discrete(labels = mappingTable$sample) +
  labs(x="Sample", y="Distribution of reads", title="uCHOOSE study - Mapping statistics") 
#ggsave("uCHOOSE_121samples_MappingStats_label90degree.pdf", width = 18, height = 10)

# histogram
# removed 'ID' from header, that is first element in the first row.
mappingStats <- read.table("uCHOOSE_121samples_mapping_stats.txt", row.names = 1, header = TRUE, sep = "\t", check.names = FALSE)
head(mappingStats)
mappingStats$`Uniquely mapped reads %` <- as.numeric(gsub("%", "", mappingStats$`Uniquely_mapped_reads%`))

pdf("uChoose_121_hist_uniq-mapped-reads.pdf", width = 8, height = 5)
hist(mappingStats$`Uniquely mapped reads %`, breaks = seq(0,100,by=5), labels = TRUE, xlab = "Uniquely mapped reads (%)", ylab = "No. of samples", main = "Histogram of uniquely mapped reads", col = "#FF6666", ylim = c(0,55))
dev.off()

# low number of reads
head(statsReads[order(statsReads$host),], n=11)
```

* low read samples
```{r low_read_samples, eval=FALSE, results='hide'}
head(statsReads[order(statsReads$host),], n=11)
#                                   sample     host PCYB ERCC ambiguous noFeature multimapping unmapped      cum
#18_Batch1_044_2  Sample11_GGCTAC_S11_L001  1251231    0    0     86614      7495       476098  4956381  6777819
#56_Batch1_063_1  Sample14_AGTTCC_S14_L004  2609082    0    0    141134      5795       915050  3313834  6984895
#33_Batch1_003_2  Sample15_ATGTCA_S15_L002  4315746    0    0    268737     19997      1028899  1685971  7319350
#91_Batch2_015_2  Sample22_CGTACG_S21_L006  4676854    0    0    287632     14522      1524502  4141754 10645264
#55_Batch1_042_2  Sample13_AGTCAA_S13_L004  5637672    0    0    391227     22016      2082188  2696816 10829919
#118_Batch2_096_1 Sample27_ATTCCT_S24_L007  6298090    0    0    351989     25162      1921794  3022278 11619313
#95_Batch2_037_2    Sample2_CGATGT_S2_L006  7073941    0    0    317418     39928      3450302  2441408 13322997
#4_Batch1_046_2     Sample6_GCCAAT_S6_L001  7929953    0    0    483285     21382      2601145   344180 11379945
#64_Batch2_016_2  Sample11_GGCTAC_S11_L004  9814859    0    0    730329     44543      2932058   194576 13716365
#66_Batch2_072_2  Sample21_GTTTCG_S20_L004  9836545    0    0    635879     16798      3569649   151678 14210549
#3_Batch1_018_2     Sample5_ACAGTG_S5_L001 10080981    0    0    597317     27469      1635854   338475 12680096

View(sampleTable)
as.character(sampleTable["18_Batch1_044_2",c("composite_feature")])
#"44_2_Pill"

# Print low-read samples with less than 10 million reads mapped to host
lowread_samples <- row.names(head(statsReads[order(statsReads$host),], n=10))

sampleTable[row.names(sampleTable) %in% lowread_samples, c("SubjectID","Visit_no","HC_type_actual_prior")]
```

## Obtain gene counts' files

Format gene counts files generated by STAR (STAR/SampleID/SampleID_ReadsPerGene.out.tab) to HTSeq format (contains two columns, namely gene names and gene counts):

* Script: **/illumina/genomes/STARCounts2htseq.pl**
* Input parameters to the script:
    + -c 1; It is an unstranded run, we will pick up the 1st numerical column in *ReadsPerGene.out.tab file.
    + -r; to remove ERCC entries
    + -t gene_name_table.txt; Convert NCBI Gene IDs to Gene symbols. DESeq prefers non-numbers at the beginning of a line. Located the gene_name_table.txt using the entry for option -GenomeDir while running STAR, from notes in README file 
* The script also modifies N_ to __ and removes ERCC entries.
* The output files which are HTSeq-format count files are saved as *alaHtseq.txt

Format count files
```{bash Format_count_files, eval = FALSE}
# /yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/ 

# Obtain list of 121 true samples as extra entries for non-existing entries were also created
awk '{print $2}' ../Unaligned_lanes_2-7/uCHOOSE/uCHOOSE_wSampleIDs_lanes_1-7_ReadsPerGene_sorted.txt > Samples_list_121.txt

# Count files with Gene symbols
# ../Unaligned_lanes_2-7/uCHOOSE/ directory contains fastq.gz for all samples, i.e. lanes 1-7. But, not all files are true samples, hence there are extra files in STAR alignment folder. We need generate count files for only 121 samples.
# for sid in `cat Samples_list_121.txt` ; do echo ${sid}; perl /illumina/genomes/STARCounts2htseq.pl -c 1 -r -t /yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt ../Unaligned_lanes_2-7/uCHOOSE/S2.5.2b_GRCh38_ERCC/${sid}/${sid}_ReadsPerGene.out.tab > ${sid}_alaHtseq.txt; done
# moved the above files to Htseq_gene-symbols/

# Count files with Ensembl Gene IDs
for sid in `cat Samples_list_121.txt` ; do echo ${sid}; perl /illumina/genomes/STARCounts2htseq.pl -c 1 -r ../Unaligned_lanes_2-7/uCHOOSE/S2.5.2b_GRCh38_ERCC/${sid}/${sid}_ReadsPerGene.out.tab > ${sid}_alaHtseq.txt; done

```

***
## Generation of sample table

* Details for some of the samples were updated on Jul2 (uCHOOSE_sampleID_and_info_CB.xlsx).
* Working with the uCHOOSE_SampleMetadataTable.xlsx file. Sheet 'sampleID_and_info_CB_edited' contains information copied from /yrknas.yerkes.emory.edu/yerkes/genomelab/share/BL_Projects/BL33_Bosinger_121uCHOOSE/'uCHOOSE_sampleID_and_info_CB.xlsx' 
* The mapping between SampleID and ExtractionID is done using uCHOOSE_SampleBarcodeIndexTable.txt
* Fields were input from these tables in the desired order, working shown in the rough sheets and saved as values in sheet 'SampleTable'. This table is saved as a text file **uCHOOSE_SampleTable.txt**.

```{r SampleTable, eval=FALSE}
#setwd("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/")
sampleTable <- read.table("uCHOOSE_SampleTable.txt", header = TRUE, row.names = 1, check.names = FALSE)

# Relevel the factors accordingly
sampleTable$Batch <- factor(sampleTable$Batch, levels = c("1", "2"))
sampleTable$Visit_no <- factor(sampleTable$Visit_no, levels = c("1", "2"))
sampleTable$Group <- factor(sampleTable$Group, levels = c("Pill", "Injection", "Ring", "Mismatch"))
sampleTable$HC_type_actual_prior <- factor(sampleTable$HC_type_actual_prior, levels = c("Pill", "Injection_NETEN", "Ring", "Naive", "None", "Unknown"))
sampleTable$composite_feature <- factor(paste(sampleTable$SubjectID, sampleTable$Visit_no, sampleTable$HC_type_actual_prior, sep = "_"))

# Pre-treatment as baseline
# ----------
sampleTable$Treatment <- sampleTable$HC_type_actual_prior
sampleTable$Treatment <- gsub("None", "Pretreatment", sampleTable$Treatment)
sampleTable$Treatment <- gsub("Naive", "Pretreatment", sampleTable$Treatment)
sampleTable$Treatment <- gsub("Unknown", "Pretreatment", sampleTable$Treatment)
sampleTable$Treatment <- factor(sampleTable$Treatment, levels = c("Pretreatment", "Injection_NETEN", "Pill", "Ring"))

# Display table in html output
# [1] Using kable
# library(knitr)
# library(kableExtra)
# kable(sampleTable) %>%
#  kable_styling(bootstrap_options = "striped", full_width = F)

# [2] Using DT
# DT table = interactive
library(DT)
datatable(sampleTable, options = list(pageLength = 21)) 

```

* Supertable containing RIN scores, mapping stats, Picard RNAseq metrics
```{r supertable, eval=FALSE}
# ----------
# uChoose generate super-table for the samples
# ----------
# Working on sampleTable where baseline samples are clubbed together as PreTreatment
uchoose_superTable <- sampleTable
superTable <- sampleTable
View(uchoose_superTable)
#row.names(superTable)
#head(rin_scores)

library(lessR)
library(tibble)

# Add RIN scores, and other sequencing details
rin_scores <- read.table("samples_rin_n_other_seq_details.txt", header = TRUE, row.names = NULL, check.names = FALSE, sep = "\t")
#colnames(rin_scores)
#colnames(superTable)
superTable <- Merge(superTable, rin_scores, by="Sample_ID")
row.names(superTable) <- superTable$Sample_ID
View(superTable)

# Add mapping statistics
statsReads <- read.table("uCHOOSE_121samples_wSampleIDs_ReadsPerGene.txt", header = TRUE, row.names = 1, check.names = FALSE, sep = "\t")
#row.names(statsReads)
statsReads$sample <- row.names(statsReads)
#mappingTable <- statsReads[, c("sample","ambiguous","noFeature","multimapping","unmapped","host")]

superTable <- Merge(superTable, statsReads, by="row.names")
row.names(superTable) <- superTable$Sample_ID
# remove redundant columns
superTable$Row.names <- NULL
superTable$sample <- NULL

superTable$PCT_uniq_host_mapped_reads <- superTable$host/superTable$cum*100

# Add PICARD/multiQC metrics
# detailed table
picard_rnaseqmetrics_det <- read.table("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_RNAseqMetrics/AlignmentPicard_stats/multiqc_data/multiqc_picard_RnaSeqMetrics.txt", header = TRUE, row.names = NULL, check.names = FALSE, sep = "\t")
# Change the Sample column to ExtractionID, so Merge can be done
picard_rnaseqmetrics_det$Sample <- gsub("_GRCh23_ERCC_sorted", "", picard_rnaseqmetrics_det$Sample)
names(picard_rnaseqmetrics_det)[names(picard_rnaseqmetrics_det)=="Sample"] <- "ExtractionID"
#colnames(picard_rnaseqmetrics_det)
picard_rnaseqmetrics_det <- picard_rnaseqmetrics_det[,c(1,2,3,4,5,6,7,18,19,20,21,22,25,26,27,32)]
superTable <- Merge(superTable, picard_rnaseqmetrics_det, by="ExtractionID")
row.names(superTable) <- superTable$Sample_ID
# remove redundant columns

head(superTable)
head(picard_rnaseqmetrics_det)
```

* Plots on the feature columns of supertable
```{r generate_plots_on_supertable, eval=FALSE}
# Violin plot
library(reshape2)
library(ggplot2)
library(ggforce)

colnames(superTable)
tmp <- head(sample_distr[order(sample_distr$host),], n=4)
head(sample_distr_melted)

#sample_distr <- superTable[,c("SampleIndex","RIN","PCT_CODING_BASES","PCT_uniq_host_mapped_reads", "host", "cum", "CODING_BASES")]
sample_distr <- superTable[,c("SampleIndex","host")]
#sample_distr <- tmp
names(sample_distr)[names(sample_distr) == "SampleIndex"] <- "sample"
sample_distr$sample <- gsub("I", "", sample_distr$sample)
sample_distr_melted <- melt(sample_distr, measure.vars = c("host"))

ggplot(sample_distr_melted, aes(x=variable, y=value)) + 
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5), axis.text.x = element_text(size = 8)) +
  labs(x="Factor", y="Percent variable", title="uCHOOSE study - Mapping Metrics") +
  geom_violin() +
  #geom_dotplot(binaxis='y', stackdir='center', dotsize=0.4)
  geom_jitter(shape=10, position=position_jitter(0.1)) +
  #geom_sina() +
  #geom_dotplot(binaxis='y', stackdir='center', dotsize=1) 
  #geom_dotplot(binaxis='y', stackdir='center', dotsize=0.4, binwidth = 3) +
  geom_boxplot(width=0.1)

# Generating stacked bar plot and histogram 
# ----------------
library(ggplot2)
library(reshape2)

superTable

statsReads <- read.table("uCHOOSE_121samples_wSampleIDs_ReadsPerGene.txt", header = TRUE, row.names = 1, check.names = FALSE, sep = "\t")
# DT table = interactive
#library(DT)
#datatable(statsReads, options = list(pageLength = 12))
#statsReads[order(statsReads$host),]
#statsReads$sample <- NULL
#statsReads$sample <- row.names(statsReads)
statsReads <- statsReads[order(statsReads$host),]
head(statsReads)
head(mappingTable, n = 11)
mappingTable <- statsReads

mappingTable$sample <- rownames(mappingTable)
mappingTable$sample <- lapply(strsplit(mappingTable$sample, "_"), '[[', 1)
mappingTable <- mappingTable[, c(1, 5, 6, 7, 8, 2)]
mappingTable <- mappingTable[order(mappingTable$host),]
mappingTable$sample <- factor(mappingTable$sample, levels = mappingTable$sample)
mappingTable <- melt(mappingTable, id.var="sample")
ggplot(mappingTable, aes(x = sample, y = value, fill = variable)) + 
  geom_bar(stat = "identity") + 
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5), axis.text.x = element_text(angle = 90, hjust = 1, size = 6)) +
  #scale_y_continuous(limits = c(0,100), breaks = seq(0,100, by = 20)) +
  #scale_x_discrete(labels = mappingTable$sample) +
  labs(x="Sample", y="Distribution of reads", title="uCHOOSE study - Mapping statistics") 
#ggsave("uCHOOSE_121samples_MappingStats_label90degree.pdf", width = 18, height = 10)

# histogram
# removed 'ID' from header, that is first element in the first row.
mappingStats <- read.table("uCHOOSE_121samples_mapping_stats.txt", row.names = 1, header = TRUE, sep = "\t", check.names = FALSE)
head(mappingStats)
mappingStats$`Uniquely mapped reads %` <- as.numeric(gsub("%", "", mappingStats$`Uniquely_mapped_reads%`))

pdf("uChoose_121_hist_uniq-mapped-reads.pdf", width = 8, height = 5)
hist(mappingStats$`Uniquely mapped reads %`, breaks = seq(0,100,by=5), labels = TRUE, xlab = "Uniquely mapped reads (%)", ylab = "No. of samples", main = "Histogram of uniquely mapped reads", col = "#FF6666", ylim = c(0,55))
dev.off()

# low number of reads
head(statsReads[order(statsReads$host),], n=13)

# for stacked bar plot
picard_rnaseqmetrics <- read.table("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_RNAseqMetrics/AlignmentPicard_stats/multiqc_data/mqc_picard_rnaseqmetrics_assignment_plot_1.txt", header = TRUE, row.names = NULL, check.names = FALSE, sep = "\t")
picard_rnaseqmetrics$Sample <- gsub("_GRCh23_ERCC_sorted", "", picard_rnaseqmetrics$Sample)
names(picard_rnaseqmetrics)[names(picard_rnaseqmetrics)=="Sample"] <- "ExtractionID"

# Histogram for RIN scores
#mappingStats$`Uniquely mapped reads %` <- as.numeric(gsub("%", "", mappingStats$`Uniquely_mapped_reads%`))

#pdf("uChoose_121_hist_uniq-mapped-reads.pdf", width = 8, height = 5)
hist(superTable$RIN, 
     breaks = seq(0,60,by=1), 
     labels = TRUE, xlab = "RIN score", ylab = "No. of samples", main = "Histogram of RIN scores", 
     col = "#FF6666" 
#     ylim = c(0,55)
     )
#dev.off()

```

## Preliminary analysis to check which samples/genes to remove.
```{r prelim_low_qual_samples, eval=FALSE}
superTable[order(superTable$MEDIAN_CV_COVERAGE, decreasing = TRUE),]

# Removed samples based on following criteria:
# mapping stats, host < 5 million
# RIN scores <= 2
# RLE plots
# Median CV coverage 1.3
# RLE var 
as.numeric(superTable$RIN)
superTable$host
superTable$MEDIAN_CV_COVERAGE
samples_with_lowRIN <- superTable[as.numeric(superTable$RIN) <= 2,]$Sample_ID
samples_with_lowhostreads <- superTable[superTable$host <= 5000000,]$Sample_ID
samples_with_highmedCV <- superTable[superTable$MEDIAN_CV_COVERAGE > 1.3,]$Sample_ID

samples_with_lowRIN
low_qual_samples_v2 <- union(samples_with_lowRIN, union(samples_with_lowhostreads, samples_with_highmedCV))
length(low_qual_samples_v2)
# 9 samples

low_qual_samples_v3 <- union(low_qual_samples_v2, samples_with_highRLEvar)
length(low_qual_samples_v3)

# 12 samples
colnames(superTable_metadata)
length(rownames(superTable_metadata))
length(superTable_metadata[(superTable_metadata$Treatment == "Pretreatment"), c("1st_HC_assigned")])

superTable_metadata$Assigned_study_arm
a <- superTable_metadata[superTable_metadata$Sample_ID %in% low_qual_samples_v3,c("Sample_ID", "1st_HC_assigned", "Treatment", "RIN", "host", "cum", "CODING_BASES", "MEDIAN_CV_COVERAGE", "Infl_category_Kmeans")]
write.table(a, "details_low_qual_samples.txt", row.names = FALSE, quote = FALSE, col.names = TRUE, sep = "\t")

superTable_filtered_v2 <- superTable[!(superTable$Sample_ID %in% low_qual_samples_v2),]
# to arrange cols so that DESeq can take it in
superTable_filtered_v2$ExtractionID_copy <- superTable_filtered_v2$ExtractionID
superTable_filtered_v2$ExtractionID <- NULL

# obtained after removing samples with high RLE from superTable_filtered_v2
superTable_filtered_v3 <- superTable_filtered_v2[!(superTable_filtered_v2$Sample_ID %in% low_qual_samples_v3),]
head(superTable_filtered_v3)

target_genes <- scan("/yerkes-cifs/runs/Genome_references/homo_sapiens/TruSeq_RNA_Exome_kit/probe-targets/probe_exon_overlaps_bedtools_pcg-names_hg38.txt", what = "character")
# 19543 target genes which are protein-coding, obtained using bedtools on probe coordinates and exon coordinates of hg38

# * remove genes with low expression
# * forward counts table to Christina
# * get inflammation data

library(DESeq2)
dds1q <- DESeqDataSetFromHTSeqCount(superTable_filtered_v2, ".", design =~ Treatment)
dds1q_copy <- dds1q
dds1q <- dds1q[row.names(dds1q) %in% target_genes,]
dds1q <- DESeq(dds1q)
vsd1q <- varianceStabilizingTransformation(dds1q, blind = FALSE, fitType = "parametric")
#rld1q <- rlog(dds1q, blind = FALSE, fitType = "parametric", betaPriorVar = 10)


head(target_genes)
# having removed 12 samples
dds1r <- DESeqDataSetFromHTSeqCount(superTable_filtered_v3, ".", design =~ Treatment)
dds1r_copy <- dds1r
dds1r <- dds1r[row.names(dds1r) %in% target_genes,]
dds1r <- DESeq(dds1r)
vsd1r <- varianceStabilizingTransformation(dds1r, blind = FALSE, fitType = "parametric")
#rld1r <- rlog(dds1r, blind = FALSE, fitType = "parametric", betaPriorVar = 10)

library(ggplot2)
library(EDASeq)
plotPCA(vsd1r, intgroup = c("Treatment"))

# Generate RLE/PCA for filtered set
sample_details <- colData(dds1q)
sample_order <- rownames(sample_details[order(sample_details$Treatment),])
sample_details[order(sample_details$Treatment),]$Treatment
uchoose_raw_counts <- counts(dds1q)[, sample_order]
uchoose_norm_counts <- counts(dds1q, normalize = TRUE)[,sample_order]

colnames(uchoose_norm_counts) <- lapply(strsplit(colnames(uchoose_norm_counts), "_"), '[[', 1)
colnames(uchoose_raw_counts) <- lapply(strsplit(colnames(uchoose_raw_counts), "_"), '[[', 1)

colnames(uchoose_norm_counts)
par(mar=c(3,3,3,3))
# On raw counts
par(mfrow = c(1, 2), las=2, cex.axis = 0.4)
plotRLE(uchoose_raw_counts, outline=FALSE, 
        ylim=c(-4, 4), col=as.numeric(sample_details[order(sample_details$Treatment),]$Treatment), 
        main = 'RLE\nRaw counts')
plotPCA(uchoose_raw_counts, col = as.numeric(sample_details[order(sample_details$Treatment),]$Treatment), 
        main = 'PCA\nRaw counts')

# FINAL RLE plot, ordered by Treatment 
# On normalized counts
par(mfrow = c(1, 2), las=2, cex.axis = 0.4)
plotRLE(uchoose_norm_counts, outline=FALSE, 
        ylim=c(-4, 4), col=as.numeric(sample_details[order(sample_details$Treatment),]$Treatment), 
        main = 'RLE\nNorm counts')
plotPCA(uchoose_norm_counts, col = as.numeric(sampleTable$Treatment), main = 'PCA\nNorm counts')

par(mfrow = c(1, 1), las=2, cex.axis = 0.4)
plotRLE(uchoose_norm_counts[!(colnames(uchoose_norm_counts) %in% c(15,90,46)),], outline=FALSE, 
        ylim=c(-4, 4), col=as.numeric(sample_details[order(sample_details$Treatment),]$Treatment), 
        main = 'RLE\nNorm counts')
abline(h=c(-2,2))
# samples 15, 90, 46

# SAMPLE outliers based on RLE plots
samples_with_highRLEvar <- c("15_Batch1_167_2", "90_Batch2_163_2", "46_Batch1_143_2")

# samples 15, 90, 46
uchoose_norm_counts_edit <- counts(dds1q, normalize = TRUE)[,sample_order]
uchoose_norm_counts_edit <- uchoose_norm_counts_edit[,!(colnames(uchoose_norm_counts_edit) %in% samples_with_highRLEvar)]
colnames(uchoose_norm_counts_edit) <- lapply(strsplit(colnames(uchoose_norm_counts_edit), "_"), '[[', 1)
sample_details_edit <- sample_details[!(rownames(sample_details) %in% samples_with_highRLEvar),]

# RLE plot after removing highRLEvar samples
par(mfrow = c(1, 1), las=2, cex.axis = 0.4)
plotRLE(uchoose_norm_counts_edit, outline=FALSE, 
        ylim=c(-4, 4), col=as.numeric(sample_details_edit[order(sample_details_edit$Treatment),]$Treatment), 
        main = 'RLE\nNorm counts')
abline(h=c(-2,2))



#PCA
PCA_plot <- function(pcaData, main_title, filename, category, legend_title) {
  percentVar <- round(100 * attr(pcaData, "percentVar"))
  ggplot(pcaData, aes(PC1, PC2)) +
    #scale_color_manual(name="feature", values = c("Control"="#fb8072", "Factor"="#0571b0")) +
    #scale_shape_manual(name="Batch", values = c("1"=16, "2"=17)) +
    theme_bw() +
    theme(axis.text.x = element_text(size = 7), axis.text.y = element_text(size = 7), plot.title = element_text(hjust = 0.5), panel.grid.minor = element_blank(), strip.background = element_blank(), strip.text.y = element_text(size=7), legend.key.size = unit(1, 'lines'), legend.text=element_text(size=7)) + 
    #geom_point(aes(color = feature1, shape = feature2), size=3) +
    geom_point(aes(color = category), size=3) +
    labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance"), 
         title=main_title, color = legend_title) + 
    coord_fixed() 
#    ggsave(filename)
}

vsd1p$Treatment 

pcaData <- plotPCA(vsd1r, intgroup=c("Treatment"), returnData=TRUE)  
plot <- PCA_plot(pcaData, main_title = "Filtered data set", filename = "PCA_109-samples_19543-genes_treatment.pdf", category = vsd1r$Treatment, legend_title = "Treatment")
library(cowplot)
ggdraw(plot)
# saved screenshot and not through filename

# filtering out lowly expressed genes
uchoose_raw_counts <- counts(dds1r)
# get the total counts for a gene across all samples
allgenes_rcnt_tot <- rowSums(uchoose_raw_counts)
allgenes_rcnt_mean <- rowMeans(uchoose_raw_counts)

plot(sort(allgenes_rcnt_mean[allgenes_rcnt_mean < 20]))
length(allgenes_rcnt_mean[allgenes_rcnt_mean > 5])

# get the list of genes with mean counts > 5
filtered_rcnt_target_genes <- names(allgenes_rcnt_mean[allgenes_rcnt_mean > 5])
length(filtered_rcnt_target_genes)
# 14226 genes

filtered_rcnt_target_genes

# --------
# FINAL DATA-SET
# --------
# 109 samples, 14226 genes which are protein-coding, probe-targets with raw mean >5 across all samples
dds1s <- DESeqDataSetFromHTSeqCount(superTable_filtered_v3, ".", design =~ Treatment)
dds1s_copy <- dds1s
dds1s <- dds1s[row.names(dds1s) %in% filtered_rcnt_target_genes,]
dds1s <- DESeq(dds1s)
vsd1s <- varianceStabilizingTransformation(dds1s, blind = FALSE, fitType = "parametric")
#rld1s <- rlog(dds1s, blind = FALSE, fitType = "parametric", betaPriorVar = 10)

pcaData <- plotPCA(vsd1s, intgroup=c("Treatment"), returnData=TRUE)  
plot <- PCA_plot(pcaData, main_title = "Filtered data set", filename = "PCA_109-samples_14226-genes_treatment.pdf", category = vsd1s$Treatment, legend_title = "Treatment")

ggdraw(plot)

# --------
# Prepare files to send
# --------

gene_name_mapping <- read.table("/yerkes-cifs/runs/Genome_references/homo_sapiens/TruSeq_RNA_Exome_kit/probe-targets/gene_name_mapping_for-targeted-pcg_hg38.txt", header = TRUE, row.names = NULL, check.names = FALSE)

raw_counts_1s <- as.data.frame(counts(dds1s))
norm_counts_1s <- as.data.frame(counts(dds1s, normalize=TRUE))
norm_counts_1s <- round(norm_counts_1s, digits = 3)

raw_counts_1s$EnsemblID <- rownames(raw_counts_1s)
raw_counts_1s$GeneSymbol <- sapply(raw_counts_1s$EnsemblID, function(i) as.character(gene_name_table[i,]))

norm_counts_1s$EnsemblID <- rownames(norm_counts_1s)
norm_counts_1s$GeneSymbol <- sapply(norm_counts_1s$EnsemblID, function(i) as.character(gene_name_table[i,]))

library(dplyr)
raw_counts_1s <- raw_counts_1s %>% select(GeneSymbol, everything())
raw_counts_1s <- raw_counts_1s %>% select(EnsemblID, everything())

norm_counts_1s <- norm_counts_1s %>% select(GeneSymbol, everything())
norm_counts_1s <- norm_counts_1s %>% select(EnsemblID, everything())

norm_counts_1s
# Add a row of gene-names in the raw counts
superTable
# highlight low-quality samples


write.table(raw_counts_1s, "uChoose_raw-counts_filtered-data.txt", row.names = FALSE, quote = FALSE, col.names = TRUE, sep = "\t")
write.table(norm_counts_1s, "uChoose_norm-counts_filtered-data.txt", row.names = FALSE, quote = FALSE, col.names = TRUE, sep = "\t")
write.table(superTable, "uChoose_super-table.txt", row.names = FALSE, quote = FALSE, col.names = TRUE, sep = "\t")

write.table(low_qual_samples_v3, "uChoose_list_of_low-quality-samples.txt", row.names = FALSE, quote = FALSE, col.names = FALSE, sep = "\t")

```

## Filtered-data: 109 samples, 14226 genes

### Low-quality samples

* RIN <= 2
* No. of reads uniquely mapped to the host <=5000000
* Haywire gene coverage, median CV coverage > 1.3
* Outliers in RLE, RLE var <-2 or >2

```{r define_low_qual_samples, eval=FALSE}
# Removed samples based on following criteria:
samples_with_lowRIN <- superTable[as.numeric(superTable$RIN) <= 2,]$Sample_ID
samples_with_lowhostreads <- superTable[superTable$host <= 5000000,]$Sample_ID
samples_with_highmedCV <- superTable[superTable$MEDIAN_CV_COVERAGE > 1.3,]$Sample_ID
samples_with_highRLEvar <- c("15_Batch1_167_2", "90_Batch2_163_2", "46_Batch1_143_2")

low_qual_samples_v2 <- union(samples_with_lowRIN, union(samples_with_lowhostreads, samples_with_highmedCV))
low_qual_samples_v3 <- union(low_qual_samples_v2, samples_with_highRLEvar)
length(low_qual_samples_v3)
# 12 samples
```

### Final set of genes

* Probe-target genes found using bedtools
* protein-coding genes based on GTF annotation file
* Mean raw counts across all samples >5

```{r filter_lowly_expressed_genes, eval=FALSE}

gene_protein_coding <- scan("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/GRCh38_protein_coding_genes.txt", what = "character")

target_genes <- scan("/yerkes-cifs/runs/Genome_references/homo_sapiens/TruSeq_RNA_Exome_kit/probe-targets/probe_exon_overlaps_bedtools_pcg-names_hg38.txt", what = "character")
# 19543 target genes which are protein-coding, obtained using bedtools on probe coordinates and exon coordinates of hg38

superTable_filtered_v2 <- superTable[!(superTable$Sample_ID %in% low_qual_samples_v2),]
# to arrange cols so that DESeq can take it in
superTable_filtered_v2$ExtractionID_copy <- superTable_filtered_v2$ExtractionID
superTable_filtered_v2$ExtractionID <- NULL

superTable_filtered_v3 <- superTable_filtered_v2[!(superTable_filtered_v2$Sample_ID %in% low_qual_samples_v3),]
nrow(superTable_filtered_v3)

# having removed 12 samples
dds1r <- DESeqDataSetFromHTSeqCount(superTable_filtered_v3, ".", design =~ Treatment)
dds1r_copy <- dds1r
dds1r <- dds1r[row.names(dds1r) %in% target_genes,]
dds1r <- DESeq(dds1r)
vsd1r <- varianceStabilizingTransformation(dds1r, blind = FALSE, fitType = "parametric")
#rld1r <- rlog(dds1r, blind = FALSE, fitType = "parametric", betaPriorVar = 10)

# raw counts
uchoose_raw_counts <- counts(dds1r)
uchoose_norm_counts <- counts(dds1r, normalize=TRUE)

a <- data.frame("Gene_EnsemblID" = rownames(uchoose_raw_counts), uchoose_raw_counts)
colnames(a) <- gsub("^X", "", colnames(a))
a$GeneSymbol <- sapply(as.character(a$Gene_EnsemblID), function(i) as.character(gene_name_table[i,]))
b <- cbind(a$Gene_EnsemblID, a$GeneSymbol, a[,2:110])
head(b[,1:5])
write.table(b, file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/UChoose_RawCounts_pre-analysis_19543-genes.txt", sep="\t", quote=FALSE, row.names = FALSE)


a <- data.frame("Gene_EnsemblID" = rownames(uchoose_norm_counts), uchoose_norm_counts)
colnames(a) <- gsub("^X", "", colnames(a))
a$GeneSymbol <- sapply(as.character(a$Gene_EnsemblID), function(i) as.character(gene_name_table[i,]))
b <- cbind(a$Gene_EnsemblID, a$GeneSymbol, a[,2:110])
head(b[,1:5])
write.table(b, file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/UChoose_NormCounts_pre-analysis_19543-genes.txt", sep="\t", quote=FALSE, row.names = FALSE)


# filtering out lowly expressed genes
# get the total counts for a gene across all samples
allgenes_rcnt_tot <- rowSums(uchoose_raw_counts)
allgenes_rcnt_mean <- rowMeans(uchoose_raw_counts)
allgenes_rcnt_mean

allgenes_rcnt_tot["ENSG00000112486"]
allgenes_rcnt_tot["ENSG00000184451"]
allgenes_rcnt_mean["ENSG00000112486"]
allgenes_rcnt_mean["ENSG00000184451"]

plot(sort(allgenes_rcnt_mean[allgenes_rcnt_mean < 20]))
length(allgenes_rcnt_mean[allgenes_rcnt_mean > 5])

# get the list of genes with mean counts > 5
filtered_rcnt_target_genes <- names(allgenes_rcnt_mean[allgenes_rcnt_mean > 5])
length(filtered_rcnt_target_genes)
# 14226 genes

```

### Include metadata in the filtered data-set, 109 samples, 14226 genes
```{r include_metadata, eval=FALSE}
# import metadata that contain BV and inflammation data

metadata <- read.table("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Sep2020/Metadata_v2_withSampledetails.txt", header = TRUE, row.names = NULL, check.names = FALSE, sep = "\t")

#metadata_v2 <- read.table("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/Metadata_v3_withSampledetails_n_InfPAM.txt", header = TRUE, row.names = NULL, check.names = FALSE, sep = "\t")

library(plyr)
superTable_metadata <- join(superTable, metadata[,c(2,17:51)], by="Sample_ID")
row.names(superTable_metadata) <- superTable_metadata$Sample_ID

superTable_metadata_filtered <- superTable_metadata[!(superTable_metadata$Sample_ID %in% low_qual_samples_v3),]
# bring extraction col to te last
superTable_metadata_filtered$ExtractionID_copy <- superTable_metadata_filtered$ExtractionID
superTable_metadata_filtered$ExtractionID <- NULL

superTable_metadata_filtered$BV <- factor(superTable_metadata_filtered$BV, levels = c("Neg","Int","Pos"))
superTable_metadata_filtered$Infl_category_Kmeans <- factor(superTable_metadata_filtered$Infl_category_Kmeans, levels = c("Low", "High"))

length(superTable_metadata_filtered[(superTable_metadata_filtered$Treatment == "Injection_NETEN") 
                             & (superTable_metadata_filtered$BV_prior == "Int")
                             & (superTable_metadata_filtered$BV == "Pos"),"BV"])

length(superTable_metadata_filtered[(superTable_metadata_filtered$Treatment == "Pill") 
                             & (superTable_metadata_filtered$BV_prior == "Int")
                             & (superTable_metadata_filtered$BV == "Pos"),"BV"])

length(superTable_metadata_filtered[(superTable_metadata_filtered$Treatment == "Ring") 
                             & (superTable_metadata_filtered$BV_prior == "Int")
                             & (superTable_metadata_filtered$BV == "Pos"),"BV"])

```


### Differential expression analysis on filtered-data

#### Run DESeq, get counts table, 109 samples, 14226 genes
```{r run_DESeq_109samples_14226genes, eval=FALSE}
# 109 samples

library(DESeq2)

dds1t <- DESeqDataSetFromHTSeqCount(superTable_metadata_filtered, ".", design =~ Treatment)
dds1t_copy <- dds1t
dds1t <- dds1t[row.names(dds1t) %in% filtered_rcnt_target_genes,]
dds1t <- DESeq(dds1t)
vsd1t <- varianceStabilizingTransformation(dds1t, blind = FALSE, fitType = "parametric")
rld1t <- rlog(dds1t, blind = FALSE, fitType = "parametric", betaPriorVar = 10)

par_dir <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/"

gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)
# ------------
atable <- superTable_metadata_filtered
length(atable[atable$Treatment == "Pretreatment" & atable$CST == "CST-I","CST"])
length(atable[atable$Treatment == "Pretreatment" & atable$CST == "CST-III","CST"])
length(atable[atable$Treatment == "Pretreatment" & atable$CST == "CST-IV","CST"])
# ------------

# Write raw counts to a file
a <- data.frame("Gene_EnsemblID" = rownames(counts(dds1t)), counts(dds1t))
colnames(a) <- gsub("^X", "", colnames(a))
a$GeneSymbol <- sapply(as.character(a$Gene_EnsemblID), function(i) as.character(gene_name_table[i,]))
write.table(a, file=paste0(par_dir, "uChoose_Analysis_Oct2020/data_raw_counts.txt"), 
            sep="\t", quote=FALSE, row.names = FALSE)
# NOTE: the counts files have gene symbols as the last cols

# Write normalized counts to a file
b <- data.frame("Gene_EnsemblID" = rownames(counts(dds1t)), round(counts(dds1t, normalize = TRUE), digits = 3))
colnames(b) <- gsub("^X", "", colnames(b))
b$GeneSymbol <- sapply(as.character(a$Gene_EnsemblID), function(i) as.character(gene_name_table[i,]))
write.table(b, file=paste0(par_dir, "uChoose_Analysis_Oct2020/data_normalized_counts.txt"), 
            sep="\t", quote=FALSE, row.names = FALSE)

# Write rlog data to a file
c <- data.frame("Gene_EnsemblID" = rownames(assay(rld1t)), round(assay(rld1t), digits = 3))
colnames(c) <- gsub("^X", "", colnames(c))
write.table(c, file=paste0(par_dir, "uChoose_Analysis_Oct2020/data_rlog.txt"), 
            sep="\t", quote=FALSE, row.names = FALSE)

# Write vst data to a file
d <- data.frame("Gene_EnsemblID" = rownames(assay(vsd1t)), round(assay(vsd1t), digits = 3))
colnames(d) <- gsub("^X", "", colnames(d))
write.table(d, file=paste0(par_dir, "uChoose_Analysis_Oct2020/data_vst.txt"), 
            sep="\t", quote=FALSE, row.names = FALSE)


# DESeq Results
# Injection vs. Pre
res1t_InjectionvsPre <- results(dds1t, contrast = c("Treatment", "Injection_NETEN", "Pretreatment"), independentFiltering = FALSE)
summary(res1t_InjectionvsPre, alpha = 0.05)
# up - 65, down - 17

# Ring vs. Pre
res1t_RingvsPre <- results(dds1t, contrast = c("Treatment", "Ring", "Pretreatment"), independentFiltering = FALSE)
summary(res1t_RingvsPre, alpha = 0.05)
# up - 334, down - 66

# Pill vs. Pre
res1t_PillvsPre <- results(dds1t, contrast = c("Treatment", "Pill", "Pretreatment"), independentFiltering = FALSE)
summary(res1t_PillvsPre, alpha = 0.05)
# up - 244, down - 38

# Injection vs. Pill
res1t_InjectionvsPill <- results(dds1t, contrast = c("Treatment", "Injection_NETEN", "Pill"), independentFiltering = FALSE)
summary(res1t_InjectionvsPill, alpha = 0.05)
# up - 1, down - 13

# Injection vs. Ring
res1t_InjectionvsRing <- results(dds1t, contrast = c("Treatment", "Injection_NETEN", "Ring"), independentFiltering = FALSE)
summary(res1t_InjectionvsRing, alpha = 0.05)
# up - 94, down - 134

# Ring vs. Pill
res1t_RingvsPill <- results(dds1t, contrast = c("Treatment", "Ring", "Pill"), independentFiltering = FALSE)
summary(res1t_RingvsPill, alpha = 0.05)
# up - 114, down - 64

# Ring vs. Injection
res1t_RingvsInjection <- results(dds1t, contrast = c("Treatment", "Ring", "Injection_NETEN"), independentFiltering = FALSE)
summary(res1t_RingvsInjection, alpha = 0.05)
# up - 134, down - 94

# Pill vs. Injection
res1t_PillvsInjection <- results(dds1t, contrast = c("Treatment", "Pill", "Injection_NETEN"), independentFiltering = FALSE)
summary(res1t_PillvsInjection, alpha = 0.05)
# up - 13, down - 1

# Pill vs. Ring 
res1t_PillvsRing <- results(dds1t, contrast = c("Treatment", "Pill", "Ring"), independentFiltering = FALSE)
summary(res1t_PillvsRing, alpha = 0.05)
# up - 64, down - 114

# merge DESeq results into one table

library(lessR)
library(tibble)

# Merge results by gene names
merge_df <- function(resA, resB) {
  merged_results <- Merge(resA, resB, by="row.names")
  row.names(merged_results) <- merged_results$Row.names
  merged_results[,"Row.names"] <- NULL
  return(merged_results)
}

# vector with all results names 
res_vector <- c("res1t_InjectionvsPre", "res1t_PillvsPre", "res1t_RingvsPre",
                "res1t_RingvsInjection", "res1t_PillvsInjection",
                "res1t_InjectionvsPill", "res1t_RingvsPill",
                "res1t_PillvsRing", "res1t_InjectionvsRing")
# first result
resP <- get(res_vector[1])
res_ext <- unlist(strsplit(res_vector[1], "1t"))[2]
colnames(resP) <- gsub("$", res_ext, colnames(resP))

head(resP)
# merge results 
for (r in 2:length(res_vector)) {
  resQ <- get(res_vector[r])
  res_ext <- unlist(strsplit(res_vector[r], "1t"))[2]
  colnames(resQ) <- gsub("$", res_ext, colnames(resQ))
  merged_resPQ <- merge_df(resP, resQ)
  resP <- merged_resPQ
}
res_merged <- resP

head(res_merged)
res_merged$EnsemblID <- row.names(res_merged)

# Add a column with gene symbols
# gene name table: Ensembl gene id as rownames, gene symbol 1st col
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)
res_merged$GeneSymbol <- sapply(res_merged$EnsemblID, function(i) as.character(gene_name_table[i,]))

write.table(res_merged, file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/DESeq_results_all.txt", sep="\t", quote=FALSE, row.names = FALSE)

```

#### PCA, 109 samples, 14226 genes
```{r PCA_109samples, eval=FALSE}

unique(superTable_metadata_filtered[order(superTable_metadata_filtered$BV),c("BV","CST")])
vsd1t_copy <- vsd1t
vsd1t_copy$BV <- factor(vsd1t_copy$BV, levels = c("Neg", "Int", "Pos"))

pcaData <- plotPCA(vsd1t_copy, intgroup=c("Treatment","BV", "CST"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData, aes(PC1, PC2, color = Treatment, group = Treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  #geom_point(aes(color = Treatment, shape = CST), size =3) +
  geom_point(aes(color = Treatment, shape = CST, size = BV)) +
  labs(title = "UCHOOSE study", color = "legend") +
  scale_colour_manual(name = "Treatment", 
                      labels = c("Pretreatment", "Injection_NETEN","Pill","Ring"),
                      values = c("#636363","#b2182b","#008000","#2166ac")) +
  #stat_ellipse()
  #scale_shape_manual(name = "BV", 
  #                    labels = c("Int", "Neg", "Pos"),
  #                    values = c(19, 1, 15)) 
  scale_shape_manual(name = "CST", 
                      labels = c("CST-I", "CST-III", "CST-IV", "NA"),
                      values = c(19, 17, 15, 1)) +
  scale_size_manual(name = "BV", 
                      labels = c("Neg", "Int", "Pos"),
                      values = c(4,6,6.5)) 

#ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Sep2020/PCA_109samples14226genes_Treatment.pdf", height = 6, width = 10)

ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_109samples14226genes_Treatment_BV_CST.pdf", height = 6, width = 10)



```

#### Volcano plots, 109 samples, 14226 genes
```{r VolcanoPlots_109samples_14226genes, eval=FALSE}
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)

fc <- 0.58
modify_result <- function(res){
  res$gene <- row.names(res)
  res$negLogPval <- -log10(res$pvalue)
  # threshold to define significant DEGs
  res <- res[!(is.na(res$log2FoldChange) | is.na(res$pvalue) | is.na(res$padj)),]
  res$sigGenes <- ifelse((res$padj < 0.05 & abs(res$log2FoldChange) > fc), 1, 0)
  res$deg <- ifelse((res$padj < 0.05 & res$log2FoldChange > fc), 1, ifelse((res$padj < 0.05 & res$log2FoldChange < -fc), -1, 0))
  res <- res[res$lfcSE < 1,]
  #res$gene_symbol <- sapply(res$gene, function(i) as.character(gene_name_table[i,]))
  return(res)
}

volcano_plot <- function(res, pline, main_title){
  ggplot(res, aes(x=log2FoldChange, y=negLogPval)) + 
    scale_color_manual(name="sigGenes", breaks = c(1,0), values = c("1"="red", "0"="black"), labels = c("DEGs", "not significant")) +
    scale_x_continuous(limits = c(-7,7), breaks = c(-6,-4,-2,0,2,4,6)) +
    scale_y_continuous(limits = c(-1,17), breaks = c(0,pline,5,10,15,20)) +
    geom_point(aes(color=as.factor(sigGenes)), size = 1) +
    theme_bw() +
    theme(axis.text = element_text(size = 9), axis.title = element_text(size = 12), plot.title = element_text(hjust = 0.5, size = 12), 
          panel.grid.minor = element_blank(), 
          legend.title = element_blank(), legend.position = c(1,1), legend.justification = c(1,1), 
          legend.background = element_blank(), legend.key = element_blank()) +
    geom_vline(xintercept = fc, linetype = "dotdash", color = "#636363") +
    geom_vline(xintercept = -fc, linetype = "dotdash", color = "#636363") +
    geom_hline(yintercept = pline, linetype = "dotdash", color = "#636363") +
    #geom_text_repel(data=res[res$sigGenes == 1,], aes(label=gene_symbol), size = 4) +
    labs(x=expression(fold~change~(log[2])), y=expression(p~value~(-log[10])), title=main_title)
}

# For volcano plots
res1 <- modify_result(as.data.frame(res1t_InjectionvsPre))
res2 <- modify_result(as.data.frame(res1t_RingvsPre))
res3 <- modify_result(as.data.frame(res1t_PillvsPre))
res4 <- modify_result(as.data.frame(res1t_InjectionvsPill))
res5 <- modify_result(as.data.frame(res1t_RingvsPill))
res6 <- modify_result(as.data.frame(res1t_RingvsInjection))

res_for_volplot <- c("res1", "res2", "res3", "res4", "res5", "res6")
res_for_volplot_labels <- c("Injection vs. Pre-treatment", 
                            "Ring vs. Pre-treatment", 
                            "Pill vs. Pre-treatment",
                            "Injection vs. Pill",
                            "Ring vs. Pill", 
                            "Ring vs. Injection")

# find the x, y limits for volcano plots, and y-intercepts
min_x <- c()
max_x <- c()
max_y <- c()
yintercept_vector <- c()
genes_up <- c()
genes_down <- c()
for (r in 1:length(res_for_volplot)) {
  resQ <- get(res_for_volplot[r])
  # to find p-value equivalent to 0.05, whichever yields same # of sig genes
  min_x <- append(min_x, min(resQ$log2FoldChange), after = length(min_x))
  max_x <- append(max_x, max(resQ$log2FoldChange), after = length(max_x))
  max_y <- append(max_y, max(resQ$negLogPval), after = length(max_y))
  yintercept <- as.numeric(format(round(min(resQ[resQ$sigGenes == 1 | resQ$sigGenes == -1,]$negLogPval), 2), nsmall = 2))
  yintercept_vector <- append(yintercept_vector, yintercept, after = length(yintercept_vector))
  genes_up <- append(genes_up, sum(resQ$deg[resQ$deg == 1]), after = length(genes_up))
  genes_down <- append(genes_down, abs(sum(resQ$deg[resQ$deg == -1])), after = length(genes_down))
}

# Helps to define axes limits
min(min_x)
max(max_x)
max(max_y)
yintercept_vector

# ACTION: MANUALLY adjust axes limits in volcano_plot function

volplot1 <- volcano_plot(res1, yintercept_vector[1], res_for_volplot_labels[1])
volplot2 <- volcano_plot(res2, yintercept_vector[2], res_for_volplot_labels[2])
volplot3 <- volcano_plot(res3, yintercept_vector[3], res_for_volplot_labels[3])
volplot4 <- volcano_plot(res4, yintercept_vector[4], res_for_volplot_labels[4])
volplot5 <- volcano_plot(res5, yintercept_vector[5], res_for_volplot_labels[5])
volplot6 <- volcano_plot(res6, yintercept_vector[6], res_for_volplot_labels[6])

ggdraw(volplot6)

# Table for up and down-regulated genes
# ---------
results_details <- matrix(c(res_for_volplot_labels, genes_down, genes_up), 
                          ncol = 3, byrow = FALSE)
rownames(results_details) <- res_for_volplot_labels
colnames(results_details) <- c("comparison", "down-regulated genes", "up-regulated genes")
results_details <- as.data.frame(results_details)

# add DEGs table to group plot
t1 <- tableGrob(results_details, theme=ttheme_minimal(base_size = 10), rows=NULL) 
t1 <- gtable::gtable_add_grob(t1, grobs = rectGrob(gp=gpar(fill=NA, lwd=2)), t = 2, b = nrow(t1), l = 2, r = ncol(t1))

r1 <- rectGrob(gp=gpar(fill="white", col="white"))

grid.arrange(volplot1, volplot2, volplot3, volplot4, volplot5, volplot6, 
             arrangeGrob(r1, r1, t1, ncol = 3), 
             nrow=3, top = "\nVolcano plots, all comparisons\n") %>%
  ggsave(filename = "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/VolcanoPlots_grouped.pdf", width = 10, height = 10)
  
```

#### Venn, 109 samples, 14226 genes
```{r venn_109samples_14226genes, eval=FALSE}

fc <- 0.58
modify_result <- function(res){
  res$gene <- row.names(res)
  res$negLogPval <- -log10(res$pvalue)
  # threshold to define significant genes
  res <- res[!(is.na(res$log2FoldChange) | is.na(res$pvalue) | is.na(res$padj)),]
  res$sigGenes <- ifelse((res$padj < 0.05 & abs(res$log2FoldChange) > fc), 1, 0)
  res$deg <- ifelse((res$padj < 0.05 & res$log2FoldChange > fc), 1, ifelse((res$padj < 0.05 & res$log2FoldChange < -fc), -1, 0))
  res <- res[res$lfcSE < 1,]
  #res$gene_symbol <- sapply(res$gene, function(i) as.character(gene_name_table[i,]))
  return(res)
}

# -------
# Modify results to get DEGs
# -------
res1 <- modify_result(as.data.frame(res1t_InjectionvsPre))
res2 <- modify_result(as.data.frame(res1t_RingvsPre))
res3 <- modify_result(as.data.frame(res1t_PillvsPre))

res4 <- modify_result(as.data.frame(res1t_InjectionvsPill))
res5 <- modify_result(as.data.frame(res1t_RingvsPill))

res6 <- modify_result(as.data.frame(res1t_InjectionvsRing))
res7 <- modify_result(as.data.frame(res1t_PillvsRing))

res8 <- modify_result(as.data.frame(res1t_RingvsInjection))
res9 <- modify_result(as.data.frame(res1t_PillvsInjection))

res_for_venn <- c("res1", "res2", "res3", "res4", "res5", "res6", "res7", "res8", "res9")
res_for_venn_labels <- c("Injection vs. Pre", "Ring vs. Pre", "Pill vs. Pre",
                         "Injection vs. Pill", "Ring vs. Pill", 
                         "Injection vs. Ring", "Pill vs. Ring", 
                         "Ring vs. Injection", "Pill vs. Injection")

# Function to get DEG stats for venn-diagram
deg_cnt <- c()
deg_up_cnt <- c()
deg_down_cnt <- c()
res_for_venn_labels
deg_names <- list()
deg_up_names <- list()
deg_down_names <- list()
#deg_names <- sapply(res_for_venn_labels,function(x) NULL)
#deg_up_names <- sapply(res_for_venn_labels,function(x) NULL)
#deg_down_names <- sapply(res_for_venn_labels,function(x) NULL)

for (r in 1:length(res_for_venn)) {
  resQ <- get(res_for_venn[r])
  
  #deg_cnt <- append(deg_cnt, sum(resQ$sigGenes[resQ$sigGenes == 1]), after = length(deg_cnt))
  #deg_up_cnt <- append(deg_up_cnt, sum(resQ$deg[resQ$deg == 1]), after = length(deg_up_cnt))
  #deg_down_cnt <- append(deg_down_cnt, abs(sum(resQ$deg[resQ$deg == -1])), after = length(deg_down_cnt))
  
  deg_cnt[r] <- sum(resQ$sigGenes[resQ$sigGenes == 1])
  deg_up_cnt[r] <- sum(resQ$deg[resQ$deg == 1])
  deg_down_cnt[r] <- abs(sum(resQ$deg[resQ$deg == -1]))
  
  deg_names[[r]] <- resQ$gene[resQ$sigGenes == 1]
  deg_up_names[[r]] <- resQ$gene[resQ$deg == 1]
  deg_down_names[[r]] <- resQ$gene[resQ$deg == -1]
}

deg_names
deg_cnt

library(venn)
# colors for Inj, Ring and Pill
color_vector <- c("#b2182b", "#2166ac", "#008000", "#decbe4", "#fdc086", "#8dd3c7")

# Pre-post, all degs
# Left to right, clockwise, Inj, ring, pill
#list("Pill\n281 genes" = deg3_names, "Injection_NETEN\n79 genes" = deg1_names, "Ring\n392 genes" = deg2_names)
#labels <- paste(res_for_venn_labels[1:3], deg_cnt[1:3], sep = "\n")
labels <- as.character(deg_cnt[1:3])
#png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_post-vs-pre.png", width = 1400, height = 1200)
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_post-vs-pre_no-setnames.png", width = 1400, height = 1200)
venn(x = deg_names[1:3], ilabels = TRUE, zcolor = color_vector[1:3], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Post vs. pre, upregulated 
#labels <- paste(res_for_venn_labels[1:3], deg_up_cnt[1:3], sep = "\n")
labels <- as.character(deg_up_cnt[1:3])
#png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_post-vs-pre_upregulated.png", width = 1400, height = 1200)
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_post-vs-pre_upregulated_no-setnames.png", width = 1400, height = 1200)
venn(x = deg_up_names[1:3], ilabels = TRUE, zcolor = color_vector[1:3], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Post vs. pre, downregulated 
#labels <- paste(res_for_venn_labels[1:3], deg_down_cnt[1:3], sep = "\n")
labels <- as.character(deg_down_cnt[1:3])
#png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_post-vs-pre_downregulated.png", width = 1400, height = 1200)
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_post-vs-pre_downregulated_no-setnames.png", width = 1400, height = 1200)
venn(x = deg_down_names[1:3], ilabels = TRUE, zcolor = color_vector[1:3], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# ------------
# cross-sectional comparisons
# ------------
res_for_venn_labels

# All 3 comparisons together, all degs
#labels <- paste(res_for_venn_labels[4:6], deg_cnt[4:6], sep = "\n")
labels <- as.character(deg_cnt[4:6])
#png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_cross-sec_all3.png", width = 1400, height = 1200)
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_cross-sec_all3_no-setnames.png", width = 1400, height = 1200)
venn(x = deg_names[4:6], ilabels = TRUE, zcolor = color_vector[4:6], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# 2-way venns
# Injection and Ring, all
labels <- paste(res_for_venn_labels[4:5], deg_cnt[4:5], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjRing.png", width = 1400, height = 1200)
venn(x = deg_names[4:5], ilabels = TRUE, zcolor = color_vector[4:5], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Injection and Ring, up
labels <- paste(res_for_venn_labels[4:5], deg_up_cnt[4:5], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjRing_up.png", width = 1400, height = 1200)
venn(x = deg_up_names[4:5], ilabels = TRUE, zcolor = color_vector[4:5], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Injection and Ring, down
labels <- paste(res_for_venn_labels[4:5], deg_down_cnt[4:5], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjRing_down.png", width = 1400, height = 1200)
venn(x = deg_down_names[4:5], ilabels = TRUE, zcolor = color_vector[4:5], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Injection and Pill, all
labels <- paste(res_for_venn_labels[6:7], deg_cnt[6:7], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjPill.png", width = 1400, height = 1200)
venn(x = deg_names[6:7], ilabels = TRUE, zcolor = color_vector[c(6,5)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Injection and Pill, up
labels <- paste(res_for_venn_labels[6:7], deg_up_cnt[6:7], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjPill_up.png", width = 1400, height = 1200)
venn(x = deg_up_names[6:7], ilabels = TRUE, zcolor = color_vector[c(6,5)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Injection and Pill, down
labels <- paste(res_for_venn_labels[6:7], deg_down_cnt[6:7], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjPill_down.png", width = 1400, height = 1200)
venn(x = deg_down_names[6:7], ilabels = TRUE, zcolor = color_vector[c(6,5)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Ring and Pill, all
labels <- paste(res_for_venn_labels[8:9], deg_cnt[8:9], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_RingPill.png", width = 1400, height = 1200)
venn(x = deg_names[8:9], ilabels = TRUE, zcolor = color_vector[c(6,4)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Ring and Pill, up
labels <- paste(res_for_venn_labels[8:9], deg_up_cnt[8:9], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_RingPill_up.png", width = 1400, height = 1200)
venn(x = deg_up_names[8:9], ilabels = TRUE, zcolor = color_vector[c(6,4)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Ring and Pill, down
labels <- paste(res_for_venn_labels[8:9], deg_down_cnt[8:9], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_RingPill_down.png", width = 1400, height = 1200)
venn(x = deg_down_names[8:9], ilabels = TRUE, zcolor = color_vector[c(6,4)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

```

#### Heatmaps for DEGs

* Heatmaps using in-house code
```{r heatmaps_109samples_14226genes, eval=FALSE}
# Heatmaps
# -----

# Get list of DEGs
fc <- 0.58
get_sigGenes <- function(res){
  # threshold to define significant genes
  res <- res[!(is.na(res$log2FoldChange) | is.na(res$pvalue) | is.na(res$padj)),]
  res <- res[(res$padj < 0.05) & (abs(res$log2FoldChange) > fc),]
  res <- res[res$lfcSE < 1,]
  # add gene name as a col
  res$gene <- row.names(res)
  return(res$gene)
}


# Function: Subset and edit rlog
subsetnedit_rlog <- function(rlog_input, gene_list, sample_list){
  # Subset for genes and samples
  rlog_matrix <- rlog_input[gene_list, sample_list]
  # Edit sample names --> colnames
  for (i in 1:length(colnames(rlog_matrix))){
    sample_name <- colnames(rlog_matrix)[i]
    colnames(rlog_matrix)[i] <-  as.character(sampleTable2_names[sample_name,]$composite_feature)
  }
  # Modify gene names into symbols --> rownames
  rlog_matrix$gene_name <- rownames(rlog_matrix)
  rlog_matrix$gene_symbol <- sapply(rlog_matrix$gene_name, function(i) as.character(gene_name_table[i,]))
  # Append repeating gene symbols with counter
  rlog_matrix$gene_symbol <- make.names(rlog_matrix$gene_symbol, unique = T)
  rownames(rlog_matrix) <- rlog_matrix$gene_symbol
  rlog_matrix[,c("gene_name", "gene_symbol")] <- NULL
  # Normalize the rlog expressions
  # By mean of all samples
  rlog_matrix <- sweep(rlog_matrix, 1, rowMeans(rlog_matrix), "-")
  # By mean of baseline samples
  #rlog_matrix <- sweep(rlog_matrix, 1, rowMeans(rlog_subset[,colnames(rlog_matrix[grepl("Pretreatment", colnames(rlog_matrix))])]), "-")
  # Return rlog subset with edited row/col names
  # Return where to draw the vlines or hlines
  return(rlog_matrix)
}

# Generate heatmap
source("/yerkes-cifs/runs/MSigDB/Heatmap_finalv2.R")

# -----
# Input data
# -----
# Get the rlog expressions for all samples/genes
rlog_190s_14226g <- rld1t
rlog_whole <- as.data.frame(assay(rld1t))

# gene name table: Ensembl gene id as rownames, gene symbol 1st col
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)

# Metadata
atable <- superTable_metadata_filtered
# Generate new sample names
sampleTable2_names <- atable[,c("composite_feature", "Treatment", "BV", "CST")]
#sampleTable2_names$new_name <- paste(sampleTable2_names$SampleIndex, sampleTable2_names$Treatment, sep = "_")

# Sample groups
gp_baseline <- atable$Sample_ID[atable$Treatment == "Pretreatment"]
gp_Injection <- atable$Sample_ID[atable$Treatment == "Injection_NETEN"]
gp_Pill <- atable$Sample_ID[atable$Treatment == "Pill"]
gp_Ring <- atable$Sample_ID[atable$Treatment == "Ring"]

# Initialize variables
par_dir <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/"

# Generate heatmaps
# ---------
# Injection_NETEN vs. Pill
# ---------
sigGenes_InjvsPill <- get_sigGenes(as.data.frame(res1t_InjectionvsPill))
samples_of_int <- c(as.character(gp_Pill), as.character(gp_Injection))
rlog_subset <- subsetnedit_rlog(rlog_whole, sigGenes_InjvsPill, samples_of_int)

# Generate heatmap
save_file_as <- paste0(par_dir, "uChoose_Analysis_Oct2020/Heatmap_InjvsPill_sigGenes.pdf")
main_title <- "Heatmap for DEGs \n(Injection_NETEN vs. Pill)"
plot_InjvsPill <- heatmap_clust_vline_order(rlog_subset, dendro = "y", title = main_title, 
                                            limits = c(-1,1), 
                                            vlines = c(length(gp_Pill) + 0.5), vline_size=1.5)
ggdraw(plot_InjvsPill)
ggsave(plot=plot_InjvsPill, filename = save_file_as, title = main_title, height = 4, width = 12, limitsize = FALSE)

# ---------
# Ring vs. Pill
# ---------
sigGenes_RingvsPill <- get_sigGenes(as.data.frame(res1t_RingvsPill))
samples_of_int <- c(as.character(gp_Pill), as.character(gp_Ring))
rlog_subset <- subsetnedit_rlog(rlog_whole, sigGenes_RingvsPill, samples_of_int)

# Generate heatmap
save_file_as <- paste0(par_dir, "uChoose_Analysis_Oct2020/Heatmap_RingvsPill_sigGenes.pdf")
main_title <- "Heatmap for DEGs \n(Ring vs. Pill)"
plot_RingvsPill <- heatmap_clust_vline_order(rlog_subset, dendro = "y", title = main_title, 
                                            limits = c(-1,1), 
                                            vlines = c(length(gp_Pill) + 0.5), vline_size=1.5)
ggdraw(plot_RingvsPill)
ggsave(plot=plot_RingvsPill, filename = save_file_as, title = main_title, height = 20, width = 10, limitsize = FALSE)

# ---------
# Ring vs. Injection_NETEN 
# ---------
sigGenes_RingvsInj <- get_sigGenes(as.data.frame(res1t_RingvsInjection))
samples_of_int <- c(as.character(gp_Injection), as.character(gp_Ring))
rlog_subset <- subsetnedit_rlog(rlog_whole, sigGenes_RingvsInj, samples_of_int)

# Generate heatmap
save_file_as <- paste0(par_dir, "uChoose_Analysis_Oct2020/Heatmap_RingvsInj_sigGenes.pdf")
main_title <- "Heatmap for DEGs \n(Ring vs. Injection_NETEN)"
plot_RingvsInj <- heatmap_clust_vline_order(rlog_subset, dendro = "y", title = main_title, 
                                            limits = c(-1,1), 
                                            vlines = c(length(gp_Injection) + 0.5), vline_size=1.5)
ggdraw(plot_RingvsInj)
ggsave(plot=plot_RingvsInj, filename = save_file_as, title = main_title, height = 25, width = 12, limitsize = FALSE)

# ---------
# All three treatments vs. Pre
# ---------
sigGenes_InjvsPre <- get_sigGenes(as.data.frame(res1t_InjectionvsPre))
sigGenes_RingvsPre <- get_sigGenes(as.data.frame(res1t_RingvsPre))
sigGenes_PillvsPre <- get_sigGenes(as.data.frame(res1t_PillvsPre))
sigGenes_allvsPre <- union(sigGenes_InjvsPre, union(sigGenes_RingvsPre, sigGenes_PillvsPre))
samples_of_int <- c(as.character(gp_baseline), as.character(gp_Injection), as.character(gp_Ring), as.character(gp_Pill))
rlog_subset <- subsetnedit_rlog(rlog_whole, sigGenes_allvsPre, samples_of_int)

# Generate heatmap
save_file_as <- paste0(par_dir, "uChoose_Analysis_Oct2020/Heatmap_unionDEGs_longitudinal.pdf")
main_title <- "Heatmap for DEGs \n(Union of DEGs from all post vs. pre comparisons)"
a <- length(gp_baseline) + 0.5
b <- a + length(gp_Injection)
c <- b + length(gp_Ring)
plot_allvsPre <- heatmap_clust_vline_order(rlog_subset, dendro = "y", title = main_title, 
                                            limits = c(-1,1), 
                                            vlines = c(a, b, c), vline_size=1.5)
ggdraw(plot_allvsPre)
ggsave(plot=plot_allvsPre, filename = save_file_as, title = main_title, height = 100, width = 30, limitsize = FALSE)

# clustered by samples
save_file_as <- paste0(par_dir, "uChoose_Analysis_Oct2020/Heatmap_unionDEGs_longitudinal_clus.pdf")
plot_allvsPre <- heatmap_clust_vline_order(rlog_subset, dendro = "both", title = main_title, 
                                           limits = c(-1,1))
ggsave(plot=plot_allvsPre, filename = save_file_as, title = main_title, height = 100, width = 30, limitsize = FALSE)


```

* Complex heatmap
```{r complex_heatmap_109samples_14226genes, eval=FALSE}
# ----------------

# Heatmaps using complex heatmap package
# -----
# Functions
# -----
# Get list of DEGs
fc <- 0.58
get_sigGenes <- function(res){
  # threshold to define significant genes
  res <- res[!(is.na(res$log2FoldChange) | is.na(res$pvalue) | is.na(res$padj)),]
  res <- res[(res$padj < 0.05) & (abs(res$log2FoldChange) > fc),]
  res <- res[res$lfcSE < 1,]
  # add gene name as a col
  res$gene <- row.names(res)
  return(res$gene)
}


# Function: Subset and edit rlog
subsetnedit_rlog <- function(rlog_input, gene_list, sample_list){
  
  # Subset for genes and samples
  rlog_matrix <- rlog_input[gene_list, sample_list]
  
  # Edit sample names --> colnames
  #for (i in 1:length(colnames(rlog_matrix))){
  #  sample_name <- colnames(rlog_matrix)[i]
  #  colnames(rlog_matrix)[i] <-  as.character(sampleTable2_names[sample_name,]$composite_feature)
  #}
  
  # Modify gene names into symbols --> rownames
  rlog_matrix$gene_name <- rownames(rlog_matrix)
  rlog_matrix$gene_symbol <- sapply(rlog_matrix$gene_name, function(i) as.character(gene_name_table[i,]))
  # Append repeating gene symbols with counter
  rlog_matrix$gene_symbol <- make.names(rlog_matrix$gene_symbol, unique = T)
  rownames(rlog_matrix) <- rlog_matrix$gene_symbol
  rlog_matrix[,c("gene_name", "gene_symbol")] <- NULL
  
  # Normalize the rlog expressions
  # By mean of all samples
  rlog_matrix <- sweep(rlog_matrix, 1, rowMeans(rlog_matrix), "-")
  # By mean of baseline samples
  #rlog_matrix <- sweep(rlog_matrix, 1, rowMeans(rlog_subset[,colnames(rlog_matrix[grepl("Pretreatment", colnames(rlog_matrix))])]), "-")
  
  return(rlog_matrix)
}

# Generate heatmap
#source("/yerkes-cifs/runs/MSigDB/Heatmap_finalv2.R")

# -----
# Input data
# -----
# Get the rlog expressions for all samples/genes
rlog_190s_14226g <- rld1t
rlog_whole <- as.data.frame(assay(rld1t))

# gene name table: Ensembl gene id as rownames, gene symbol 1st col
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)

# Metadata
atable <- superTable_metadata_filtered
# Generate new sample names
sampleTable2_names <- atable[,c("composite_feature", "Treatment", "BV", "CST", "SampleIndex")]

# Sample groups
gp_baseline <- atable$Sample_ID[atable$Treatment == "Pretreatment"]
gp_Injection <- atable$Sample_ID[atable$Treatment == "Injection_NETEN"]
gp_Pill <- atable$Sample_ID[atable$Treatment == "Pill"]
gp_Ring <- atable$Sample_ID[atable$Treatment == "Ring"]

# Initialize variables
par_dir <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/"

# define colors for the the scale
library(circlize)
library(ComplexHeatmap)
col_fun = colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))
col_fun(seq(-2, 2))

robust_dist = function(x, y) {
  qx = quantile(x, c(0.1, 0.9))
  qy = quantile(y, c(0.1, 0.9))
  l = x > qx[1] & x < qx[2] & y > qy[1] & y < qy[2]
  x = x[l]
  y = y[l]
  sqrt(sum((x - y)^2))
}

# Generate heatmaps
# ---------
# Injection_NETEN vs. Pill
# ---------
sigGenes_InjvsPill <- get_sigGenes(as.data.frame(res1t_InjectionvsPill))
samples_of_int <- c(as.character(gp_Injection), as.character(gp_Pill), as.character(gp_baseline))
rlog_subset <- subsetnedit_rlog(rlog_whole, sigGenes_InjvsPill, samples_of_int)
sigGenes_InjvsPill
rlog_subset

#write.table(sigGenes_InjvsPill, file = "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/DEGs_IvsP.txt", quote = FALSE, row.names = FALSE, col.names = FALSE)


study_arm <- as.character(sapply(colnames(rlog_subset), 
                                 function(i) as.character(sampleTable2_names[i,"Treatment"])))

column_ha = HeatmapAnnotation( Treatment = as.character(sapply(colnames(rlog_subset), 
                                                  function(i) as.character(sampleTable2_names[i,"Treatment"]))),
                               BV = sapply(colnames(rlog_subset), function(i) as.character(sampleTable2_names[i,"BV"])),
                               CST = sapply(colnames(rlog_subset), function(i) as.character(sampleTable2_names[i,"CST"])),
                               col = list(Treatment = c("Injection_NETEN" = "#b2182b", "Pill" = "#008000", "Pretreatment" = "#636363"), 
                                          CST = c("CST-I" = "#fee391", "CST-III" = "#fec44f", "CST-IV" = "#998ec3") ))

#colnames(rlog_subset) <- sapply(colnames(rlog_subset), function(i) as.character(sampleTable2_names[i,"composite_feature"]))

#Ring: "#2166ac"
ht2 = Heatmap(rlog_subset, name = "Injection vs. Pill", 
              col = col_fun,
              column_title_side = "bottom", column_title_gp = gpar(fontsize = 14), 
              row_title = "DEGs", row_title_gp = gpar(fontsize = 14),
              show_column_names = FALSE,
              row_title_rot = 0, 
              cluster_rows = TRUE, cluster_columns = TRUE, 
              show_column_dend = TRUE, column_dend_side = "top", 
              column_dend_height = unit(2, "cm"), row_dend_width = unit(1, "cm"), 
              row_dend_reorder = TRUE,
              clustering_distance_rows = robust_dist, 
              clustering_distance_columns = robust_dist,
              top_annotation = column_ha,
              column_split = study_arm
              )

draw(ht2)

# ---------
# Injection_NETEN vs. Ring
# ---------
sigGenes_InjvsRing <- get_sigGenes(as.data.frame(res1t_InjectionvsRing))
samples_of_int <- c(as.character(gp_Injection), as.character(gp_Ring), as.character(gp_baseline))
rlog_subset <- subsetnedit_rlog(rlog_whole, sigGenes_InjvsRing, samples_of_int)

write.table(sigGenes_InjvsRing, file = "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/DEGs_IvsR.txt", quote = FALSE, row.names = FALSE, col.names = FALSE)

study_arm <- as.character(sapply(colnames(rlog_subset), 
                                 function(i) as.character(sampleTable2_names[i,"Treatment"])))

column_ha = HeatmapAnnotation( Treatment = as.character(sapply(colnames(rlog_subset), 
                                                               function(i) as.character(sampleTable2_names[i,"Treatment"]))),
                               BV = sapply(colnames(rlog_subset), function(i) as.character(sampleTable2_names[i,"BV"])),
                               CST = sapply(colnames(rlog_subset), function(i) as.character(sampleTable2_names[i,"CST"])),
                               col = list(Treatment = c("Injection_NETEN" = "#b2182b", "Ring" = "#2166ac", "Pretreatment" = "#636363"), 
                                          CST = c("CST-I" = "#fee391", "CST-III" = "#fec44f", "CST-IV" = "#998ec3") ))

#colnames(rlog_subset) <- sapply(colnames(rlog_subset), function(i) as.character(sampleTable2_names[i,"composite_feature"]))

#Ring: "#2166ac"
ht3 = Heatmap(rlog_subset, name = "Injection vs. Ring", 
              col = col_fun,
              column_title_side = "bottom", column_title_gp = gpar(fontsize = 14), 
              row_title = "DEGs", row_title_gp = gpar(fontsize = 14),
              show_column_names = FALSE,
              row_title_rot = 0, 
              cluster_rows = TRUE, cluster_columns = TRUE, 
              show_column_dend = TRUE, column_dend_side = "top", 
              column_dend_height = unit(2, "cm"), row_dend_width = unit(1, "cm"), 
              row_dend_reorder = TRUE,
              clustering_distance_rows = robust_dist, 
              clustering_distance_columns = robust_dist,
              top_annotation = column_ha,
              column_split = study_arm
)

draw(ht3)

# ---------
# Ring vs. Pill
# ---------
sigGenes_RingvsPill <- get_sigGenes(as.data.frame(res1t_RingvsPill))

write.table(sigGenes_RingvsPill, file = "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/DEGs_RvsP.txt", quote = FALSE, row.names = FALSE, col.names = FALSE)

# =======
eht_list = ht1 + ht2
grid.newpage()
pushViewport(viewport(layout=grid.layout(2, 1)))
draw(ht_list, newpage = FALSE, ht_gap = unit(1, "cm"))
popViewport()

```

#### GSEA

Location: **/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA**

##### Generate rlog and phenotype class files, run GSEA on standalone tool
```{r gsea_109samples_14226genes, eval=FALSE}

# /yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/
# mkdir GSEA

# [1] Expression dataset file
# rlog files contain col names as 'sample names' and row names are the 'genes'
# For GSEA, we need to add two new cols at the beginning, 'Name' and 'Description'
#gawk 'BEGIN{FS="\t"}NR == 1{printf("Name\tDescription\t%s\n",$0)}NR > 1{printf("%s\tna",$1); for(fi=2;fi<=NF;fi++) printf("\t%s",$fi); printf("\n")}' BL33_uCHOOSE_DESeq2_des2_rlog.txt > GSEA/BL33_uCHOOSE_DESeq2_des2_rlog_forGSEA.txt

gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)

rlog_matrix <- as.data.frame(assay(rld1t))
gene_details <- data.frame(row.names(rlog_matrix), 
                           sapply(rownames(rlog_matrix), function(i) as.character(gene_name_table[i,])))

colnames(gene_details) <- c("Name", "Description")
rlog_for_gsea <- cbind(gene_details, rlog_matrix)
write.table(rlog_for_gsea, file = paste0(par_dir, "uChoose_Analysis_Oct2020/GSEA/data_109s_rlog_for_GSEA.txt"), sep = "\t", quote = FALSE, row.names = FALSE)


# [2] Phenotype label file
# Function to create Phenotype label file for a specific condition
#Line format:      (number of samples) (space) (number of classes) (space) 1
df_samples <- superTable_metadata_filtered[,c("Sample_ID","Treatment")] 
df_samples$feature <- df_samples$Treatment
length(df_samples$Sample_ID)
line1 <- paste(length(df_samples$Sample_ID), length(unique(df_samples$feature)), 1, sep = " ")
line2 <- paste0("# ", gsub(",", "", toString(unique(df_samples$feature))))
line3 <- gsub(",", "", toString(df_samples$feature))

write.table(line1, file = paste0(par_dir, "uChoose_Analysis_Oct2020/GSEA/des109s_14226g_phenotype.cls"), 
            quote = FALSE, eol = "\n", row.names = FALSE, col.names = FALSE)
write.table(line2, file = paste0(par_dir, "uChoose_Analysis_Oct2020/GSEA/des109s_14226g_phenotype.cls"), 
            quote = FALSE, eol = "\n", row.names = FALSE, col.names = FALSE, append = TRUE)
write.table(line3, file = paste0(par_dir, "uChoose_Analysis_Oct2020/GSEA/des109s_14226g_phenotype.cls"), 
            quote = FALSE, eol = "\n", row.names = FALSE, col.names = FALSE, append = TRUE)


# [3] Run GSEA 
# Downloaded GSEA command line tool GSEAv4.1.0 in /yerkes-cifs/runs/tools/GSEA_4.1.0.zip
# Unzipped the file
# Command: gsea-cli.sh GSEA [parameters] 
# To change memory usage update -Xmx8g (8g stands for 8 GB)
# parameters - To find the parameters for an analysis, open the GSEA application, display the page that runs the analysis, enter the parameters that you want to use, and click the Command button at the bottom of the page 

# GSEA for Injection vs. baseline, Hallmark gene sets
system("/yerkes-cifs/runs/tools/GSEA_4.1.0/gsea-cli.sh GSEA -res /yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA/data_109s_rlog_for_GSEA.txt -cls /yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA/des109s_14226g_phenotype.cls#Injection_NETEN_versus_Pretreatment -gmx ftp.broadinstitute.org://pub/gsea/gene_sets/h.all.v7.2.symbols.gmt -collapse Collapse -mode Max_probe -norm meandiv -nperm 1000 -permute phenotype -rnd_type no_balance -scoring_scheme weighted -rpt_label Injection_vs_baseline_Hallmark -metric Signal2Noise -sort real -order descending -chip ftp.broadinstitute.org://pub/gsea/annotations_versioned/Human_ENSEMBL_Gene_ID_MSigDB.v7.0.chip -create_gcts false -create_svgs false -include_only_symbols true -make_sets true -median false -num 100 -plot_top_x 500 -rnd_seed timestamp -save_rnd_lists false -set_max 1000 -set_min 15 -zip_report true -out /yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA")

#-gmx /yerkes-cifs/runs/MSigDB/h.all.v7.2.symbols.gmt
# -chip ftp.broadinstitute.org://pub/gsea/annotations_versioned/Human_ENSEMBL_Gene_ID_MSigDB.v7.0.chip

#Using system JDK.
#Unrecognized option: --module-path=/yerkes-cifs/runs/tools/GSEA_4.1.0/modules
#Error: Could not create the Java Virtual Machine.
#Error: A fatal exception has occurred. Program will exit.

# GSEA results for the following were saved in '/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA'

# Pathway collections
# Hallmark, Kegg, BioCarta, Custom gene-sets (including Serpins, Integrins, epithelial barrier), Th17

# Comparisons
# Injection vs baseline 
# Ring vs baseline 
# Pill vs baseline 
# InjectionvsPill 
# RingvsPill 
# RingvsInjection

```


##### ITT - Generate rlog and phenotype class files, run GSEA on standalone tool

``` {r gsea_ITT, eval=FALSE}

# /yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/
# mkdir GSEA_ITT

# [1] Expression dataset file
# rlog files contain col names as 'sample names' and row names are the 'genes'
# For GSEA, we need to add two new cols at the beginning, 'Name' and 'Description'

gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)

rlog_matrix <- as.data.frame(assay(rlditt))
gene_details <- data.frame(row.names(rlog_matrix), 
                           sapply(rownames(rlog_matrix), function(i) as.character(gene_name_table[i,])))
colnames(gene_details) <- c("Name", "Description")
rlog_for_gsea <- cbind(gene_details, rlog_matrix)
write.table(rlog_for_gsea, file = paste0(par_dir, "uChoose_Analysis_Oct2020/GSEA_ITT/rlog_for_GSEA_ITT.txt"), sep = "\t", quote = FALSE, row.names = FALSE)


# [2] Phenotype label file
# Function to create Phenotype label file for a specific condition
#Line format:      (number of samples) (space) (number of classes) (space) 1
df_samples <- superTable_itt[,c("Sample_ID","Treatment")] 
df_samples$feature <- df_samples$Treatment
length(df_samples$Sample_ID)
line1 <- paste(length(df_samples$Sample_ID), length(unique(df_samples$feature)), 1, sep = " ")
line2 <- paste0("# ", gsub(",", "", toString(unique(df_samples$feature))))
line3 <- gsub(",", "", toString(df_samples$feature))

write.table(line1, file = paste0(par_dir, "uChoose_Analysis_Oct2020/GSEA_ITT/itt_phenotype.cls"), 
            quote = FALSE, eol = "\n", row.names = FALSE, col.names = FALSE)
write.table(line2, file = paste0(par_dir, "uChoose_Analysis_Oct2020/GSEA_ITT/itt_phenotype.cls"), 
            quote = FALSE, eol = "\n", row.names = FALSE, col.names = FALSE, append = TRUE)
write.table(line3, file = paste0(par_dir, "uChoose_Analysis_Oct2020/GSEA_ITT/itt_phenotype.cls"), 
            quote = FALSE, eol = "\n", row.names = FALSE, col.names = FALSE, append = TRUE)


# [3] Run GSEA 
# GSEA results for the following were saved in '/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT'

# Pathway collections
# Hallmark, Kegg, BioCarta, Custom gene-sets (including Serpins, Integrins, epithelial barrier), Th17

# Comparisons
# Injection vs baseline 
# Ring vs baseline 
# Pill vs baseline 
# InjectionvsPill 
# RingvsPill 
# RingvsInjection

```

##### Combine GSEA stats for all comparisons for each pathway collection
```{r gsea_combine_stats, eval=FALSE}
# -----------------
# For each pathway collection, combine stats from each cross-treatment comparison
# -----------------

# Merge all results for a pathway collection
merge_df <- function(resA, resB) {
  merged_results <- Merge(resA, resB, by="row.names")
  row.names(merged_results) <- merged_results$Row.names
  merged_results[,"Row.names"] <- NULL
  return(merged_results)
}

modify_gsea_report <- function(resA, res_ext) {
  row.names(resA) <- resA$NAME
  resA <- resA[,c("NES", "FDR q-val")]
  #resA <- resA[,c("SIZE", "NES", "NOM p-val", "FDR q-val")]
  #resA[,c("NAME","GS<br> follow link to MSigDB", "GS DETAILS")]<- NULL
  colnames(resA) <- gsub("$", paste0("_", res_ext), colnames(resA))
  return(resA)
}


# Compile GSEA stats into one table for each pathway collection
gsea_dir <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA/"
comparison_order <- c("Injection_vs_baseline", "Pill_vs_baseline", "Ring_vs_baseline", 
                      "Injection_vs_Pill", "Ring_vs_Pill", "Ring_vs_Injection")
pathway_coll <- c("Hallmark", "Kegg", "BioCarta", "CustomGeneSets", "c7Th17")

data_list <- list()
for (i in 1:length(pathway_coll)) {
  pathway <- pathway_coll[i]
  # merge results for each pathway coll
  for (j in 1:length(comparison_order)) {
    comparison <- comparison_order[j]
    dir_name <- paste(comparison, pathway, sep = "_")
    a <- dir(path = gsea_dir, recursive = FALSE, pattern = gsub("$", "*", dir_name))
    f <- list.files(path = paste0(gsea_dir, a), recursive = FALSE, pattern = "gsea_report_for.*\\.xls$", full.names = TRUE)
    
    data <- read.table(f[1], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")
    data <- rbind(data, read.table(f[2], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t"))
    
    # modify this gsea report
    edited_data <- modify_gsea_report(data, dir_name)
    # merge data
    if (j == 1) { merged_data <- edited_data}
    else {  merged_data <- merge_df(merged_data, edited_data)}
  }
  # push merged data in a list
  merged_data <- data.frame("NAME" = row.names(merged_data), merged_data)
  results_file <- paste0(gsea_dir, "compiled_gsea_NES_FDR_", pathway, ".txt")
  write.table(merged_data, results_file, quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")
  data_list[[i]] <- merged_data
}
names(data_list) <- pathway_coll

```

##### GSEA enrichment statistics dot plots
```{r gsea_enrichment_dot_plots_v1, eval=FALSE}
# Greg's code for enrichment dotplot
#gseaTable <- read.table("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_extra/exampleGSEAtable.txt", check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")

#View(gseaTable)
#gseaTable %>% ggplot(aes(x=topPhenotype, y=GENESET, color=NES, size=1-log(`NOM p-val`))) + 
#  geom_point(na.rm = TRUE) + 
#  scale_color_gradientn(colors = c("blue","grey75","grey75","red"), limits = c(-2.8,2.8), breaks = c(-2,-1,0,1,2)) + 
#  theme_bw() + 
#  theme(axis.text.y = element_text(angle = 30, vjust = 0.5, hjust=1, size = 4, face = "bold")) + 
#  scale_radius("1-log(NOM p-val)", 
#               range = c(1,(12-0.3*length(unlist(c(na.omit(GSEA_MET[,name])))))), 
#               breaks =  c(2,4,6,8), limits = c(1,8.2), 
#               labels = c("0.37","0.05","0.007","<0.001"))



# -----------------
# For dot plot
# -----------------
# Compile GSEA stats into one table for each pathway collection
gsea_dir <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA/"

# Input parameters
comparison_order_all <- c("Injection_vs_baseline", "Pill_vs_baseline", "Ring_vs_baseline", "Injection_vs_Pill", "Ring_vs_Pill", "Ring_vs_Injection")
comparison_order_lng <- c("Injection_vs_baseline", "Pill_vs_baseline", "Ring_vs_baseline")
comparison_order_cross <- c("Injection_vs_Pill", "Ring_vs_Pill", "Ring_vs_Injection")

pathway_coll <- c("Hallmark", "Kegg", "BioCarta", "CustomGeneSets", "c7Th17")

# Functions
modify_gsea_report2 <- function(resA, res_ext) {
  resA <- resA[,c("NAME", "NES", "NOM p-val")]
  resA$comparison <- rep(res_ext, length(resA$NES))
  return(resA)
}

# ---------
# FOR ALL 6 comparisons
# ---------
# to get table for generating dotplot
comparison_order <- comparison_order_all
data_list2 <- list()
for (i in 1:length(pathway_coll)) {
  pathway <- pathway_coll[i]
  # merge results for each pathway coll
  for (j in 1:length(comparison_order)) {
    comparison <- comparison_order[j]
    dir_name <- paste(comparison, pathway, sep = "_")
    a <- dir(path = gsea_dir, recursive = FALSE, pattern = gsub("$", "*", dir_name))
    f <- list.files(path = paste0(gsea_dir, a), recursive = FALSE, pattern = "gsea_report_for.*\\.xls$", full.names = TRUE)
    
    data <- read.table(f[1], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")
    data <- rbind(data, read.table(f[2], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t"))
    
    edited_data <- modify_gsea_report2(data, dir_name)
    # merge data
    if (j == 1) { merged_data <- edited_data}
    else {  merged_data <- rbind(merged_data, edited_data)}
  }
  # push merged data in a list
  #merged_data <- data.frame("NAME" = row.names(merged_data), merged_data)
  data_list2[[i]] <- merged_data
}
names(data_list2) <- pathway_coll

names(data_list2)
data_list2
colnames(data_list2[[1]])

a <- data_list2[["Hallmark"]]
enriched_pathways <- unique(as.character(a[a$`NOM p-val` < 0.1,"NAME"]))
a[a$NAME %in% enriched_pathways,]

# edit data_list2
#names(data_list2)
# "Hallmark"       "Kegg"           "BioCarta"       "CustomGeneSets" "c7Th17" 

# ---------
# LONGITUDINAL COMPARISONS
# ---------
# to get table for generating dotplot
comparison_order <- comparison_order_lng
data_list_lng <- list()
for (i in 1:length(pathway_coll)) {
  pathway <- pathway_coll[i]
  # merge results for each pathway coll
  for (j in 1:length(comparison_order)) {
    comparison <- comparison_order[j]
    dir_name <- paste(comparison, pathway, sep = "_")
    a <- dir(path = gsea_dir, recursive = FALSE, pattern = gsub("$", "*", dir_name))
    f <- list.files(path = paste0(gsea_dir, a), recursive = FALSE, pattern = "gsea_report_for.*\\.xls$", full.names = TRUE)
    
    data <- read.table(f[1], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")
    data <- rbind(data, read.table(f[2], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t"))
    
    edited_data <- modify_gsea_report2(data, dir_name)
    # merge data
    if (j == 1) { merged_data <- edited_data}
    else {  merged_data <- rbind(merged_data, edited_data)}
  }
  # push merged data in a list
  #merged_data <- data.frame("NAME" = row.names(merged_data), merged_data)
  data_list_lng[[i]] <- merged_data
}
names(data_list_lng) <- pathway_coll

# ---------
# CROSS-SECTIONAL COMPARISONS
# ---------
# to get table for generating dotplot
comparison_order <- comparison_order_cross
data_list_cross <- list()
for (i in 1:length(pathway_coll)) {
  pathway <- pathway_coll[i]
  # merge results for each pathway coll
  for (j in 1:length(comparison_order)) {
    comparison <- comparison_order[j]
    dir_name <- paste(comparison, pathway, sep = "_")
    a <- dir(path = gsea_dir, recursive = FALSE, pattern = gsub("$", "*", dir_name))
    f <- list.files(path = paste0(gsea_dir, a), recursive = FALSE, pattern = "gsea_report_for.*\\.xls$", full.names = TRUE)
    
    data <- read.table(f[1], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")
    data <- rbind(data, read.table(f[2], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t"))
    
    edited_data <- modify_gsea_report2(data, dir_name)
    # merge data
    if (j == 1) { merged_data <- edited_data}
    else {  merged_data <- rbind(merged_data, edited_data)}
  }
  # push merged data in a list
  #merged_data <- data.frame("NAME" = row.names(merged_data), merged_data)
  data_list_cross[[i]] <- merged_data
}
names(data_list_cross) <- pathway_coll


# -----------
# FUNCTION, to get gsea table for a collection and specific comparisons
# -----------
get_enrichment_dotplot <- function(collection, comparison_order, data_list_input, outputname){
  gsea_stats <- data_list_input[[collection]]
  r <- as.numeric(length(unique(gsea_stats$NAME)))
  h <- r/5
  if (h < 5) {h <- 4}
  gsea_stats$comparison <- gsub(paste0("_", collection), "", gsea_stats$comparison)
  gsea_stats$comparison <- factor(gsea_stats$comparison, 
                                levels = comparison_order)
  gsea_stats$NAME <- gsub(paste0(toupper(collection), "_"), "", gsea_stats$NAME)
  b <- gsea_stats[gsea_stats$comparison == rev(comparison_order)[1],]
  # define pathway order in dotplot
  gsea_stats$NAME <- factor(gsea_stats$NAME, levels = rev(b[order(b$`NOM p-val`),"NAME"]))
  
  gsea_stats %>% ggplot(aes(x=comparison, y=NAME, color=NES, size=1-log(`NOM p-val`))) + 
    geom_point(na.rm = TRUE) + 
    scale_color_gradientn(colors = c("blue","grey75","grey75","red"), limits = c(-2.8,2.8), breaks = c(-2,-1,0,1,2)) + 
    theme_bw() + 
    theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5),
          axis.text.x = element_text(angle = 30, hjust=1), 
          axis.text.y = element_text(vjust = 0.5, hjust=1)) + 
    labs(title=paste0("GSEA - ", collection, " pathways"), x="Contraceptive contrast", y="Pathway") +
    scale_radius("1-log(NOM p-val)", 
                 range = c(1,5), 
                 breaks = c(2,4,6,8), limits = c(1,8.2), 
                 labels = c("0.37","0.05","0.007","<0.001"))
  ggsave(paste(save_dir, collection, outputname,"stats.pdf", sep = "_"), height = h, width = 8)
}

# --------
# Enrichment dotplots 
# --------
save_dir <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA"

pathway_coll
# [1] "Hallmark"       "Kegg"           "BioCarta"       "CustomGeneSets" "c7Th17"    

# Get enrichment dot plots
get_enrichment_dotplot("Hallmark", comparison_order_all, data_list2, "all6")
get_enrichment_dotplot("Hallmark", comparison_order_lng, data_list_lng, "longitudinal")
get_enrichment_dotplot("Hallmark", comparison_order_cross, data_list_cross, "crosssectional")

get_enrichment_dotplot("Kegg", comparison_order_all, data_list2, "all6")
get_enrichment_dotplot("Kegg", comparison_order_lng, data_list_lng, "longitudinal")
get_enrichment_dotplot("Kegg", comparison_order_cross, data_list_cross, "crosssectional")

get_enrichment_dotplot("BioCarta", comparison_order_all, data_list2, "all6")
get_enrichment_dotplot("BioCarta", comparison_order_lng, data_list_lng, "longitudinal")
get_enrichment_dotplot("BioCarta", comparison_order_cross, data_list_cross, "crosssectional")

get_enrichment_dotplot("CustomGeneSets", comparison_order_all, data_list2, "all6")
get_enrichment_dotplot("CustomGeneSets", comparison_order_lng, data_list_lng, "longitudinal")
get_enrichment_dotplot("CustomGeneSets", comparison_order_cross, data_list_cross, "crosssectional")

get_enrichment_dotplot("c7Th17", comparison_order_all, data_list2, "all6")
get_enrichment_dotplot("c7Th17", comparison_order_lng, data_list_lng, "longitudinal")
get_enrichment_dotplot("c7Th17", comparison_order_cross, data_list_cross, "crosssectional")

```

* Using functions to generate enrichment stats dotplots
```{r gsea_enrichment_dot_plots_v2, eval=FALSE}
# Compile GSEA stats into one table for each pathway collection
# Put all those tables in a list

# ---------
# Functions
# ---------
# Function: To extract only required stats from GSEA output files
modify_gsea_report2 <- function(resA, res_ext) {
  resA <- resA[,c("NAME", "NES", "NOM p-val")]
  resA$comparison <- rep(res_ext, length(resA$NES))
  return(resA)
}

# Function: To combine GSEA stats for different comparison under the same pathway set into one table, row wise, and then put all such tables for different pathway sets into a list.
compile_gsea_stats <- function(comparison_order, pathway_coll, gsea_dir) {
  data_list <- list()
  for (i in 1:length(pathway_coll)) {
    # merge results for each pathway coll
    for (j in 1:length(comparison_order)) {
      #i <- 1
      #j <- 1
      #pathway_coll <- pathway_coll2
      #pathway_coll
      #dir_name
      #a
      #f
      #gsea_dir <- gsea_dir_itt2
      #list.files(path = paste0(gsea_dir, a), recursive = FALSE, pattern = "gsea_report_for.*\\.tsv$", full.names = TRUE)
      #paste0(gsea_dir, a)
      #f
      #list.files( "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Injection_vs_baseline_CSG.Gsea.1606966230762", pattern = "gsea_report")
      #f
      dir_name <- paste(comparison_order[j], pathway_coll[i], sep = "_")
      a <- dir(path = gsea_dir, recursive = FALSE, pattern = gsub("$", "*", dir_name))
      f <- list.files(path = paste0(gsea_dir, a), recursive = FALSE, pattern = "gsea_report_for.*\\.(xls|tsv)$", full.names = TRUE)
      
      data <- read.table(f[1], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")
      data <- rbind(data, read.table(f[2], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t"))
      
      edited_data <- modify_gsea_report2(data, dir_name)
      # merge data from different comparisons but same collection
      if (j == 1) { merged_data <- edited_data}
      else {  merged_data <- rbind(merged_data, edited_data)}
    }
    # push merged data in a list
    #merged_data <- data.frame("NAME" = row.names(merged_data), merged_data)
    data_list[[i]] <- merged_data
  }
  names(data_list) <- pathway_coll
  return(data_list)
}

# Function: pull gsea stats table for a pathway set and specific comparisons, and generate a dotplot
get_enrichment_dotplot <- function(collection, comparison_order, data_list_input, outputname, save_dir){
  #collection <- "GO"
  #comparison_order <- comparison_order_lng
  #data_list_input <- data_list_lng_itt2
  #outputname <- "longitudinal"
  #save_dir <- gsea_dir_itt_fig
  
  #gsea_stats
  
  gsea_stats <- data_list_input[[collection]]
  enriched_pathways <- unique(as.character(gsea_stats[gsea_stats$`NOM p-val` < 0.1,"NAME"]))
  gsea_stats <- gsea_stats[gsea_stats$NAME %in% enriched_pathways,]
  
  r <- as.numeric(length(unique(gsea_stats$NAME)))
  h <- r/5
  if (h < 5) {h <- 4}
  gsea_stats$comparison <- gsub(paste0("_", collection), "", gsea_stats$comparison)
  gsea_stats$comparison <- factor(gsea_stats$comparison, 
                                  levels = comparison_order)
  gsea_stats$NAME <- gsub(paste0(toupper(collection), "_"), "", gsea_stats$NAME)
  b <- gsea_stats[gsea_stats$comparison == rev(comparison_order)[1],]
  # define pathway order in dotplot
  gsea_stats$NAME <- factor(gsea_stats$NAME, levels = rev(b[order(b$`NOM p-val`),"NAME"]))
  
  gsea_stats %>% ggplot(aes(x=comparison, y=NAME, color=NES, size=1-log(`NOM p-val`))) + 
    geom_point(na.rm = TRUE) + 
    scale_color_gradientn(colors = c("blue","grey75","grey75","red"), limits = c(-2.8,2.8), breaks = c(-2,-1,0,1,2)) + 
    theme_bw() + 
    theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5),
          axis.text.x = element_text(angle = 30, hjust=1), 
          axis.text.y = element_text(vjust = 0.5, hjust=1)) + 
    labs(title=paste0("GSEA - ", collection, " pathways"), x="Contraceptive contrast", y="Pathway") +
    scale_radius("1-log(NOM p-val)", 
                 range = c(1,5), 
                 breaks = c(2,4,6,8), limits = c(1,8.2), 
                 labels = c("0.37","0.05","0.007","<0.001"))
  ggsave(paste0(save_dir, "GSEA_", collection, "_",outputname,"_stats_enriched.pdf"), height = h, width = 20, limitsize = FALSE)
  
}



#names(data_list2)
# "Hallmark"       "Kegg"           "BioCarta"       "CustomGeneSets" "c7Th17" 

# -------------
# Generate stats tables
# -------------
# to get input table for generating dotplot using ggplot
gsea_dir_pp <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA/"
gsea_dir_itt <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/"
gsea_dir_itt2 <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/"
gsea_dir_itt_fig <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Enrichment_stats_dotplots_ITT/"

pathway_coll1 <- c("Hallmark", "Kegg", "BioCarta", "CustomGeneSets", "c7Th17")
pathway_coll2 <- c("CSG", "EBF", "REACTOME", "PID", "GO")
pathway_coll3 <- c("GO")

comparison_order_all <- c("Injection_vs_baseline", "Pill_vs_baseline", "Ring_vs_baseline", 
                          "Injection_vs_Pill", "Ring_vs_Pill", "Ring_vs_Injection")
comparison_order_lng <- c("Injection_vs_baseline", "Pill_vs_baseline", "Ring_vs_baseline")
comparison_order_cross <- c("Injection_vs_Pill", "Ring_vs_Pill", "Ring_vs_Injection")

data_list_all_pp <- compile_gsea_stats(comparison_order_all, pathway_coll, gsea_dir_pp)
data_list_lng_pp <- compile_gsea_stats(comparison_order_lng, pathway_coll, gsea_dir_pp)
data_list_cross_pp <- compile_gsea_stats(comparison_order_cross, pathway_coll, gsea_dir_pp)

data_list_all_itt <- compile_gsea_stats(comparison_order_all, pathway_coll, gsea_dir_itt)
data_list_lng_itt <- compile_gsea_stats(comparison_order_lng, pathway_coll, gsea_dir_itt)
data_list_cross_itt <- compile_gsea_stats(comparison_order_cross, pathway_coll, gsea_dir_itt)

data_list_lng_itt2 <- compile_gsea_stats(comparison_order_lng, pathway_coll = pathway_coll2, gsea_dir = gsea_dir_itt2)
data_list_cross_itt2 <- compile_gsea_stats(comparison_order_cross, pathway_coll2, gsea_dir_itt2)

names(data_list_lng_itt2)
data_list_lng_itt2[5]
# --------
# Enrichment dotplots 
# --------
# cd GSEA
# mkdir Enrichment_stats_dotplots

for (p in pathway_coll) {
  print(p)
  #get_enrichment_dotplot(p, comparison_order_all, data_list_all_pp, "all6", gsea_dir_pp)
  get_enrichment_dotplot(p, comparison_order_lng, data_list_lng_pp, "longitudinal", gsea_dir_pp)
  get_enrichment_dotplot(p, comparison_order_cross, data_list_cross_pp, "crosssectional", gsea_dir_pp)
  
  #get_enrichment_dotplot(p, comparison_order_all, data_list_all_itt, "all6", gsea_dir_itt)
  get_enrichment_dotplot(p, comparison_order_lng, data_list_lng_itt, "longitudinal", gsea_dir_itt)
  get_enrichment_dotplot(p, comparison_order_cross, data_list_cross_itt, "crosssectional", gsea_dir_itt)
  
}

pathway_coll
# set 2 collections
for (p in pathway_coll2) {
  print(p)
  get_enrichment_dotplot(p, comparison_order_lng, data_list_lng_itt2, "longitudinal", gsea_dir_itt_fig)
  get_enrichment_dotplot(p, comparison_order_cross, data_list_cross_itt2, "crosssectional", gsea_dir_itt_fig)
  
}


# Greg's code for enrichment dotplot
#gseaTable <- read.table("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_extra/exampleGSEAtable.txt", check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")

#View(gseaTable)
#gseaTable %>% ggplot(aes(x=topPhenotype, y=GENESET, color=NES, size=1-log(`NOM p-val`))) + 
#  geom_point(na.rm = TRUE) + 
#  scale_color_gradientn(colors = c("blue","grey75","grey75","red"), limits = c(-2.8,2.8), breaks = c(-2,-1,0,1,2)) + 
#  theme_bw() + 
#  theme(axis.text.y = element_text(angle = 30, vjust = 0.5, hjust=1, size = 4, face = "bold")) + 
#  scale_radius("1-log(NOM p-val)", 
#               range = c(1,(12-0.3*length(unlist(c(na.omit(GSEA_MET[,name])))))), 
#               breaks =  c(2,4,6,8), limits = c(1,8.2), 
#               labels = c("0.37","0.05","0.007","<0.001"))


```

#### Enrichment plots, in-house code
```{r enrichment_plots, eval=FALSE}
# SAMPLE CODES
# -----------------
# Generate single-enrichment plots 
# -----------------
setwd("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis")

# Function: generate gsea plots
#source("/yerkes-cifs/runs/MSigDB/GSEA_enrich_auto.R")
# modified the code for ggtitle
source("/yerkes-cifs/runs/MSigDB/GSEA_enrich_auto_vX.R")

# Generate enrichment plots for a specific dataset 
# -----------------------
# ISG_SET
#One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/GSEA/InjectionvsPill_CustomGeneSets.Gsea.1568321700481", geneset="ISG_SET", color1="red", phen1="Higher in Injection", phen2="Higher in Pill", label1_y_pos = 0.6, filename="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/GSEA/ISG_Set_InjvsPill", mainTitle = "Injection_NETEN vs. Pill\nISG Set")

source("/yerkes-cifs/runs/MSigDB/GSEA_enrich_auto_vX.R")

# -----------------
# Dual enrichment plots
# --------------
#NFKB_TARGET_GENES_SET

#NFKB_TARGET_GENES_SET
# Pill vs Injection and Ring vs Injection 
#One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/GSEA/PillvsInjection_CustomGeneSets.Gsea.1568730913275", dir2="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/GSEA/RingvsInjection_CustomGeneSets.Gsea.1568743836870", geneset="NFKB_TARGET_GENES_SET", color1="blue", color2="red", phen1="Higher in Ring/Pill", phen2="Higher in Injection", label1_y_pos = -0.2, label2_y_pos = 0.2, label1="Pill vs Injection", label2="Ring vs Injection", filename = "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/GSEA/Enrichment_plots/NFKB_TARGET_GENES_SET_PillvsInjandRingvsInj", mainTitle = "NFKB_TARGET_GENES_SET\n Pill and Ring compared to Injection_NETEN")

```

### Inflammation analysis, 106 samples dataset
* Inflammation analysis
```{r DESeq_infl_analysis, eval=FALSE}
# Inlcude inflammation category in the design
# 106 samples
samples_to_remove <- c(low_qual_samples_v3, "57_Batch1_063_2",  "66_Batch2_072_2",  "114_Batch2_177_1")
#a <- superTable_metadata[superTable_metadata$Sample_ID %in% samples_to_remove,c("Sample_ID", "RIN","host", "MEDIAN_CV_COVERAGE", "Infl_category_Kmeans","Treatment")]
#a[order(a$Treatment),]

superTable_metadata_filtered_v4 <- superTable_metadata_filtered[!is.na(superTable_metadata_filtered$Infl_category_Kmeans),]
superTable_metadata_filtered_v4$BV
# DESeq run shown in previous chunk
dds1u <- DESeqDataSetFromHTSeqCount(superTable_metadata_filtered_v4, ".", design =~ Infl_category_Kmeans)
dds1u_copy <- dds1u
dds1u <- dds1u[row.names(dds1u) %in% filtered_rcnt_target_genes,]
dds1u <- DESeq(dds1u)
vsd1u <- varianceStabilizingTransformation(dds1u, blind = FALSE, fitType = "parametric")
rld1u <- rlog(dds1u, blind = FALSE, fitType = "parametric", betaPriorVar = 10)

resultsNames(dds1u)

library(DESeq2)

#dds1v <- DESeqDataSetFromHTSeqCount(superTable_metadata_filtered_v4, ".", design =~ Treatment * Infl_category_Kmeans)
#dds1v_copy <- dds1v
#dds1v <- dds1v[row.names(dds1v) %in% filtered_rcnt_target_genes,]
#dds1v <- DESeq(dds1v)
#vsd1v <- varianceStabilizingTransformation(dds1v, blind = FALSE, fitType = "parametric")
#rld1t <- rlog(dds1t, blind = FALSE, fitType = "parametric", betaPriorVar = 10)

colData(dds1u)
# PCA
vsd1u$Infl_category_Kmeans

# PCA
# generate PCA with treatment:inflammation coloring.
# volcano plots for H vs L
# and within each study-arm
# heatmaps for DEGs within each study-arm
#vsd1v$BV
#pcaData <- plotPCA(vsd1v, intgroup=c("Treatment", "BV","Infl_category_Kmeans"), returnData=TRUE)
#percentVar <- round(100 * attr(pcaData, "percentVar"))

#ggplot(pcaData, aes(PC1, PC2)) + theme_bw() + 
#  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
#  labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance")) +
#  coord_fixed() +
#  geom_point(aes(color = Treatment, shape = Infl_category_Kmeans, size =BV)) +
#  labs(title = "UCHOOSE study", color = "legend") +
#  scale_colour_manual(name = "Treatment", 
#                      labels = c("Pretreatment", "Injection_NETEN","Pill","Ring"),
#                      values = c("#636363","#b2182b","#008000","#2166ac"))  +
#  scale_shape_manual(name = "Inflammation", 
#                      labels = c("High", "Low"),
#                      values = c(15,20)) 


# Get results
resultsNames(dds1v)

# Pretreatment, High vs. Low
# Injection_NETEN, High vs. Low
# Pill, High vs. Low
# Ring, High vs. Low

# PCA for High vs. Low
pcaData <- plotPCA(vsd1u, intgroup=c("Treatment","Infl_category_Kmeans"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData, aes(PC1, PC2)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  geom_point(aes(color = Treatment, shape = Infl_category_Kmeans), size =3) +
  labs(title = "UCHOOSE study", color = "legend") +
  scale_colour_manual(name = "Treatment", 
                      labels = c("Pretreatment", "Injection_NETEN","Pill","Ring"),
                      values = c("#636363","#b2182b","#008000","#2166ac"))  +
  scale_shape_manual(name = "Inflammation", 
                      labels = c("High", "Low"),
                      values = c(15,20)) 
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Sep2020/Inflammation_analysis/PCA_infl-analysis.pdf", height = 6, width = 10)

resultsNames(dds1u)
res1u_HvsL <- results(dds1u, contrast = c("Infl_category_Kmeans", "High", "Low"))
summary(res1u_HvsL, alpha = 0.05)

# Write results to file
res1u_HvsL_gene_names
#round(res1u_HvsL, digits = 3)
res1u_HvsL_gene_names <- res1u_HvsL
res1u_HvsL_gene_names$EnsemblID <- rownames(res1u_HvsL_gene_names)
res1u_HvsL_gene_names$GeneSymbol <- sapply(res1u_HvsL_gene_names$EnsemblID, function(i) as.character(gene_name_table[i,]))
colnames(res1u_HvsL_gene_names)
res1u_HvsL_gene_names <- res1u_HvsL_gene_names[,c(7,8,1,2,3,4,5,6)]
res1u_HvsL_gene_names
write.table(res1u_HvsL_gene_names, "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Sep2020/Inflammation_analysis/uChoose_DESeq-results_inflammation-data.txt", row.names = FALSE, quote = FALSE, col.names = TRUE, sep = "\t")

# Volcano plot High vs. Low
library(ggplot2)
library(cowplot)

modify_result <- function(res){
  res <- res[!(is.na(res$log2FoldChange) | is.na(res$pvalue) | is.na(res$padj)),]
  res$gene <- row.names(res)
  res$negLogPval <- -log10(res$pvalue)
  # threshold to define significant genes
  res$sigGenes <- ifelse((res$padj < 0.05 & abs(res$log2FoldChange) > 1), 1, 0)
  res$deg <- ifelse((res$padj < 0.05 & res$log2FoldChange > 1), 1, ifelse((res$padj < 0.05 & res$log2FoldChange < -1), -1, 0))
  # filter results 
  #res <- res[res$baseMean > 10,]
  res <- res[res$lfcSE < 1,]
  return(res)
}

volcano_plot <- function(res, pline, main_title){
  ggplot(res, aes(x=log2FoldChange, y=negLogPval)) + 
    scale_color_manual(name="sigGenes", breaks = c(1,0), values = c("1"="red", "0"="black"), labels = c("DEGs", "not significant")) +
    scale_x_continuous(limits = c(-3,3), breaks = c(-3,-2,-1,0,1,2,3)) +
    scale_y_continuous(limits = c(-1,12), breaks = c(0,pline,5,10,15)) +
    geom_point(aes(color=as.factor(sigGenes)), size = 1) +
    theme_bw() +
    theme(axis.text = element_text(size = 9), axis.title = element_text(size = 12), plot.title = element_text(hjust = 0.5, size = 12), 
          panel.grid.minor = element_blank(), 
          legend.title = element_blank(), legend.position = c(1,1), legend.justification = c(1,1), 
          legend.background = element_blank(), legend.key = element_blank()) +
    geom_vline(xintercept = 1, linetype = "dotdash", color = "#636363") +
    geom_vline(xintercept = -1, linetype = "dotdash", color = "#636363") +
    geom_hline(yintercept = pline, linetype = "dotdash", color = "#636363") +
    geom_text_repel(data=res[res$sigGenes == 1,], aes(label=gene_symbol), size = 4) +
    labs(x=expression(fold~change~(log[2])), y=expression(p~value~(-log[10])), title=main_title)
}

res <- modify_result(as.data.frame(res1u_HvsL))

gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)
res$gene_symbol <- sapply(res$gene, function(i) as.character(gene_name_table[i,]))

up <- sum(res$deg[res$deg == 1])
down <- abs(sum(res$deg[res$deg == -1]))
up
# to find p-value equivalent to 0.05, whichever yields same # of sig genes
yintercept <- as.numeric(format(round(min(res[res$sigGenes == 1 | res$sigGenes == -1,]$negLogPval), 2), nsmall = 2))
summary(res$log2FoldChange)
summary(res$negLogPval)
#yintercept
#x - [-2.36,2.68], y - [0,12], yintercept - 3.85
# generate volcano plot
volplot1u <- volcano_plot(res, yintercept, "Volcano plot for Inflammation - High vs. Low")
ggdraw(volplot1u)

ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Sep2020/Inflammation_analysis/VolcanoPlot_infl-analysis_labeled.pdf", height = 8, width = 10)

#ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Sep2020/Inflammation_analysis/VolcanoPlot_infl-analysis.pdf", height = 8, width = 10)

colData(dds1u)
plotCounts(dds1u, "ENSG00000167779", intgroup = "Infl_category_Kmeans")
# Heatmaps for DEGs High vs. Low
# Extract rlog expressions 
rld1u_copy <- rld1u
rlog_whole <- assay(rld1u)
rlog_whole <- as.data.frame(rlog_whole)

# Define order of the samples
atable <- superTable_metadata_filtered_v4

gp_baseline <- atable$Sample_ID[atable$Treatment == "Pretreatment"]
gp_Injection <- atable$Sample_ID[atable$Treatment == "Injection_NETEN"]
gp_Pill <- atable$Sample_ID[atable$Treatment == "Pill"]
gp_Ring <- atable$Sample_ID[atable$Treatment == "Ring"]

gp_High <- atable$Sample_ID[atable$Infl_category_Kmeans == "High"]
gp_Low <- atable$Sample_ID[atable$Infl_category_Kmeans == "Low"]
length(gp_High)

#samples_of_int <- c(as.character(gp_baseline), as.character(gp_Injection), as.character(gp_Pill), as.character(gp_Ring))
#samples_of_int <- factor(samples_of_int, levels = samples_of_int)

#samples_of_int <- c(as.character(gp_Low), as.character(gp_High))
samples_of_int <- c(as.character(gp_High), as.character(gp_Low))
samples_of_int <- factor(samples_of_int, levels = samples_of_int)

rlog_whole <- rlog_whole[,levels(samples_of_int)]

# Label rlog rows/cols with more meaningful names
# More meaningful sample names (column names), important to do this before adding any gene names cols to rlog
#colnames(rlog_whole_norm_copy) <- sampleTable_visit2$composite_feature
sampleTable2_names <- atable[,c("Sample_ID", "SubjectID", "SampleIndex","Visit_no", "Treatment", "Infl_category_Kmeans")]
sampleTable2_names$new_name <- paste(sampleTable2_names$SampleIndex, sampleTable2_names$Infl_category_Kmeans, sep = "_")

# Normalize the rlog
rowmeans_whole <- rowMeans(rlog_whole)
rlog_whole_norm <- sweep(rlog_whole, 1, rowmeans_whole, "-")
rlog_whole_norm_copy <- rlog_whole_norm

# Add the Ensembl IDs as a column
rlog_whole_norm$gene_name <- row.names(rlog_whole_norm)
rlog_whole_norm_copy$gene_name <- row.names(rlog_whole_norm_copy)
rlog_whole_norm_copy$gene_name <- factor(rlog_whole_norm_copy$gene_name, levels = rlog_whole_norm_copy$gene_name)

# Func1: Get the list of differentially expressed genes
get_sigGenes <- function(resObject){
  resObject <- resObject[!(is.na(resObject$log2FoldChange) | is.na(resObject$pvalue) | is.na(resObject$padj)),]
  resObject <- resObject[(abs(resObject$log2FoldChange) > 1) & (resObject$padj < 0.05),]
  resObject <- resObject[resObject$lfcSE < 1,]
  resObject <- resObject[order(resObject$padj),]
  sigGenes <- row.names(resObject)
  return(sigGenes)
}

# Ensembl gene IDs to Gene symbols
# gene name table: Ensembl gene id as rownames, gene symbol 1st col
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)

# Func2: Subset rlog expressions for a gene list
modify_rlog <- function(rlog_matrix, gene_list){
  rlog_subset <- rlog_matrix[rlog_matrix$gene_name %in% gene_list,]
  #rlog_subset$gene_name <- factor(rlog_subset$gene_name, levels = gene_list)
  #Order genes by levels
  #rlog_subset <- rlog_subset[as.character(gene_list),]
  #delete non-numerical columns
  rlog_subset[,"gene_name"] <- NULL
  #rlog_subset[,"gene_symbol"] <- NULL
  #rlog_subset <- rlog_subset[, -122]
  return(rlog_subset)
}

source("/yerkes-cifs/runs/MSigDB/Heatmap_finalv2.R")

# High vs. Low
# ---------
# Obtain list of DEGs
res <- as.data.frame(res1u_HvsL)
sigGenes_HvsL <- get_sigGenes(res)
# Subset rlog
rlog_subset <- modify_rlog(rlog_whole_norm_copy, sigGenes_HvsL)

# Modify colnames
for (i in 1:length(colnames(rlog_subset))){
  sample_name <- colnames(rlog_subset)[i]
  colnames(rlog_subset)[i] <- sampleTable2_names[sample_name,]$new_name
}

# Add gene symbols as a col and make that into rownames
rlog_subset$gene_name <- rownames(rlog_subset)
rlog_subset$gene_symbol <- sapply(rlog_subset$gene_name, function(i) as.character(gene_name_table[i,]))
# Append repeating gene symbols with counter
#rlog_whole_norm_copy$gene_symbol <- make.names(rlog_whole_norm_copy$gene_symbol, unique = T)
rownames(rlog_subset) <- rlog_subset$gene_symbol

rlog_subset[,"gene_name"] <- NULL
rlog_subset[,"gene_symbol"] <- NULL
length(gp_Low)
# Generate heatmap with samples from all 3 groups
save_file_as <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Sep2020/Inflammation_analysis/Heatmap_HvsL_sigGenes.pdf"
main_title <- "Heatmap for DEGs \n(Inflammation, High vs. Low)"
plot_HvsL <- heatmap_clust_vline_order(rlog_subset, dendro = "y", title = main_title, limits = c(-1.5,1.5), vlines = c(41.5), vline_size = 1)

ggdraw(plot_HvsL)
ggsave(plot=plot_HvsL, filename = save_file_as, title = main_title, height = 6, width = 16, limitsize = FALSE)

# cluster samples
save_file_as <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Sep2020/Inflammation_analysis/Heatmap_HvsL_sigGenes_clustered-samples2.pdf"
plot_HvsL <- heatmap_clust_vline_order(rlog_subset, dendro = "both", title = main_title, limits = c(-1.5,1.5))


ggdraw(plot_HvsL)
ggsave(plot=plot_HvsL, filename = save_file_as, title = main_title, height = 10, width = 16, limitsize = FALSE)

View(superTable_metadata)
```

* inflammation and treatment as design factors
```{r infl_and_treatment, eval=FALSE}
resultsNames(dds1v)

# Pre-treatment, High vs Low
# Injection, High vs. Low
# Pill, High vs. Low
# Ring, High vs. Low

res1u_HvsL <- results(dds1u, contrast = c("Infl_category_Kmeans", "High", "Low"))
summary(res1u_HvsL, alpha = 0.05)

res1v_Pretret_HvsL <- results(dds1v, contrast = c("Infl_category_Kmeans","High","Low"))
summary(res1v_Pretret_HvsL, alpha = 0.05)
# 10, 0

res1v_Inj_HvsL <- results(dds1v, contrast = list(c("Infl_category_Kmeans_High_vs_Low","TreatmentInjection_NETEN.Infl_category_KmeansHigh")))
summary(res1v_Inj_HvsL, alpha=0.05)

res1v_Pill_HvsL <- results(dds1v, contrast = list(c("Infl_category_Kmeans_High_vs_Low","TreatmentPill.Infl_category_KmeansHigh")))
summary(res1v_Pill_HvsL, alpha=0.05)
# up- 5, down - 34

res1v_Ring_HvsL <- results(dds1v, contrast = list(c("Infl_category_Kmeans_High_vs_Low","TreatmentRing.Infl_category_KmeansHigh")))
summary(res1v_Ring_HvsL, alpha=0.05)
# 4, 2

# Volcano plot High vs. Low
library(ggplot2)
library(cowplot)

modify_result <- function(res){
  res <- res[!(is.na(res$log2FoldChange) | is.na(res$pvalue) | is.na(res$padj)),]
  res$gene <- row.names(res)
  res$negLogPval <- -log10(res$pvalue)
  # threshold to define significant genes
  res$sigGenes <- ifelse((res$padj < 0.05 & abs(res$log2FoldChange) > 1 & res$lfcSE < 1), 1, 0)
  res$deg <- ifelse((res$padj < 0.05 & res$log2FoldChange > 1 & res$lfcSE < 1), 1, 
                    ifelse((res$padj < 0.05 & res$log2FoldChange < -1 & res$lfcSE < 1), -1, 0))
  # filter results 
  #res <- res[res$baseMean > 10,]
  res <- res[res$lfcSE < 1,]
  return(res)
}

volcano_plot <- function(res, pline, main_title){
  ggplot(res, aes(x=log2FoldChange, y=negLogPval)) + 
    scale_color_manual(name="sigGenes", breaks = c(1,0), values = c("1"="red", "0"="black"), labels = c("DEGs", "not significant")) +
    scale_x_continuous(limits = c(-4,4), breaks = c(-4,-3,-2,-1,0,1,2,3,4)) +
    scale_y_continuous(limits = c(-1,7), breaks = c(0,2,4,pline,6)) +
    geom_point(aes(color=as.factor(sigGenes)), size = 1) +
    theme_bw() +
    theme(axis.text = element_text(size = 9), axis.title = element_text(size = 12), plot.title = element_text(hjust = 0.5, size = 12), 
          panel.grid.minor = element_blank(), 
          legend.title = element_blank(), legend.position = c(1,1), legend.justification = c(1,1), 
          legend.background = element_blank(), legend.key = element_blank()) +
    geom_vline(xintercept = 1, linetype = "dotdash", color = "#636363") +
    geom_vline(xintercept = -1, linetype = "dotdash", color = "#636363") +
    geom_hline(yintercept = pline, linetype = "dotdash", color = "#636363") +
    #geom_text_repel(data=res[res$sigGenes == 1,], aes(label=gene_symbol), size = 4) +
    labs(x=expression(fold~change~(log[2])), y=expression(p~value~(-log[10])), title=main_title)
}

res <- modify_result(as.data.frame(res1v_Pretret_HvsL))
res1 <- res
res <- modify_result(as.data.frame(res1v_Inj_HvsL))
res2 <- res
res <- modify_result(as.data.frame(res1v_Pill_HvsL))
res3 <- res
res <- modify_result(as.data.frame(res1v_Ring_HvsL))
res4 <- res

up <- sum(res$deg[res$deg == 1])
down <- abs(sum(res$deg[res$deg == -1]))
up
down
# to find p-value equivalent to 0.05, whichever yields same # of sig genes
yintercept <- as.numeric(format(round(min(res[res$sigGenes == 1 | res$sigGenes == -1,]$negLogPval), 2), nsmall = 2))
yintercept1 <- yintercept
yintercept2 <- yintercept
yintercept3 <- yintercept
yintercept4 <- yintercept


summary(res$log2FoldChange)
summary(res$negLogPval)
yintercept
#x - [-2.75,2.88], y - [0,4.6], yintercept - 4.61
#x - [-3.43,3.85], y - [0,5.4], yintercept - Inf
#x - [-3.92,3.73], y - [0,6.06], yintercept - Inf

# generate volcano plot
volplot1v_Pretret_HvsL <- volcano_plot(res1, yintercept1, "Pre-treatment, Inflammation - High vs. Low")
volplot1v_Inj_HvsL <- volcano_plot(res2, yintercept2, "Injection_NETEN, Inflammation - High vs. Low")
volplot1v_Pill_HvsL <- volcano_plot(res3, yintercept3, "Pill, Inflammation - High vs. Low")
volplot1v_Ring_HvsL <- volcano_plot(res4, yintercept4, "Ring, Inflammation - High vs. Low")
ggdraw(volplot1v_Pretret_HvsL)
sum(res1$deg[res1$deg == 1])

treatment <- c("Pretreatment", "Injection_NETEN", "Pill", "Ring")
genes_up <- c(sum(res1$deg[res1$deg == 1]), sum(res2$deg[res2$deg == 1]), 
              sum(res3$deg[res3$deg == 1]), sum(res4$deg[res4$deg == 1]))
genes_down <- c(abs(sum(res1$deg[res1$deg == -1])), 
                abs(sum(res2$deg[res2$deg == -1])), 
                abs(sum(res3$deg[res3$deg == -1])),
                abs(sum(res4$deg[res4$deg == -1])))

genes_up
results_details <- matrix(c(treatment, genes_down, genes_up), ncol = 3, byrow = FALSE)
results_details
rownames(results_details) <- treatment
colnames(results_details) <- c("comparison", "down-regulated genes", "up-regulated genes")
results_details <- as.data.frame(results_details)

# add DEGs table to group plot
t1 <- tableGrob(results_details, theme=ttheme_minimal(base_size = 10), rows=NULL) 
t1 <- gtable::gtable_add_grob(t1, grobs = rectGrob(gp=gpar(fill=NA, lwd=2)), t = 2, b = nrow(t1), l = 2, r = ncol(t1))

pdf("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Sep2020/Inflammation_analysis/VolcanoPlot_treatment-infl-analysis.pdf", width = 10, height = 8)
grid.arrange(volplot1v_Pretret_HvsL, volplot1v_Pill_HvsL, volplot1v_Ring_HvsL, t1, nrow=2, top = "Volcano plots") 
dev.off()

# Heatmaps
```


### Intention-to-treat

#### Run DESeq, ITT
```{run_DESeq_109samples_14226genes_ITT, eval=FALSE}
# Intent-to-treat
superTable_itt <- superTable_metadata_filtered
superTable_itt$Assigned_treatment <- superTable_itt$`1st_HC_assigned`
superTable_itt$Assigned_treatment <- factor(superTable_itt$Assigned_treatment, levels = levels(superTable_itt$Treatment))
superTable_itt[superTable_itt$Visit_no == 1, "Assigned_treatment"] <- "Pretreatment"
superTable_itt[superTable_itt$Assigned_treatment == "Pretreatment" & superTable_itt$Visit_no == 1,]

write.table(superTable_itt, file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/UChoose_ITT_SampleTable_109samples.txt", sep="\t", quote=FALSE, row.names = FALSE)

View(superTable_itt)

View(superTable_itt)
# DESeq
ddsitt <- DESeqDataSetFromHTSeqCount(superTable_itt, ".", design =~ Assigned_treatment)
ddsitt_copy <- ddsitt
ddsitt <- ddsitt[row.names(ddsitt) %in% filtered_rcnt_target_genes,]
ddsitt <- DESeq(ddsitt)
vsditt <- varianceStabilizingTransformation(ddsitt, blind = FALSE, fitType = "parametric")
rlditt <- rlog(ddsitt, blind = FALSE, fitType = "parametric", betaPriorVar = 10)

write.table(counts(ddsitt), file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/UChoose_ITT_DESeq2_RawCounts.txt", sep="\t", quote=FALSE)
write.table(counts(ddsitt, normalize=TRUE), file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/UChoose_ITT_DESeq2_NormCounts.txt", sep="\t", quote=FALSE)
write.table(assay(rlditt), file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/UChoose_ITT_DESeq2_rlog.txt", sep="\t", quote=FALSE)
write.table(assay(vsditt), file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/UChoose_ITT_DESeq2_vst.txt", sep="\t", quote=FALSE)

# ------- #
# get count matrix for each treatment group.
# or change the names so we can identify which group it belongs.
uchoose_itt_ncnt <- counts(ddsitt, normalize=TRUE)
tmp <- colnames(uchoose_itt_ncnt)

colnames(uchoose_itt_ncnt) <- sapply(tmp, 
       function(i) paste0(superTable_itt[i, "Assigned_treatment"], "_",superTable_itt[i, "SubjectID"]))

write.table(uchoose_itt_ncnt, file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/UChoose_ITT_NormCounts_edited-sample-names.txt", sep="\t", quote=FALSE)


# ------- #
# DESeq results

# Injection vs. Pre
resitt_InjectionvsPre <- results(ddsitt, contrast = c("Assigned_treatment", "Injection_NETEN", "Pretreatment"), independentFiltering = FALSE)
summary(resitt_InjectionvsPre, alpha = 0.05)
# up - 30, down - 7
# actual treatment: up - 65, down - 17


# Ring vs. Pre
resitt_RingvsPre <- results(ddsitt, contrast = c("Assigned_treatment", "Ring", "Pretreatment"), 
                            independentFiltering = FALSE)
summary(resitt_RingvsPre, alpha = 0.05)
# up - 685, down - 70
# actual treatment: up - 334, down - 66

# Pill vs. Pre
resitt_PillvsPre <- results(ddsitt, contrast = c("Assigned_treatment", "Pill", "Pretreatment"), 
                            independentFiltering = FALSE)
summary(resitt_PillvsPre, alpha = 0.05)
# up - 325, down 62
# actual treatment: up - 244, down - 38

# Cross-sectional comparisons
# Injection vs. Pill
resitt_InjectionvsPill <- results(ddsitt, contrast = c("Assigned_treatment", "Injection_NETEN", "Pill"), 
                                  independentFiltering = FALSE)
summary(resitt_InjectionvsPill, alpha = 0.05)

# up - 2, down - 8
# actual treatment: up - 1, down - 13

# Injection vs. Ring
resitt_InjectionvsRing <- results(ddsitt, contrast = c("Assigned_treatment", "Injection_NETEN", "Ring"), 
                                  independentFiltering = FALSE)
summary(resitt_InjectionvsRing, alpha = 0.05)
# up - 72, down - 203
# actual treatment: up - 94, down - 134

# Ring vs. Pill
resitt_RingvsPill <- results(ddsitt, contrast = c("Assigned_treatment", "Ring", "Pill"), 
                             independentFiltering = FALSE)
summary(resitt_RingvsPill, alpha = 0.05)
# up - 82, down - 96
# actual treatment: up - 114, down - 64

# Ring vs. Injection
resitt_RingvsInjection <- results(ddsitt, contrast = c("Assigned_treatment", "Ring", "Injection_NETEN"), 
                                  independentFiltering = FALSE)
summary(resitt_RingvsInjection, alpha = 0.05)
# up - 203, down - 72
# actual treatment: up - 134, down - 94

# Pill vs. Injection
resitt_PillvsInjection <- results(ddsitt, contrast = c("Assigned_treatment", "Pill", "Injection_NETEN"), 
                                  independentFiltering = FALSE)
summary(resitt_PillvsInjection, alpha = 0.05)
# up - 8, down - 2
# actual treatment: up - 13, down - 1

# Pill vs. Ring 
resitt_PillvsRing <- results(ddsitt, contrast = c("Assigned_treatment", "Pill", "Ring"), 
                             independentFiltering = FALSE)
summary(resitt_PillvsRing, alpha = 0.05)
# up - 96, down - 82
# actual treatment: up - 64, down - 114

# merge DESeq results into one table

library(lessR)
library(tibble)

# Merge results by gene names
merge_df <- function(resA, resB) {
  merged_results <- Merge(resA, resB, by="row.names")
  row.names(merged_results) <- merged_results$Row.names
  merged_results[,"Row.names"] <- NULL
  return(merged_results)
}

# vector with all results names 
res_vector <- c("resitt_InjectionvsPre", "resitt_PillvsPre", "resitt_RingvsPre",
                "resitt_RingvsInjection", "resitt_PillvsInjection",
                "resitt_InjectionvsPill", "resitt_RingvsPill",
                "resitt_PillvsRing", "resitt_InjectionvsRing")
# first result
resP <- get(res_vector[1])
res_ext <- unlist(strsplit(res_vector[1], "itt"))[2]
colnames(resP) <- gsub("$", res_ext, colnames(resP))

# merge results 
for (r in 2:length(res_vector)) {
  resQ <- get(res_vector[r])
  res_ext <- unlist(strsplit(res_vector[r], "itt"))[2]
  colnames(resQ) <- gsub("$", res_ext, colnames(resQ))
  merged_resPQ <- merge_df(resP, resQ)
  resP <- merged_resPQ
}
res_merged <- resP

head(res_merged)
res_merged$EnsemblID <- row.names(res_merged)

# Add a column with gene symbols
# gene name table: Ensembl gene id as rownames, gene symbol 1st col
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)
res_merged$GeneSymbol <- sapply(res_merged$EnsemblID, function(i) as.character(gene_name_table[i,]))

write.table(res_merged, file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/ITT_DESeq_results_all.txt", sep="\t", quote=FALSE, row.names = FALSE)

```

#### Power analysis parameters

```{r power_analysis, eval=FALSE}

a <- rowData(ddsitt)
head(log(a$baseMean))
head(log(a$dispersion))

resitt_InjectionvsPill$log2FoldChange
# Find what log base is used by PROPER

```

#### Sample details, ITT

```{r sample_details_itt, eval=FALSE}
length(superTable_itt[superTable_itt$Assigned_treatment == "Pretreatment","Sample_ID"])
length(superTable_itt[superTable_itt$Assigned_treatment == "Injection_NETEN","Sample_ID"])
length(superTable_itt[superTable_itt$Assigned_treatment == "Pill","Sample_ID"])
length(superTable_itt[superTable_itt$Assigned_treatment == "Ring","Sample_ID"])

#BV status
superTable_itt$BV
arm <- "Pretreatment"
arm <- "Injection_NETEN"
arm <- "Pill"
arm <- "Ring"
length(superTable_itt[superTable_itt$Assigned_treatment == arm & superTable_itt$BV == "Neg","Sample_ID"])
length(superTable_itt[superTable_itt$Assigned_treatment == arm & superTable_itt$BV == "Int","Sample_ID"])
length(superTable_itt[superTable_itt$Assigned_treatment == arm & superTable_itt$BV == "Pos","Sample_ID"])

#CST 
superTable_itt$CST
arm <- "Pretreatment"
arm <- "Injection_NETEN"
arm <- "Pill"
arm <- "Ring"
length(superTable_itt[superTable_itt$Assigned_treatment == arm & superTable_itt$CST == "CST-I","CST"])
length(superTable_itt[superTable_itt$Assigned_treatment == arm & superTable_itt$CST == "CST-III","CST"])
length(superTable_itt[superTable_itt$Assigned_treatment == arm & superTable_itt$CST == "CST-IV","Sample_ID"])

#Inf category PAM
metadata_v2 <- read.table("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/Metadata_v3_withSampledetails_n_InfPAM.txt", header = TRUE, row.names = NULL, check.names = FALSE, sep = "\t")
rownames(metadata_v2) <- metadata_v2$Sample_ID

x <- metadata_v2[,c("Sample_ID", "Infl_category_PAM", "1st_HC_assigned", "Visit_no")]
arm <- "Injection_NETEN"
arm <- "Pill"
arm <- "Ring"
arm
sort(x[x$`1st_HC_assigned` == arm & x$Visit_no == 2, "Infl_category_PAM"])

```
#### PCA, ITT

```{r PCA_ITT, eval=FALSE}
# -------
# PCA
# -------
vsditt_copy <- vsditt

atable <- colData(vsditt_copy)

atable$match <- "yes"
atable[atable$Assigned_treatment != atable$Treatment,"match"] <- "no"

metadata_v2 <- read.table("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/Metadata_v3_withSampledetails_n_InfPAM.txt", header = TRUE, row.names = NULL, check.names = FALSE, sep = "\t")
rownames(metadata_v2) <- metadata_v2$Sample_ID
atable$Infl_category_PAM <- sapply(rownames(atable), function(i) as.character(metadata_v2[i, "Infl_category_PAM"]))

colData(vsditt_copy) <- atable

vsditt_copy$BV <- factor(vsditt_copy$BV, levels = c("Neg", "Int", "Pos"))
vsditt_copy$CST <- factor(vsditt_copy$CST, levels = c("CST-I", "CST-III", "CST-IV", "<NA>"))
vsditt_copy$match <- factor(atable$match, levels = c("yes", "no"))
vsditt_copy$Infl_category_PAM <- factor(vsditt_copy$Infl_category_PAM, levels = c("Low", "High", "<NA>"))
vsditt_copy$Infl_category_Kmeans <- factor(vsditt_copy$Infl_category_Kmeans, levels = c("Low", "High", "<NA>"))
vsditt_copy$Treatment
vsditt_copy$Assigned_treatment

# Run PCA
pcaData <- plotPCA(vsditt_copy, intgroup=c("Treatment", "Assigned_treatment", "BV", "CST", "match", "Infl_category_Kmeans", "Infl_category_PAM"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

# PCA without baseline samples
# NOTE how the feature col is filtered as a second field
vsditt_nopre <- vsditt_copy[,vsditt_copy$Assigned_treatment != "Pretreatment"]
vsditt_nopre$Assigned_treatment <- factor(vsditt_nopre$Assigned_treatment, levels = c("Injection_NETEN", "Pill", "Ring"))
vsditt_nopre$Treatment <- factor(vsditt_nopre$Treatment, levels = c("Injection_NETEN", "Pill", "Ring"))

pcaData_nopre <- plotPCA(vsditt_nopre, intgroup=c("Treatment", "Assigned_treatment", "BV", "CST", "match", "Infl_category_Kmeans", "Infl_category_PAM"), returnData=TRUE)
percentVar_nopre <- round(100 * attr(pcaData_nopre, "percentVar"))

# [1a] Showing only assigned treatment, all 4 study-arms
pcaData$name
ggplot(pcaData, aes(PC1, PC2, color = Assigned_treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  geom_point(color = 'gray40', size = 4, show_guide = FALSE) +
  geom_point(aes(color = Assigned_treatment), shape = 19, size = 3) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA") +
  stat_ellipse() +
  scale_color_manual(name = "Assigned Study-arm", 
                      labels = c("Baseline", "Net-En","COC","CCVR"),
                      values = c("dimgrey","coral3","dodgerblue3","seagreen")) 
  #geom_text_repel(data=pcaData, aes(label=name), size = 4) 
#ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgnd-treatment.pdf", height = 6, width = 8)
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgnd-treatment_ellipse.pdf", height = 6, width = 8)

# [1b] Showing only assigned treatment, no baseline samples
ggplot(pcaData_nopre, aes(PC1, PC2, color = Assigned_treatment)) + theme_bw() +
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar_nopre[1], "% variance"), y=paste0("PC2: ", percentVar_nopre[2], "% variance")) +
  coord_fixed() +
  geom_point(color = 'gray40', size = 4, show_guide = FALSE) +
  geom_point(aes(color = Assigned_treatment), shape = 19, size = 3) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA") +
  stat_ellipse() +
  scale_color_manual(name = "Assigned Study-arm", 
                      labels = c("Net-En","COC","CCVR"),
                      values = c("coral3","dodgerblue3","seagreen")) 
#ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgnd-treatment_noPre.pdf", height = 6, width = 8)
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgnd-treatment_noPre_ellipse.pdf", height = 6, width = 8)

# [2] Showing assigned and actual treatments
ggplot(pcaData, aes(PC1, PC2, group = Assigned_treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  geom_point(aes(color = Assigned_treatment), shape = 19, size = 6) +
  geom_point(aes(fill = Treatment), shape = 21, size = 3) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA") +
  scale_colour_manual(name = "Assigned Study-arm", 
                      labels = c("Baseline","Net-En","COC","CCVR"),
                      values = c("dimgrey","coral3","dodgerblue3","seagreen")) +
  scale_fill_manual(name = "Actual Study-arm", 
                    labels = c("Baseline", "Net-En","COC","CCVR"),
                      values = c("dimgrey","coral3","dodgerblue3","seagreen")) 

ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgnd-actual-treatment.pdf", height = 6, width = 8)

# [3a] Showing CST, BV as features
ggplot(pcaData, aes(PC1, PC2, color = Assigned_treatment, group = Assigned_treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  geom_point(aes(color = Assigned_treatment, shape = CST, size = BV)) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA", color = "legend") +
  scale_colour_manual(name = "Study-arm", 
                      labels = c("Baseline", "Net-En","COC","CCVR"),
                      values = c("dimgrey","coral3","dodgerblue3","seagreen")) +
  #stat_ellipse() +
  scale_shape_manual(name = "CST", 
                     labels = c("CST-I", "CST-III", "CST-IV", "NA"),
                     values = c(19, 17, 15, 1)) +
  scale_size_manual(name = "BV", 
                    labels = c("Neg", "Int", "Pos"),
                    values = c(4,6,6.5)) 
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgndtrt-and-other-features.pdf", height = 6, width = 8)

# [3b] Showing CST, BV as features
ggplot(pcaData_nopre, aes(PC1, PC2, color = Assigned_treatment, group = Assigned_treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar_nopre[1], "% variance"), y=paste0("PC2: ", percentVar_nopre[2], "% variance")) +
  coord_fixed() +
  geom_point(aes(color = Assigned_treatment, shape = CST, size = BV)) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA", color = "legend") +
  scale_colour_manual(name = "Study-arm", 
                      labels = c("Net-En","COC","CCVR"),
                      values = c("coral3","dodgerblue3","seagreen")) +
  #stat_ellipse() +
  scale_shape_manual(name = "CST", 
                     labels = c("CST-I", "CST-III", "CST-IV", "NA"),
                     values = c(19, 17, 15, 1)) +
  scale_size_manual(name = "BV", 
                    labels = c("Neg", "Int", "Pos"),
                    values = c(4,6,6.5)) 
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgndtrt-and-other-features_noPre.pdf", height = 6, width = 8)

# [4a] Showing CST with assigned treatment
ggplot(pcaData, aes(PC1, PC2, color = Assigned_treatment, group = Assigned_treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  geom_point(color = 'gray40', size = 4, show_guide = FALSE, aes(shape=CST)) +
  geom_point(aes(color = Assigned_treatment, shape = CST), size = 3) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA", color = "legend") +
  scale_colour_manual(name = "Study-arm", 
                      labels = c("Baseline", "Net-En","COC","CCVR"),
                      values = c("dimgrey","coral3","dodgerblue3","seagreen")) +
  scale_shape_manual(name = "Community-state-type", 
                     labels = c("CST-I", "CST-III", "CST-IV", "NA"),
                     values = c(19, 17, 15, 1)) 
  #geom_text_repel(data=pcaData, aes(label=name), size = 4) 
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgndtrt_n_CST.pdf", height = 6, width = 8)

# [4b] Showing CST with assigned treatment, no pre
ggplot(pcaData_nopre, aes(PC1, PC2, color = Assigned_treatment, group = Assigned_treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar_nopre[1], "% variance"), y=paste0("PC2: ", percentVar_nopre[2], "% variance")) +
  coord_fixed() +
  geom_point(color = 'gray40', size = 4, show_guide = FALSE, aes(shape=CST)) +
  geom_point(aes(color = Assigned_treatment, shape = CST), size = 3) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA", color = "legend") +
  scale_colour_manual(name = "Study-arm", 
                      labels = c("Net-En","COC","CCVR"),
                      values = c("coral3","dodgerblue3","seagreen")) +
  scale_shape_manual(name = "Community-state-type", 
                     labels = c("CST-I", "CST-III", "CST-IV", "NA"),
                     values = c(19, 17, 15, 1)) 
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgndtrt_n_CST_noPre.pdf", height = 6, width = 8)

# [5a] Showing BV with assigned treatment
ggplot(pcaData, aes(PC1, PC2, color = Assigned_treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  geom_point(color = 'gray40', size = 4, show_guide = FALSE, aes(shape=BV)) +
  geom_point(aes(color = Assigned_treatment, shape = BV), size = 3) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA", color = "legend") +
  scale_colour_manual(name = "Study-arm", 
                      labels = c("Baseline","Net-En","COC","CCVR"),
                      values = c("dimgrey","coral3","dodgerblue3","seagreen")) +
  scale_shape_manual(name = "BV", 
                     labels = c("Neg", "Int", "Pos"),
                     values = c(19, 17, 15)) 
  #geom_text_repel(data=pcaData, aes(label=name), size = 4) 
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgndtrt_n_BV.pdf", height = 6, width = 8)

# [5b] Showing BV with assigned treatment, no pre
ggplot(pcaData_nopre, aes(PC1, PC2, color = Assigned_treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar_nopre[1], "% variance"), y=paste0("PC2: ", percentVar_nopre[2], "% variance")) +
  coord_fixed() +
  geom_point(color = 'gray40', size = 4, show_guide = FALSE, aes(shape=BV)) +
  geom_point(aes(color = Assigned_treatment, shape = BV), size = 3) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA", color = "legend") +
  scale_colour_manual(name = "Study-arm", 
                      labels = c("Net-En","COC","CCVR"),
                      values = c("coral3","dodgerblue3","seagreen")) +
  scale_shape_manual(name = "BV", 
                     labels = c("Neg", "Int", "Pos"),
                     values = c(19, 17, 15)) 
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgndtrt_n_BV_noPre.pdf", height = 6, width = 8)

# [6a] Showing Infl.cat.PAM with assigned treatment, no pre
ggplot(pcaData, aes(PC1, PC2, color = Assigned_treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  geom_point(color = 'gray40', size = 4, show_guide = FALSE, aes(shape=Infl_category_PAM)) +
  geom_point(aes(color = Assigned_treatment, shape = Infl_category_PAM), size = 3) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA", color = "legend") +
  scale_colour_manual(name = "Study-arm", 
                      labels = c("Baseline","Net-En","COC","CCVR"),
                      values = c("dimgrey","coral3","dodgerblue3","seagreen")) +
  scale_shape_manual(name = "Infl category PAM", 
                     labels = c("Low", "High", "NA"),
                     values = c(19, 17, 15)) 
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgndtrt_n_Infl-by-PAM.pdf", height = 6, width = 8)

# [6b] Showing Infl.cat.PAM with assigned treatment, no pre
ggplot(pcaData_nopre, aes(PC1, PC2, color = Assigned_treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar_nopre[1], "% variance"), y=paste0("PC2: ", percentVar_nopre[2], "% variance")) +
  coord_fixed() +
  geom_point(color = 'gray40', size = 4, show_guide = FALSE, aes(shape=Infl_category_PAM)) +
  geom_point(aes(color = Assigned_treatment, shape = Infl_category_PAM), size = 3) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA", color = "legend") +
  scale_colour_manual(name = "Study-arm", 
                      labels = c("Net-En","COC","CCVR"),
                      values = c("coral3","dodgerblue3","seagreen")) +
  scale_shape_manual(name = "Infl category PAM", 
                     labels = c("Low", "High", "NA"),
                     values = c(19, 17, 15)) 
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgndtrt_n_Infl-by-PAM_noPre.pdf", height = 6, width = 8)

# [7] Showing Inf.cat.kmeans with assigned treatment
ggplot(pcaData, aes(PC1, PC2, color = Assigned_treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  geom_point(color = 'gray40', size = 4, show_guide = FALSE, aes(shape=Infl_category_Kmeans)) +
  geom_point(aes(color = Assigned_treatment, shape = Infl_category_Kmeans), size = 3) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA", color = "legend") +
  scale_colour_manual(name = "Assigned Study-arm", 
                      labels = c("Baseline", "Net-En","COC","CCVR"),
                      values = c("dimgrey","coral3","dodgerblue3","seagreen")) +
  #stat_ellipse() +
  scale_shape_manual(name = "Infl category by kmeans", 
                     labels = c("High", "Low"),
                     values = c(19, 17)) 

ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgndtrt_n_Inf.pdf", height = 6, width = 8)


# PC3 and PC4 
pcaDataGKT <- function (object, intgroup = "condition", ntop = 500) 
{
  #object <- vsd1t
  #ntop = 500
  #intgroup = c("Treatment")
  rv <- rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
                                                     length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup.df <- as.data.frame(colData(object)[, intgroup, 
                                               drop = FALSE])
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1, paste, collapse = " : "))
  }
  else {
    colData(object)[[intgroup]]
  }
  d <- data.frame(pca$x, group = group, 
                  intgroup.df, name = colnames(object))
  attr(d, "percentVar") <- percentVar
  return(d)
}

pcaPlotGKT <- function(object, intgroup = "condition", xpc = 2, ypc = 3, ntop = 500)
{
  d <- pcaDataGKT(object, intgroup, ntop)
  percentVar <- round(100 * attr(d, "percentVar"))
  ggplot(d,aes_string(x = names(d)[xpc], y = names(d)[ypc]), color = Assigned_treatment) +
    theme_bw() + 
    theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
    labs(x=paste0(names(d)[xpc],": ", percentVar[xpc], "% variance"), y=paste0(names(d)[ypc],": ", percentVar[ypc], "% variance")) +
    coord_fixed() +
    geom_point(color = 'gray40', size = 4, show_guide = FALSE) +
    geom_point(aes(color = Assigned_treatment), shape = 19, size = 3) +
    stat_ellipse(aes(color = Assigned_treatment)) +
    labs(title = "UCHOOSE study", color = "legend") +
    scale_colour_manual(name = "Assigned_treatment", 
                      labels = c("Control", "Net-En","COC","CCVR"),
                      values = c("dimgrey","coral3","dodgerblue3","seagreen")) 

#ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_ITT_assgnd-treatment_ellipse.pdf", height = 6, width = 8)
}
  
pcaPlotGKT(vsditt_copy, intgroup = "Assigned_treatment", xpc = 3, ypc = 5, ntop = 500)
# need to change the factor under ggplot inside the function
pcaPlotGKT(vsditt_copy, intgroup = "Treatment", xpc = 3, ypc = 4, ntop = 500)

```

#### Volcano plots, ITT

```{r Volcano_plots_ITT, eval=FALSE}
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)

fc <- 0.58
modify_result <- function(res){
  res$gene <- row.names(res)
  res$negLogPval <- -log10(res$pvalue)
  # threshold to define significant DEGs
  res <- res[!(is.na(res$log2FoldChange) | is.na(res$pvalue) | is.na(res$padj)),]
  res$sigGenes <- ifelse((res$padj < 0.05 & abs(res$log2FoldChange) > fc), 1, 0)
  res$deg <- ifelse((res$padj < 0.05 & res$log2FoldChange > fc), 1, ifelse((res$padj < 0.05 & res$log2FoldChange < -fc), -1, 0))
  res <- res[res$lfcSE < 1,]
  #res$gene_symbol <- sapply(res$gene, function(i) as.character(gene_name_table[i,]))
  return(res)
}

volcano_plot <- function(res, pline, main_title){
  ggplot(res, aes(x=log2FoldChange, y=negLogPval)) + 
    scale_color_manual(name="sigGenes", breaks = c(1,0), values = c("1"="red", "0"="black"), labels = c("DEGs", "not significant")) +
    scale_x_continuous(limits = c(-7,7), breaks = c(-6,-4,-2,0,2,4,6)) +
    scale_y_continuous(limits = c(-1,17), breaks = c(0,pline,5,10,15,20)) +
    geom_point(aes(color=as.factor(sigGenes)), size = 1) +
    theme_bw() +
    theme(axis.text = element_text(size = 9), axis.title = element_text(size = 12), plot.title = element_text(hjust = 0.5, size = 12), 
          panel.grid.minor = element_blank(), 
          legend.title = element_blank(), legend.position = c(1,1), legend.justification = c(1,1), 
          legend.background = element_blank(), legend.key = element_blank()) +
    geom_vline(xintercept = fc, linetype = "dotdash", color = "#636363") +
    geom_vline(xintercept = -fc, linetype = "dotdash", color = "#636363") +
    geom_hline(yintercept = pline, linetype = "dotdash", color = "#636363") +
    #geom_text_repel(data=res[res$sigGenes == 1,], aes(label=gene_symbol), size = 4) +
    labs(x=expression(fold~change~(log[2])), y=expression(p~value~(-log[10])), title=main_title)
}

# For volcano plots
res1 <- modify_result(as.data.frame(resitt_InjectionvsPre))
res2 <- modify_result(as.data.frame(resitt_RingvsPre))
res3 <- modify_result(as.data.frame(resitt_PillvsPre))
res4 <- modify_result(as.data.frame(resitt_InjectionvsPill))
res5 <- modify_result(as.data.frame(resitt_RingvsPill))
res6 <- modify_result(as.data.frame(resitt_RingvsInjection))

res_for_volplot <- c("res1", "res2", "res3", "res4", "res5", "res6")
res_for_volplot_labels <- c("Injection vs. Pre-treatment", 
                            "Ring vs. Pre-treatment", 
                            "Pill vs. Pre-treatment",
                            "Injection vs. Pill",
                            "Ring vs. Pill", 
                            "Ring vs. Injection")

# find the x, y limits for volcano plots, and y-intercepts
min_x <- c()
max_x <- c()
max_y <- c()
yintercept_vector <- c()
genes_up <- c()
genes_down <- c()
for (r in 1:length(res_for_volplot)) {
  resQ <- get(res_for_volplot[r])
  # to find p-value equivalent to 0.05, whichever yields same # of sig genes
  min_x <- append(min_x, min(resQ$log2FoldChange), after = length(min_x))
  max_x <- append(max_x, max(resQ$log2FoldChange), after = length(max_x))
  max_y <- append(max_y, max(resQ$negLogPval), after = length(max_y))
  yintercept <- as.numeric(format(round(min(resQ[resQ$sigGenes == 1 | resQ$sigGenes == -1,]$negLogPval), 2), nsmall = 2))
  yintercept_vector <- append(yintercept_vector, yintercept, after = length(yintercept_vector))
  genes_up <- append(genes_up, sum(resQ$deg[resQ$deg == 1]), after = length(genes_up))
  genes_down <- append(genes_down, abs(sum(resQ$deg[resQ$deg == -1])), after = length(genes_down))
}

# Helps to define axes limits
min(min_x)
max(max_x)
max(max_y)
yintercept_vector

# ACTION: MANUALLY adjust axes limits in volcano_plot function
volplot1 <- volcano_plot(res1, yintercept_vector[1], res_for_volplot_labels[1])
volplot2 <- volcano_plot(res2, yintercept_vector[2], res_for_volplot_labels[2])
volplot3 <- volcano_plot(res3, yintercept_vector[3], res_for_volplot_labels[3])
volplot4 <- volcano_plot(res4, yintercept_vector[4], res_for_volplot_labels[4])
volplot5 <- volcano_plot(res5, yintercept_vector[5], res_for_volplot_labels[5])
volplot6 <- volcano_plot(res6, yintercept_vector[6], res_for_volplot_labels[6])

ggdraw(volplot6)
# Table for up and down-regulated genes
# ---------
res_for_volplot_labels
genes_up
results_details <- matrix(c(res_for_volplot_labels, genes_down, genes_up), 
                          ncol = 3, byrow = FALSE)
rownames(results_details) <- res_for_volplot_labels
colnames(results_details) <- c("comparison", "down-regulated genes", "up-regulated genes")
results_details <- as.data.frame(results_details)

# add DEGs table to group plot
t1 <- tableGrob(results_details, theme=ttheme_minimal(base_size = 10), rows=NULL) 
t1 <- gtable::gtable_add_grob(t1, grobs = rectGrob(gp=gpar(fill=NA, lwd=2)), t = 2, b = nrow(t1), l = 2, r = ncol(t1))

r1 <- rectGrob(gp=gpar(fill="white", col="white"))
grid.arrange(volplot1, volplot2, volplot3, volplot4, volplot5, volplot6, 
             arrangeGrob(r1, r1, t1, ncol = 3), 
             nrow=3, top = "\nIntention-to-treat\nVolcano plots, all comparisons\n") %>%
  ggsave(filename = "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/VolcanoPlots_grouped_ITT.pdf", width = 10, height = 10)

```


#### Count plots, ITT

* Plots for all DEGs, post vs. pre and cross-contraceptive

```{r plotCounts_ITT, eval=FALSE}
# IL1B, IL1A, Serpin, CD19, MS4A1 (CD20) - B-cell infiltration, CD79B, TGFB1.
gene_name_table2 <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)
gene_name_table2$gene_id <- rownames(gene_name_table2)
colnames(gene_name_table2)
# "gene_name" "gene_id"  

#gene_ensemblid <- "ENSG00000125538"
#gene_symbol <- as.character(gene_name_table2[gene_ensemblid, "gene_name"])

gene_symbol <- "IL1B"
gene_symbol <- "IL6"
gene_symbol <- "IL1A"
gene_symbol <- "CD19"
gene_symbol <- "MS4A1"
gene_symbol <- "CD79B"
gene_symbol <- "TGFB1"
gene_symbol <- "IL19"
gene_symbol <- "IL33"
gene_symbol <- "CXCL10"
gene_symbol <- "NFKBIA"
gene_symbol <- "NFKB1"
gene_symbol <- "TNF"

gene_ensemblid <- as.character(gene_name_table2[gene_name_table2$gene_name == gene_symbol,"gene_id"])
plotCounts(ddsitt, gene=gene_ensemblid, intgroup="Assigned_treatment") 
# plotcounts function plots normalized counts with + 0.5 as transform=TRUE

#gene_symbol <- as.character(gene_name_table2[gene_ensemblid, "gene_name"])
  d <- plotCounts(ddsitt, gene=gene_ensemblid, intgroup="Assigned_treatment", returnData = TRUE) 
  plot <- ggplot(d, aes(x = Assigned_treatment, y = count, color = Assigned_treatment)) + 
    geom_point(position=position_jitter(w = 0.1,h = 0), size = 3) +
    #geom_text_repel(aes(label = rownames(d))) + 
    theme_bw() +
    ggtitle(paste0("ITT\n",gene_symbol, "- ", gene_ensemblid)) +
    coord_trans(y="log2") +
    theme(plot.title = element_text(hjust = 0.5), 
          axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x = element_blank()) + 
    scale_color_manual(name = "Assigned Treatment", 
                        labels = c("Control", "Net-En","COC","CCVR"),
                        values = c("dimgrey","coral3","dodgerblue3","seagreen")) +
    stat_summary(fun.y=mean, geom="point",shape=23,size=4, color="black", fill="#f03b20")
  ggdraw(plot)
  ggsave(plot, filename = paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PlotCounts/EnrichedGenes/", gene_symbol, ".pdf"), height = 4, width = 6, limitsize = FALSE)
  
# ----------
# From above chunk
res1 <- modify_result(as.data.frame(resitt_InjectionvsPre))
res2 <- modify_result(as.data.frame(resitt_RingvsPre))
res3 <- modify_result(as.data.frame(resitt_PillvsPre))
res4 <- modify_result(as.data.frame(resitt_InjectionvsPill))
res5 <- modify_result(as.data.frame(resitt_RingvsPill))
res6 <- modify_result(as.data.frame(resitt_RingvsInjection))

#res <- res1
#res <- res2
#res <- res3
#res <- res4
#res <- res5
res <- res6

data=res$gene[res$sigGenes == 1]
data
for (gene_ensemblid in data) {
  gene_symbol <- as.character(gene_name_table2[gene_ensemblid, "gene_name"])
  d <- plotCounts(ddsitt, gene=gene_ensemblid, intgroup="Assigned_treatment", returnData = TRUE) 
  plot <- ggplot(d, aes(x = Assigned_treatment, y = count, color = Assigned_treatment)) + 
    geom_point(position=position_jitter(w = 0.1,h = 0), size = 3) +
    #geom_text_repel(aes(label = rownames(d))) + 
    theme_bw() +
    ggtitle(paste0("ITT\n",gene_symbol, "- ", gene_ensemblid)) +
    coord_trans(y="log2") +
    theme(plot.title = element_text(hjust = 0.5), 
          axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x = element_blank()) + 
    scale_color_manual(name = "Assigned Treatment", 
                        labels = c("Baseline", "Net-En","COC","CCVR"),
                        values = c("dimgrey","coral3","dodgerblue3","seagreen")) +
    stat_summary(fun.y=mean, geom="point",shape=23,size=4, color="black", fill="#f03b20")
  ggdraw(plot)
#  ggsave(plot, filename = paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PlotCounts/EnrichedGenes/", gene_symbol, ".pdf"), height = 6, width = 6, limitsize = FALSE)
  #ggsave(plot, filename = paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PlotCounts/Injection_vs_baseline/", gene_symbol, ".pdf"), height = 6, width = 6, limitsize = FALSE)
 #ggsave(plot, filename = paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PlotCounts/Pill_vs_baseline/", gene_symbol, ".pdf"), height = 6, width = 6, limitsize = FALSE)
  #ggsave(plot, filename = paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PlotCounts/Ring_vs_baseline/", gene_symbol, ".pdf"), height = 6, width = 6, limitsize = FALSE)
  #ggsave(plot, filename = paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PlotCounts/Injection_vs_Pill/", gene_symbol, ".pdf"), height = 6, width = 6, limitsize = FALSE)
  #ggsave(plot, filename = paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PlotCounts/Ring_vs_Pill/", gene_symbol, ".pdf"), height = 6, width = 6, limitsize = FALSE)
  ggsave(plot, filename = paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PlotCounts/Ring_vs_Injection/", gene_symbol, ".pdf"), height = 6, width = 6, limitsize = FALSE)
}  

# for all degs
# results from volcano plots chunk
# Net-En vs. baseline
  gene_count <- plotCounts(dds4_LN, gene, intgroup = c("TimePoint","SIV_Genotype"), returnData = TRUE)
  ggplot(gene_count, aes(x=TimePoint, y=count, color=SIV_Genotype)) +  
    geom_point(position=position_jitter(width=.1,height=0), size=3) +
    ggtitle(paste("Gene counts for", gene, "(LN)", sep = " ")) +
    ggsave(paste("counts_LN_",gene,".pdf", sep = ""))
}


```

#### Venn, ITT

```{r venn_ITT, eval=FALSE}
# -----------
# Venn
# -----------
fc <- 0.58
modify_result <- function(res){
  res$gene <- row.names(res)
  res$negLogPval <- -log10(res$pvalue)
  # threshold to define significant genes
  res <- res[!(is.na(res$log2FoldChange) | is.na(res$pvalue) | is.na(res$padj)),]
  res$sigGenes <- ifelse((res$padj < 0.05 & abs(res$log2FoldChange) > fc), 1, 0)
  res$deg <- ifelse((res$padj < 0.05 & res$log2FoldChange > fc), 1, ifelse((res$padj < 0.05 & res$log2FoldChange < -fc), -1, 0))
  res <- res[res$lfcSE < 1,]
  #res$gene_symbol <- sapply(res$gene, function(i) as.character(gene_name_table[i,]))
  return(res)
}


# Modify results to get DEGs
res1 <- modify_result(as.data.frame(resitt_InjectionvsPre))
res2 <- modify_result(as.data.frame(resitt_RingvsPre))
res3 <- modify_result(as.data.frame(resitt_PillvsPre))

res4 <- modify_result(as.data.frame(resitt_InjectionvsPill))
res5 <- modify_result(as.data.frame(resitt_RingvsPill))

res6 <- modify_result(as.data.frame(resitt_InjectionvsRing))
res7 <- modify_result(as.data.frame(resitt_PillvsRing))

res8 <- modify_result(as.data.frame(resitt_RingvsInjection))
res9 <- modify_result(as.data.frame(resitt_PillvsInjection))

res_for_venn <- c("res1", "res2", "res3", "res4", "res5", "res6", "res7", "res8", "res9")
res_for_venn_labels <- c("Injection vs. Pre", "Ring vs. Pre", "Pill vs. Pre",
                         "Injection vs. Pill", "Ring vs. Pill", 
                         "Injection vs. Ring", "Pill vs. Ring", 
                         "Ring vs. Injection", "Pill vs. Injection")


# Function to get DEG stats for venn-diagram
deg_cnt <- c()
deg_up_cnt <- c()
deg_down_cnt <- c()
res_for_venn_labels
deg_names <- list()
deg_up_names <- list()
deg_down_names <- list()
#deg_names <- sapply(res_for_venn_labels,function(x) NULL)
#deg_up_names <- sapply(res_for_venn_labels,function(x) NULL)
#deg_down_names <- sapply(res_for_venn_labels,function(x) NULL)

for (r in 1:length(res_for_venn)) {
  resQ <- get(res_for_venn[r])
  
  #deg_cnt <- append(deg_cnt, sum(resQ$sigGenes[resQ$sigGenes == 1]), after = length(deg_cnt))
  #deg_up_cnt <- append(deg_up_cnt, sum(resQ$deg[resQ$deg == 1]), after = length(deg_up_cnt))
  #deg_down_cnt <- append(deg_down_cnt, abs(sum(resQ$deg[resQ$deg == -1])), after = length(deg_down_cnt))
  
  deg_cnt[r] <- sum(resQ$sigGenes[resQ$sigGenes == 1])
  deg_up_cnt[r] <- sum(resQ$deg[resQ$deg == 1])
  deg_down_cnt[r] <- abs(sum(resQ$deg[resQ$deg == -1]))
  
  deg_names[[r]] <- resQ$gene[resQ$sigGenes == 1]
  deg_up_names[[r]] <- resQ$gene[resQ$deg == 1]
  deg_down_names[[r]] <- resQ$gene[resQ$deg == -1]
}

deg_names
deg_cnt

library(venn)
# colors for Inj, Ring and Pill
# red, blue, green
color_vector <- c("coral3", "seagreen", "dodgerblue3", "#decbe4", "#fdc086", "#8dd3c7")
#color_vector <- c("#b2182b", "#2166ac", "#008000", "#decbe4", "#fdc086", "#8dd3c7")

# -------
# longitudinal comparisons
# -------
# Pre-post, all degs
# Left to right, clockwise, Inj, ring, pill
labels <- as.character(deg_cnt[1:3])
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_post-vs-pre_no-setnames_ITT.png", width = 1400, height = 1200)
venn(x = deg_names[1:3], ilabels = TRUE, zcolor = color_vector[1:3], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# post vs. pre, upregulated 
labels <- as.character(deg_up_cnt[1:3])
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_post-vs-pre_upregulated_no-setnames_ITT.png", width = 1400, height = 1200)
venn(x = deg_up_names[1:3], ilabels = TRUE, zcolor = color_vector[1:3], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# post vs. pre, downregulated 
labels <- as.character(deg_down_cnt[1:3])
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_post-vs-pre_downregulated_no-setnames_ITT.png", width = 1400, height = 1200)
venn(x = deg_down_names[1:3], ilabels = TRUE, zcolor = color_vector[1:3], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# ------------
# cross-sectional comparisons
# ------------
res_for_venn_labels[c(4,5,8)]

# All 3 comparisons together, all degs
labels <- as.character(deg_cnt[c(4,5,8)])
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_cross-sec_all3_no-setnames_ITT.png", width = 1400, height = 1200)
venn(x = deg_names[c(4,5,8)], ilabels = TRUE, zcolor = color_vector[4:6], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# ------
# Ring comparisons
# ------
# res 2, 5 8
res_for_venn_labels[c(2,5,8)]
labels
# All 3 comparisons together, all degs
labels <- as.character(deg_cnt[c(2,5,8)])
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_Ring_comparisons.png", width = 1400, height = 1200)
venn(x = deg_names[c(2,5,8)], ilabels = TRUE, zcolor = color_vector[c(2,5,6)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()


# -------
# 2-way venns
# -------
# 2-way venns
# Injection and Ring, all
labels <- paste(res_for_venn_labels[4:5], deg_cnt[4:5], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjRing_ITT.png", width = 1400, height = 1200)
venn(x = deg_names[4:5], ilabels = TRUE, zcolor = color_vector[4:5], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Injection and Ring, up
labels <- paste(res_for_venn_labels[4:5], deg_up_cnt[4:5], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjRing_up_ITT.png", width = 1400, height = 1200)
venn(x = deg_up_names[4:5], ilabels = TRUE, zcolor = color_vector[4:5], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Injection and Ring, down
labels <- paste(res_for_venn_labels[4:5], deg_down_cnt[4:5], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjRing_down_ITT.png", width = 1400, height = 1200)
venn(x = deg_down_names[4:5], ilabels = TRUE, zcolor = color_vector[4:5], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Injection and Pill, all
res_for_venn_labels[6:7]
labels <- paste(res_for_venn_labels[6:7], deg_cnt[6:7], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjPill_ITT.png", width = 1400, height = 1200)
venn(x = deg_names[6:7], ilabels = TRUE, zcolor = color_vector[c(6,5)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Injection and Pill, up
labels <- paste(res_for_venn_labels[6:7], deg_up_cnt[6:7], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjPill_up_ITT.png", width = 1400, height = 1200)
venn(x = deg_up_names[6:7], ilabels = TRUE, zcolor = color_vector[c(6,5)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Injection and Pill, down
labels <- paste(res_for_venn_labels[6:7], deg_down_cnt[6:7], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_InjPill_down_ITT.png", width = 1400, height = 1200)
venn(x = deg_down_names[6:7], ilabels = TRUE, zcolor = color_vector[c(6,5)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Ring and Pill, all
labels <- paste(res_for_venn_labels[8:9], deg_cnt[8:9], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_RingPill_ITT.png", width = 1400, height = 1200)
venn(x = deg_names[8:9], ilabels = TRUE, zcolor = color_vector[c(6,4)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Ring and Pill, up
labels <- paste(res_for_venn_labels[8:9], deg_up_cnt[8:9], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_RingPill_up_ITT.png", width = 1400, height = 1200)
venn(x = deg_up_names[8:9], ilabels = TRUE, zcolor = color_vector[c(6,4)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

# Ring and Pill, down
labels <- paste(res_for_venn_labels[8:9], deg_down_cnt[8:9], sep = "\n")
png("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/Venn_RingPill_down_ITT.png", width = 1400, height = 1200)
venn(x = deg_down_names[8:9], ilabels = TRUE, zcolor = color_vector[c(6,4)], cexil = 12, cexsn = 12, ilcs = 6, sncs = 6, box = FALSE, snames = labels) 
dev.off()

```

#### Heatmaps of DEGs, ITT

```{r heatmaps_degs_ITT, eval=FALSE}
# Heatmaps
# -----

# Get list of DEGs
fc <- 0.58
get_sigGenes <- function(res){
  # threshold to define significant genes
  res <- res[!(is.na(res$log2FoldChange) | is.na(res$pvalue) | is.na(res$padj)),]
  res <- res[(res$padj < 0.05) & (abs(res$log2FoldChange) > fc),]
  res <- res[res$lfcSE < 1,]
  # add gene name as a col
  res$gene <- row.names(res)
  return(res$gene)
}


# Function: Subset and edit rlog
subsetnedit_rlog <- function(rlog_input, gene_list, sample_list){
  # Subset for genes and samples
  rlog_matrix <- rlog_input[gene_list, sample_list]
  # Edit sample names --> colnames
  for (i in 1:length(colnames(rlog_matrix))){
    sample_name <- colnames(rlog_matrix)[i]
    colnames(rlog_matrix)[i] <-  as.character(sampleTable2_names[sample_name,]$composite_feature)
  }
  # Modify gene names into symbols --> rownames
  rlog_matrix$gene_name <- rownames(rlog_matrix)
  rlog_matrix$gene_symbol <- sapply(rlog_matrix$gene_name, function(i) as.character(gene_name_table[i,]))
  # Append repeating gene symbols with counter
  rlog_matrix$gene_symbol <- make.names(rlog_matrix$gene_symbol, unique = T)
  rownames(rlog_matrix) <- rlog_matrix$gene_symbol
  rlog_matrix[,c("gene_name", "gene_symbol")] <- NULL
  # Normalize the rlog expressions
  # By mean of all samples
  rlog_matrix <- sweep(rlog_matrix, 1, rowMeans(rlog_matrix), "-")
  # By mean of baseline samples
  #rlog_matrix <- sweep(rlog_matrix, 1, rowMeans(rlog_subset[,colnames(rlog_matrix[grepl("Pretreatment", colnames(rlog_matrix))])]), "-")
  # Return rlog subset with edited row/col names
  # Return where to draw the vlines or hlines
  return(rlog_matrix)
}

# Generate heatmap
source("/yerkes-cifs/runs/MSigDB/Heatmap_finalv2.R")

# -----
# Input data
# -----
# Get the rlog expressions for all samples/genes
rlog_itt <- rlditt
rlog_whole <- as.data.frame(assay(rlditt))

# gene name table: Ensembl gene id as rownames, gene symbol 1st col
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)

# Metadata
atable <- superTable_itt
atable$match <- "yes"
atable[atable$Assigned_treatment != atable$Treatment,"match"] <- "no"
atable$match <- factor(atable$match, levels = c("yes", "no"))

# Generate new sample names
atable$composite_feature_assigned <- paste(atable$SubjectID, atable$Visit_no, atable$Assigned_treatment, sep = "_")
sampleTable2_names <- atable[,c("composite_feature", "composite_feature_assigned","Assigned_treatment","Treatment", "BV", "CST", "match")]

# Sample groups
gp_baseline <- atable$Sample_ID[atable$Assigned_treatment == "Pretreatment"]
gp_Injection <- atable$Sample_ID[atable$Assigned_treatment == "Injection_NETEN"]
gp_Pill <- atable$Sample_ID[atable$Assigned_treatment == "Pill"]
gp_Pill_stayedonpill <- atable$Sample_ID[atable$Assigned_treatment == "Pill" & atable$match == "yes"]
gp_Pill_convertedfrompill <- atable$Sample_ID[atable$Assigned_treatment == "Pill" & atable$match == "no"]
gp_Ring <- atable$Sample_ID[atable$Assigned_treatment == "Ring"]
gp_Ring_stayedonring <- atable$Sample_ID[atable$Assigned_treatment == "Ring" & atable$match == "yes"]
gp_Ring_convertedfromring <- atable$Sample_ID[atable$Assigned_treatment == "Ring" & atable$match == "no"]

# Initialize variables
par_dir <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/"

# Generate heatmaps
# ---------
# Injection_NETEN vs. Pill
# ---------
sigGenes_InjvsPill <- get_sigGenes(as.data.frame(resitt_InjectionvsPill))
samples_of_int <- c(as.character(gp_Pill_convertedfrompill), as.character(gp_Pill_stayedonpill), as.character(gp_Injection))
rlog_subset <- subsetnedit_rlog(rlog_whole, sigGenes_InjvsPill, samples_of_int)

# Generate heatmap
save_file_as <- paste0(par_dir, "uChoose_Analysis_Oct2020/Heatmap_InjvsPill_sigGenes_ITT_v2.pdf")
main_title <- "Heatmap for DEGs \n(Injection_NETEN vs. Pill)"
# group_labs = c("Converted Pill", "Pill", "Injection")
plot_InjvsPill <- heatmap_clust_vline_order(rlog_subset, dendro = "y", title = main_title, 
                                            limits = c(-1,1), 
                                            vlines = c(length(gp_Pill_convertedfrompill) + 0.5, 
                                                       length(gp_Pill) + 0.5), vline_size=1.5)
ggdraw(plot_InjvsPill)
ggsave(plot=plot_InjvsPill, filename = save_file_as, title = main_title, height = 4, width = 12, limitsize = FALSE)

# ---------
# Ring vs. Pill
# ---------
sigGenes_RingvsPill <- get_sigGenes(as.data.frame(resitt_RingvsPill))
samples_of_int <- c(as.character(gp_Pill_convertedfrompill), as.character(gp_Pill_stayedonpill), 
                    as.character(gp_Ring_convertedfromring), as.character(gp_Ring_stayedonring))
rlog_subset <- subsetnedit_rlog(rlog_whole, sigGenes_RingvsPill, samples_of_int)

# Generate heatmap
save_file_as <- paste0(par_dir, "uChoose_Analysis_Oct2020/Heatmap_RingvsPill_sigGenes_ITT.pdf")
main_title <- "Heatmap for DEGs \n(Ring vs. Pill)"
plot_RingvsPill <- heatmap_clust_vline_order(rlog_subset, dendro = "y", title = main_title, 
                                             limits = c(-1,1), 
                                             vlines = c(length(gp_Pill_convertedfrompill) + 0.5, 
                                                        length(gp_Pill) + 0.5, 
                                                        length(gp_Pill) + length(gp_Ring_convertedfromring) + 0.5),
                                             vline_size=1.5)
ggdraw(plot_RingvsPill)
ggsave(plot=plot_RingvsPill, filename = save_file_as, title = main_title, height = 20, width = 10, limitsize = FALSE)

# ---------
# Ring vs. Injection_NETEN 
# ---------
sigGenes_RingvsInj <- get_sigGenes(as.data.frame(resitt_RingvsInjection))
samples_of_int <- c(as.character(gp_Injection), as.character(gp_Ring_convertedfromring), as.character(gp_Ring_stayedonring))
rlog_subset <- subsetnedit_rlog(rlog_whole, sigGenes_RingvsInj, samples_of_int)

# Generate heatmap
save_file_as <- paste0(par_dir, "uChoose_Analysis_Oct2020/Heatmap_RingvsInj_sigGenes_ITT.pdf")
main_title <- "Heatmap for DEGs \n(Ring vs. Injection_NETEN)"
plot_RingvsInj <- heatmap_clust_vline_order(rlog_subset, dendro = "y", title = main_title, 
                                            limits = c(-1,1), 
                                            vlines = c(length(gp_Injection) + 0.5,
                                                       length(gp_Injection) + length(gp_Ring_convertedfromring) + 0.5), 
                                            vline_size=1.5)
ggdraw(plot_RingvsInj)
ggsave(plot=plot_RingvsInj, filename = save_file_as, title = main_title, height = 30, width = 12, limitsize = FALSE)

# ---------
# All three treatments vs. Pre
# ---------
sigGenes_InjvsPre <- get_sigGenes(as.data.frame(resitt_InjectionvsPre))
sigGenes_RingvsPre <- get_sigGenes(as.data.frame(resitt_RingvsPre))
sigGenes_PillvsPre <- get_sigGenes(as.data.frame(resitt_PillvsPre))
sigGenes_allvsPre <- union(sigGenes_InjvsPre, union(sigGenes_RingvsPre, sigGenes_PillvsPre))
samples_of_int <- c(as.character(gp_baseline), as.character(gp_Injection), 
                    as.character(gp_Ring_convertedfromring), as.character(gp_Ring_stayedonring),
                    as.character(gp_Pill_convertedfrompill), as.character(gp_Pill_stayedonpill))
rlog_subset <- subsetnedit_rlog(rlog_whole, sigGenes_allvsPre, samples_of_int)

# Generate heatmap
save_file_as <- paste0(par_dir, "uChoose_Analysis_Oct2020/Heatmap_unionDEGs_longitudinal_ITT.pdf")
main_title <- "Heatmap for DEGs \n(Union of DEGs from all post vs. pre comparisons)"
a <- length(gp_baseline) + 0.5
b <- a + length(gp_Injection)
c <- b + length(gp_Ring_convertedfromring)
d <- c + length(gp_Ring_stayedonring)
e <- d + length(gp_Pill_convertedfrompill)
plot_allvsPre <- heatmap_clust_vline_order(rlog_subset, dendro = "y", title = main_title, 
                                           limits = c(-1,1), 
                                           vlines = c(a, b, c, d, e), vline_size=1.5)
ggdraw(plot_allvsPre)
ggsave(plot=plot_allvsPre, filename = save_file_as, title = main_title, height = 100, width = 30, limitsize = FALSE)





```

* Using complex heatmap package

```{r heatmaps_degs_complexheatmaps_ITT, eval=FALSE}

# Heatmaps using complex heatmap package
# -----
# Functions
# -----
# Get list of DEGs
fc <- 0.58
get_sigGenes <- function(res){
  # threshold to define significant genes
  res <- res[!(is.na(res$log2FoldChange) | is.na(res$pvalue) | is.na(res$padj)),]
  res <- res[(res$padj < 0.05) & (abs(res$log2FoldChange) > fc),]
  res <- res[res$lfcSE < 1,]
  # add gene name as a col
  res$gene <- row.names(res)
  return(res$gene)
}

# Function: Subset and edit rlog
subsetnedit_rlog2 <- function(rlog_input, gene_list, sample_list){
  
  # Subset for genes and samples
  rlog_matrix <- rlog_input[gene_list, sample_list]
  
  # Edit sample names --> colnames
  for (i in 1:length(colnames(rlog_matrix))){
    sample_name <- colnames(rlog_matrix)[i]
    colnames(rlog_matrix)[i] <-  as.character(sampleTable2_names[sample_name,]$composite_feature_assigned)
  }
  
  # Modify gene names into symbols --> rownames
  rlog_matrix$gene_name <- rownames(rlog_matrix)
  rlog_matrix$gene_symbol <- sapply(rlog_matrix$gene_name, function(i) as.character(gene_name_table[i,]))
  # Append repeating gene symbols with counter
  rlog_matrix$gene_symbol <- make.names(rlog_matrix$gene_symbol, unique = T)
  rownames(rlog_matrix) <- rlog_matrix$gene_symbol
  rlog_matrix[,c("gene_name", "gene_symbol")] <- NULL
  
  # Normalize the rlog expressions
  if (grepl("Baseline", colnames(rlog_matrix))) { rlog_matrix <- sweep(rlog_matrix, 1, rowMeans(rlog_matrix[,colnames(rlog_matrix[grepl("Baseline", colnames(rlog_matrix))])]), "-")
  }
  else {  rlog_matrix <- sweep(rlog_matrix, 1, rowMeans(rlog_matrix), "-")  }
  
  return(rlog_matrix)
}

# define colors for the the scale
library(circlize)
library(ComplexHeatmap)

draw_complex_heatmap <- function(rlog_subset, title, sampleTable2_names){
  btable <- sampleTable2_names
  rownames(btable) <- btable$composite_feature_assigned

  study_arm <- as.character(sapply(colnames(rlog_subset), 
                                   function(i) as.character(btable[i,"rename_assgn_treatment"])))
  column_ha = HeatmapAnnotation( Study_arm = as.character(sapply(colnames(rlog_subset), 
                                                                 function(i) as.character(btable[i,"rename_assgn_treatment"]))),
                                 #BV = sapply(colnames(rlog_subset), function(i) as.character(btable[i,"BV"])),
                                 #CST = sapply(colnames(rlog_subset), function(i) as.character(btable[i,"CST"])),
                                 #Infl_category_PAM = sapply(colnames(rlog_subset), function(i) as.character(btable[i,"Infl_category_PAM"])),
                                 Ng = sapply(colnames(rlog_subset), function(i) as.character(btable[i,"Ng"])),
                                 col = list(Study_arm = c("Baseline" = "dimgrey", "NetEn" = "coral3", "COC" = "dodgerblue3", "CCVR" = "seagreen"), 
                                            #CST = c("CST-I" = "#FDCA89", "CST-III" = "#E3EA9C", "CST-IV" = "#666666"),
                                            #BV = c("Neg" = "maroon1", "Int" = "deepskyblue", "Pos" = "yellow"),
                                            #Infl_category_PAM = c("High" = "#cab2d6", "Low" = "#33a02c"),
                                            Ng = c("0" = "#cab2d6", "1" = "#33a02c")
                                 ))
  
  col_fun = colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))
  robust_dist = function(x, y) {
    qx = quantile(x, c(0.1, 0.9))
    qy = quantile(y, c(0.1, 0.9))
    l = x > qx[1] & x < qx[2] & y > qy[1] & y < qy[2]
    x = x[l]
    y = y[l]
    sqrt(sum((x - y)^2))
  }
  #width = ncol(rlog_subset)*unit(5, "cm") 
  #height = nrow(rlog_subset)*unit(5, "cm")
  #ncol(rlog_subset)
  ht = Heatmap(rlog_subset,
               #width = ncol(rlog_subset)*unit(5, "mm"),
               #height = nrow(rlog_subset)*unit(5, "mm"),
               name = title, 
               col = col_fun,
               column_title_side = "bottom", column_title_gp = gpar(fontsize = 18),
               column_names_gp = gpar(fontsize = 12), row_names_gp = gpar(fontsize = 12),
               row_title = "DEGs", row_title_gp = gpar(fontsize = 18),
               show_column_names = FALSE,
               row_title_rot = 0, 
               cluster_rows = TRUE, cluster_columns = TRUE, 
               show_column_dend = TRUE, column_dend_side = "top", 
               column_dend_height = unit(1, "cm"), row_dend_width = unit(1, "cm"), 
               row_dend_reorder = TRUE,
               clustering_distance_rows = robust_dist, 
               clustering_distance_columns = robust_dist,
               top_annotation = column_ha,
               column_split = study_arm)
  return(ht)
}

save_complex_heatmap <- function(rlog_all, res, dir_path, gp1_samples, gp2_samples, title){
  rlog_whole <- as.data.frame(assay(rlog_all))
  
  comparison <- unlist(strsplit(res, "_"))[2]
  #gp2_title <- unlist(strsplit(comparison, "vs"))[1]
  #gp1_title <- unlist(strsplit(comparison, "vs"))[2]
  samples_of_int <- c(as.character(gp1_samples), as.character(gp2_samples))
  sigGenes <- get_sigGenes(as.data.frame(get(res)))
  rlog_subset <- subsetnedit_rlog2(rlog_whole, sigGenes, samples_of_int)
  ht <- draw_complex_heatmap(rlog_subset, title, sampleTable2_names)
  
  r <- as.numeric(length(rownames(rlog_subset)))
  h <- r/5 
  if (h < 5) {h <- 5}
  w <- as.numeric(length(colnames(rlog_subset))/4) + 1 
  
  #pdf(paste0(dir_path,"ComplexHeatmap_", comparison, ".pdf"), width = w, height = h)
  #pdf(paste0(dir_path,"ComplexHeatmap_", comparison, "_ITT_with_Ng_colorbar.pdf"), width = w, height = h)
  pdf(paste0(dir_path,"ComplexHeatmap_", comparison, "_PP_with_Ng_colorbar.pdf"), width = w, height = h)
  draw(ht)
  dev.off()
  
  return(ht)
}

# Metadata
modify_sample_table <- function(input_table){
  
  atable <- input_table
  
  # join supertable_itt and inflammation categories PAM
  metadata_v2 <- read.table("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uCHOOSE_Metadata_v3_withSampledetails_n_InfPAM.txt", header = TRUE, row.names = NULL, check.names = FALSE, sep = "\t")
  atable <- join(atable, metadata_v2[c("Sample_ID", "Infl_category_PAM")], by="Sample_ID")
  rownames(atable) <- atable$Sample_ID
  
  atable$match <- "yes"
  atable[atable$Assigned_treatment != atable$Treatment,"match"] <- "no"
  atable$match <- factor(atable$match, levels = c("yes", "no"))
  
  atable$rename_assgn_treatment <- as.character(atable$Assigned_treatment)
  atable$rename_assgn_treatment <- gsub("Injection_NETEN", "NetEn", atable$rename_assgn_treatment)
  atable$rename_assgn_treatment <- gsub("Pill", "COC", atable$rename_assgn_treatment)
  atable$rename_assgn_treatment <- gsub("Ring", "CCVR", atable$rename_assgn_treatment)
  atable$rename_assgn_treatment <- gsub("Pretreatment", "Baseline", atable$rename_assgn_treatment)
  atable$rename_assgn_treatment <- factor(atable$rename_assgn_treatment, levels = c("Baseline", "NetEn", "COC", "CCVR"))
  
  # Generate new sample names
  atable$composite_feature_assigned <- paste(atable$SubjectID, atable$rename_assgn_treatment, sep = "_")
  #atable_subset <- atable[,c("composite_feature_assigned","rename_assgn_treatment", "BV", "CST", "match", )]
  return(atable)
}

sampleTable2_names <- modify_sample_table(superTable_itt)

# Sample groups
atable <- superTable_itt
gp_baseline <- atable$Sample_ID[atable$Assigned_treatment == "Pretreatment"]
gp_Injection <- atable$Sample_ID[atable$Assigned_treatment == "Injection_NETEN"]
gp_Pill <- atable$Sample_ID[atable$Assigned_treatment == "Pill"]
gp_Ring <- atable$Sample_ID[atable$Assigned_treatment == "Ring"]

gp_Pill_stayedonpill <- atable$Sample_ID[atable$Assigned_treatment == "Pill" & atable$match == "yes"]
gp_Pill_convertedfrompill <- atable$Sample_ID[atable$Assigned_treatment == "Pill" & atable$match == "no"]
gp_Ring_stayedonring <- atable$Sample_ID[atable$Assigned_treatment == "Ring" & atable$match == "yes"]
gp_Ring_convertedfromring <- atable$Sample_ID[atable$Assigned_treatment == "Ring" & atable$match == "no"]

# gene name table: Ensembl gene id as rownames, gene symbol 1st col
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)


dir_path <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/"
rlog_itt <- rlditt
# -------------
# Generate heatmaps
# -------------
save_complex_heatmap(rlog_itt, "resitt_InjectionvsPill", dir_path, gp_Pill, gp_Injection, "NetEn vs. COC")
save_complex_heatmap(rlog_itt, "resitt_RingvsPill", dir_path, gp_Pill, gp_Ring, "CCVR vs. COC")
save_complex_heatmap(rlog_itt, "resitt_RingvsInjection", dir_path, gp_Injection, gp_Ring, "CCVR vs. NetEn")

save_complex_heatmap(rlog_itt, "resitt_InjectionvsPre", dir_path, gp_baseline, gp_Injection, "NetEn vs. baseline")
save_complex_heatmap(rlog_itt, "resitt_PillvsPre", dir_path, gp_baseline, gp_Pill, "COC vs. baseline")
save_complex_heatmap(rlog_itt, "resitt_RingvsPre", dir_path, gp_baseline, gp_Ring, "CCVR vs. baseline")

```

#### GSEA, ITT

Look at code chunk under main analysis - gsea_ITT, gsea_enrichment_dot_plots_v2

*  ITT Combine GSEA stats for all comparisons for each pathway collection

```{r gsea_combine_stats_ITT, eval=FALSE}
# -----------------
# For each pathway collection, combine stats from each cross-treatment comparison
# -----------------

# Merge all results for a pathway collection
merge_df <- function(resA, resB) {
  merged_results <- Merge(resA, resB, by="row.names")
  row.names(merged_results) <- merged_results$Row.names
  merged_results[,"Row.names"] <- NULL
  return(merged_results)
}

modify_gsea_report <- function(resA, res_ext) {
  row.names(resA) <- resA$NAME
  #resA <- resA[,c("NES", "NOM p-val", "FDR q-val")]
  resA <- resA[,c("SIZE", "NES", "NOM p-val", "FDR q-val")]
  resA <- round(resA, digits = 3)
  #resA[,c("NAME","GS<br> follow link to MSigDB", "GS DETAILS")]<- NULL
  colnames(resA) <- gsub("$", paste0("_", res_ext), colnames(resA))
  return(resA)
}

# Compile GSEA stats into one table for each pathway collection
gsea_dir_itt1 <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/"
gsea_dir_itt2 <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/"

comparison_order <- c("Injection_vs_baseline", "Pill_vs_baseline", "Ring_vs_baseline", 
                      "Injection_vs_Pill", "Ring_vs_Pill", "Ring_vs_Injection")
pathway_coll1 <- c("Hallmark", "Kegg", "BioCarta", "c7Th17")
pathway_coll2 <- c("EBF","REACTOME","PID","GO","CSG")

#data_list <- list()
#gsea_dir <- gsea_dir_itt1
#pathway_coll <- pathway_coll1
gsea_dir <- gsea_dir_itt2
pathway_coll <- pathway_coll2
for (i in 1:length(pathway_coll)) {
  #a
  #f
  #data
  #edited_data
  #merged_data
  #i <- 1
  #j <- 1
  #colnames(merged_data)
  
  pathway <- pathway_coll[i]
  # merge results for each pathway coll
  for (j in 1:length(comparison_order)) {
    comparison <- comparison_order[j]
    print(pathway)
    print(comparison)
    
    dir_name <- paste(comparison, pathway, sep = "_")
    a <- dir(path = gsea_dir, recursive = FALSE, pattern = gsub("$", "*", dir_name))
    f <- list.files(path = paste0(gsea_dir, a), recursive = FALSE, pattern = "gsea_report_for.*\\.(xls|tsv)$", full.names = TRUE)
    
    data <- read.table(f[1], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")
    data <- rbind(data, read.table(f[2], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t"))
    
    # modify this gsea report
    edited_data <- modify_gsea_report(data, dir_name)
    # merge data
    if (j == 1) { merged_data <- edited_data}
    else {  merged_data <- merge_df(merged_data, edited_data)}
  }
  # push merged data in a list
  merged_data <- data.frame("NAME" = row.names(merged_data), merged_data)
  results_file <- paste0(gsea_dir, "compiled_gsea_", pathway, ".txt")
  write.table(merged_data, results_file, quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")
  #data_list[[i]] <- merged_data
}
#names(data_list) <- pathway_coll
```

* Enrichment dot plots

```{r gsea_enrichment_dot_plots_ITT, eval=FALSE}
# Compile GSEA stats into one table for each pathway collection

# ---------
# Functions
# ---------
# Function: To extract only required stats from GSEA output files
modify_gsea_report2 <- function(resA, res_ext) {
  resA <- resA[,c("NAME", "NES", "NOM p-val")]
  resA$comparison <- rep(res_ext, length(resA$NES))
  return(resA)
}

# Function: To combine GSEA stats for different comparison under the same pathway set into one table, row wise, and then put all such tables for different pathway sets into a list
compile_gsea_stats <- function(comparison_order, pathway_coll, gsea_dir) {
  data_list <- list()
  for (i in 1:length(pathway_coll)) {
    # merge results for each pathway coll
    for (j in 1:length(comparison_order)) {
      #i <- 1
      #j <- 1
      #pathway_coll <- pathway_coll2
      #pathway_coll
      #dir_name
      #a
      #f
      #gsea_dir <- gsea_dir_itt2
      #list.files(path = paste0(gsea_dir, a), recursive = FALSE, pattern = "gsea_report_for.*\\.tsv$", full.names = TRUE)
      #paste0(gsea_dir, a)
      #f
      #list.files( "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Injection_vs_baseline_CSG.Gsea.1606966230762", pattern = "gsea_report")
      #f
      dir_name <- paste(comparison_order[j], pathway_coll[i], sep = "_")
      a <- dir(path = gsea_dir, recursive = FALSE, pattern = gsub("$", "*", dir_name))
      f <- list.files(path = paste0(gsea_dir, a), recursive = FALSE, pattern = "gsea_report_for.*\\.(xls|tsv)$", full.names = TRUE)
      
      data <- read.table(f[1], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")
      data <- rbind(data, read.table(f[2], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t"))
      
      edited_data <- modify_gsea_report2(data, dir_name)
      # merge data from different comparisons but same collection
      if (j == 1) { merged_data <- edited_data}
      else {  merged_data <- rbind(merged_data, edited_data)}
    }
    # push merged data in a list
    #merged_data <- data.frame("NAME" = row.names(merged_data), merged_data)
    data_list[[i]] <- merged_data
  }
  names(data_list) <- pathway_coll
  return(data_list)
}

# Function: pull gsea stats table for a pathway set and specific comparisons, and generate a dotplot
get_enrichment_dotplot <- function(collection, comparison_order, data_list_input, outputname, save_dir){
#get_enrichment_dotplot <- function(collection, comparison_order, data_list_input, outputname, save_dir, enriched_pathways){
  #collection <- "GO"
  #comparison_order <- comparison_order_lng
  #data_list_input <- data_list_lng_itt2
  #outputname <- "longitudinal"
  #save_dir <- gsea_dir_itt_fig
  
  print(collection)
  gsea_stats <- data_list_input[[collection]]
  
  enriched_pathways <- unique(as.character(gsea_stats[gsea_stats$`NOM p-val` < 0.1,"NAME"]))
  #enriched_pathways <- unique(as.character(gsea_stats[gsea_stats$`NOM p-val` < 0.05,"NAME"]))
  print(length(enriched_pathways))
  gsea_stats <- gsea_stats[gsea_stats$NAME %in% enriched_pathways,]
  
  r <- as.numeric(length(unique(gsea_stats$NAME)))
  h <- r/5
  if (h < 5) {h <- 4}
  gsea_stats$comparison <- gsub(paste0("_", collection), "", gsea_stats$comparison)
  gsea_stats$comparison <- factor(gsea_stats$comparison, 
                                  levels = comparison_order)
  gsea_stats$NAME <- gsub(paste0(toupper(collection), "_"), "", gsea_stats$NAME)
  b <- gsea_stats[gsea_stats$comparison == rev(comparison_order)[1],]
  # define pathway order in dotplot
  gsea_stats$NAME <- factor(gsea_stats$NAME, levels = rev(b[order(b$`NOM p-val`),"NAME"]))
  
  gsea_stats %>% ggplot(aes(x=comparison, y=NAME, color=NES, size=1-log(`NOM p-val`))) + 
    geom_point(na.rm = TRUE) + 
    scale_color_gradientn(colors = c("blue","grey75","grey75","red"), limits = c(-2.8,2.8), breaks = c(-2,-1,0,1,2)) + 
    theme_bw() + 
    theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5),
          axis.text.x = element_text(angle = 30, hjust=1), 
          axis.text.y = element_text(vjust = 0.5, hjust=1)) + 
    labs(title=paste0("GSEA - ", collection, " pathways"), x="Contraceptive contrast", y="Pathway") +
    scale_radius("1-log(NOM p-val)", 
                range = c(1,5), 
                 breaks = c(2,4,6,8), limits = c(1,8.2), 
                 labels = c("0.37","0.05","0.007","<0.001"))
  # nom p-val 0.1
  #ggsave(paste0(save_dir, "GSEA_", collection, "_",outputname,"_stats_enriched.pdf"), height = h, width = 20, limitsize = FALSE)
  # nom p-val 0.05
  #ggsave(paste0(save_dir, "GSEA_", collection, "_",outputname,"_stats_enriched_v2.pdf"), height = h, width = 20, limitsize = FALSE)
  # selected pathways
  #ggsave(paste0(save_dir, "GSEA_", collection, "_",outputname,"_stats_enriched_v3.pdf"), height = 2, width = 10, limitsize = FALSE)
  # nom p-val 0.1, only cross-comparisons
  ggsave(paste0(save_dir, "GSEA_", collection, "_",outputname,"_stats_enriched_nomp_point1_v4.pdf"), height = h, width = 8, limitsize = FALSE)
}

#names(data_list2)
# "Hallmark"       "Kegg"           "BioCarta"       "CustomGeneSets" "c7Th17" 

# -------------
# Generate stats tables
# -------------
# To get input table for generating dotplot using ggplot
gsea_dir_itt1 <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/"
gsea_dir_itt2 <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/"
gsea_dir_itt_fig <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Enrichment_stats_dotplots_ITT/"

pathway_coll1 <- c("Hallmark", "Kegg", "BioCarta", "c7Th17")
pathway_coll2 <- c("CSG", "EBF", "REACTOME", "PID", "GO")

comparison_order_all <- c("Injection_vs_baseline", "Pill_vs_baseline", "Ring_vs_baseline", 
                          "Injection_vs_Pill", "Ring_vs_Pill", "Ring_vs_Injection")
comparison_order_lng <- c("Injection_vs_baseline", "Pill_vs_baseline", "Ring_vs_baseline")
comparison_order_cross <- c("Injection_vs_Pill", "Ring_vs_Pill", "Ring_vs_Injection")

#data_list_all_itt <- compile_gsea_stats(comparison_order_all, pathway_coll, gsea_dir_itt)
data_list_lng_itt1 <- compile_gsea_stats(comparison_order_lng, pathway_coll1, gsea_dir_itt1)
data_list_cross_itt1 <- compile_gsea_stats(comparison_order_cross, pathway_coll1, gsea_dir_itt1)
names(data_list_lng_itt2)

data_list_lng_itt2 <- compile_gsea_stats(comparison_order_lng, pathway_coll = pathway_coll2, gsea_dir = gsea_dir_itt2)
data_list_cross_itt2 <- compile_gsea_stats(comparison_order_cross, pathway_coll2, gsea_dir_itt2)

names(data_list_lng_itt1)
length(unique(data_list_lng_itt1[[1]]$NAME))
length(unique(data_list_lng_itt1[[2]]$NAME))
length(unique(data_list_lng_itt1[[3]]$NAME))
length(unique(data_list_lng_itt1[[4]]$NAME))

names(data_list_lng_itt2)
length(unique(data_list_lng_itt2[[1]]$NAME))
length(unique(data_list_lng_itt2[[2]]$NAME))
length(unique(data_list_lng_itt2[[3]]$NAME))
length(unique(data_list_lng_itt2[[4]]$NAME))
length(unique(data_list_lng_itt2[[5]]$NAME))
# --------
# Enrichment dotplots 
# --------
# cd GSEA
# mkdir Enrichment_stats_dotplots
for (p in pathway_coll1) {
  print(p)
  #get_enrichment_dotplot(p, comparison_order_all, data_list_all_itt, "all6", gsea_dir_itt)
  #get_enrichment_dotplot(p, comparison_order_lng, data_list_lng_itt1, "longitudinal", gsea_dir_itt_fig)
  get_enrichment_dotplot(p, comparison_order_cross, data_list_cross_itt1, "crosssectional", gsea_dir_itt_fig)
}
# set 2 collections
for (p in pathway_coll2) {
  print(p)
  #get_enrichment_dotplot(p, comparison_order_lng, data_list_lng_itt2, "longitudinal", gsea_dir_itt_fig)
  get_enrichment_dotplot(p, comparison_order_cross, data_list_cross_itt2, "crosssectional", gsea_dir_itt_fig)
}

# get plots for selected pathways

# GO
enriched_pathways <- c("GO_POSITIVE_REGULATION_OF_KERATINOCYTE_DIFFERENTIATION", "GO_T_HELPER_17_TYPE_IMMUNE_RESPONSE", "GO_REGULATION_OF_ACUTE_INFLAMMATORY_RESPONSE")
get_enrichment_dotplot("GO", comparison_order_cross, data_list_cross_itt2, "crosssectional", gsea_dir_itt_fig, enriched_pathways)

# Reactome
names(data_list_cross_itt2)
data_list_cross_itt2[[3]][grepl("B_CELL",data_list_cross_itt2[[3]]$NAME),]
enriched_pathways <- c("REACTOME_SIGNALING_BY_THE_B_CELL_RECEPTOR_BCR_", "REACTOME_GAP_JUNCTION_ASSEMBLY", "REACTOME_REGULATION_OF_TLR_BY_ENDOGENOUS_LIGAND")
get_enrichment_dotplot("REACTOME", comparison_order_cross, data_list_cross_itt2, "crosssectional", gsea_dir_itt_fig, enriched_pathways)

# For pathways enriched in Ring
names(data_list_cross_itt2)
names(data_list_cross_itt1)
df <- data.frame()
# PID IL6, KEGG cytosolic dna sensing, Biocarta NFKB
df1 <- data_list_cross_itt2[[4]][grepl("IL6",data_list_cross_itt2[[4]]$NAME),]
df2 <- data_list_cross_itt1[[2]][grepl("CYTOSOLIC",data_list_cross_itt1[[2]]$NAME),]
df3 <- data_list_cross_itt1[[3]][grepl("NFKB",data_list_cross_itt1[[3]]$NAME),]

gsea_stats <- rbind(df1, df2, df3)
r <- as.numeric(length(unique(gsea_stats$NAME)))
h <- r/5
if (h < 5) {h <- 4}

gsea_stats
gsea_stats$comparison <- gsub("PID|Kegg|BioCarta", "", gsea_stats$comparison)
#gsea_stats$comparison <- gsub("_$", "", gsea_stats$comparison)
gsea_stats$comparison <- gsub("Injection", "Net-En", gsea_stats$comparison)
gsea_stats$comparison <- gsub("Pill", "COC", gsea_stats$comparison)
gsea_stats$comparison <- gsub("Ring", "CCVR", gsea_stats$comparison)
gsea_stats$comparison <- gsub("_", " ", gsea_stats$comparison)
gsea_stats$NAME <- factor(gsea_stats$NAME, levels = rev(c("PID_IL6_7_PATHWAY", "KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY", "BIOCARTA_NFKB_PATHWAY")))
gsea_stats$comparison <- factor(gsea_stats$comparison, levels = c("CCVR vs Net-En ", "CCVR vs COC ", "Net-En vs COC "))
  
gsea_stats %>% ggplot(aes(x=comparison, y=NAME, color=NES, size=1-log(`NOM p-val`))) + 
    geom_point(na.rm = TRUE) + 
    scale_color_gradientn(colors = c("blue","grey75","grey75","red"), limits = c(-2.8,2.8), breaks = c(-2,-1,0,1,2)) + 
    theme_bw() + 
    theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5),
          axis.text.x = element_text(angle = 30, hjust=1), 
          axis.text.y = element_text(vjust = 0.5, hjust=1)) + 
    labs(title="GSEA - Pathways enriched in CCVR", x="Contraceptive contrast", y="Pathway") +
    scale_radius("1-log(NOM p-val)", 
                range = c(1,5), 
                 breaks = c(2,4,6,8), limits = c(1,8.2), 
                 labels = c("0.37","0.05","0.007","<0.001"))
  # selected pathways
  ggsave(paste0(gsea_dir_itt_fig, "GSEA_pathways_in_ring_v3.pdf"), height = 2, width = 7, limitsize = FALSE)

  # Greg's code for enrichment dotplot
#gseaTable <- read.table("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_extra/exampleGSEAtable.txt", check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")

#View(gseaTable)
#gseaTable %>% ggplot(aes(x=topPhenotype, y=GENESET, color=NES, size=1-log(`NOM p-val`))) + 
#  geom_point(na.rm = TRUE) + 
#  scale_color_gradientn(colors = c("blue","grey75","grey75","red"), limits = c(-2.8,2.8), breaks = c(-2,-1,0,1,2)) + 
#  theme_bw() + 
#  theme(axis.text.y = element_text(angle = 30, vjust = 0.5, hjust=1, size = 4, face = "bold")) + 
#  scale_radius("1-log(NOM p-val)", 
#               range = c(1,(12-0.3*length(unlist(c(na.omit(GSEA_MET[,name])))))), 
#               breaks =  c(2,4,6,8), limits = c(1,8.2), 
#               labels = c("0.37","0.05","0.007","<0.001"))


```

* Heatmaps for all pathways

```{r gsea_heatmaps_itt, eval=FALSE}
# Generate heatmaps for leading-edge genes in enriched pathways

# Heatmap for a specific comparison

# From a GSEA folder - 
# * gsea_report_file
# * .xls files for each pathway

# Heatmaps
# --------
# Function: Subset and edit rlog gene expression for a specific set of genes and samples
subsetnedit_rlog2 <- function(rlog_input, gene_list, sample_list){
  # Subset for genes and samples
  rlog_matrix <- rlog_input[gene_list, sample_list]
  # Edit sample names --> colnames
  for (i in 1:length(colnames(rlog_matrix))){
    sample_name <- colnames(rlog_matrix)[i]
    colnames(rlog_matrix)[i] <-  as.character(sampleTable2_names[sample_name,]$composite_feature_assigned)
  }
  # Modify gene names into symbols --> rownames
  rlog_matrix$gene_name <- rownames(rlog_matrix)
  rlog_matrix$gene_symbol <- sapply(rlog_matrix$gene_name, function(i) as.character(gene_name_table[i,]))
  # Append repeating gene symbols with counter
  rlog_matrix$gene_symbol <- make.names(rlog_matrix$gene_symbol, unique = T)
  rownames(rlog_matrix) <- rlog_matrix$gene_symbol
  rlog_matrix[,c("gene_name", "gene_symbol")] <- NULL
  # Normalize the rlog expressions
  rlog_matrix <- sweep(rlog_matrix, 1, rowMeans(rlog_matrix), "-")
  #rlog_matrix <- sweep(rlog_matrix, 1, rowMeans(rlog_subset[,colnames(rlog_matrix[grepl("Pretreatment", colnames(rlog_matrix))])]), "-")
  return(rlog_matrix)
}

# Generate heatmap
source("/yerkes-cifs/runs/MSigDB/Heatmap_finalv2.R")

# -----
# Input data
# -----
# Get the rlog expressions for all samples/genes
rlog_itt <- rlditt
rlog_whole <- as.data.frame(assay(rlditt))

# gene name table: Ensembl gene id as rownames, gene symbol 1st col
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)

# Metadata
atable <- superTable_itt
atable$match <- "yes"
atable[atable$Assigned_treatment != atable$Treatment,"match"] <- "no"
atable$match <- factor(atable$match, levels = c("yes", "no"))

# Generate new sample names
atable$composite_feature_assigned <- paste(atable$SubjectID, atable$Visit_no, atable$Assigned_treatment, sep = "_")
sampleTable2_names <- atable[,c("composite_feature", "composite_feature_assigned","Assigned_treatment","Treatment", "BV", "CST", "match")]

# Sample groups
gp_baseline <- atable$Sample_ID[atable$Assigned_treatment == "Pretreatment"]
gp_Injection <- atable$Sample_ID[atable$Assigned_treatment == "Injection_NETEN"]
gp_Pill <- atable$Sample_ID[atable$Assigned_treatment == "Pill"]
gp_Pill_stayedonpill <- atable$Sample_ID[atable$Assigned_treatment == "Pill" & atable$match == "yes"]
gp_Pill_convertedfrompill <- atable$Sample_ID[atable$Assigned_treatment == "Pill" & atable$match == "no"]
gp_Ring <- atable$Sample_ID[atable$Assigned_treatment == "Ring"]
gp_Ring_stayedonring <- atable$Sample_ID[atable$Assigned_treatment == "Ring" & atable$match == "yes"]
gp_Ring_convertedfromring <- atable$Sample_ID[atable$Assigned_treatment == "Ring" & atable$match == "no"]

# Initialize variables
par_dir <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/"

# ----------
# Enrichment plots
# ---------
# Post vs. Pre comparisons
comparison_order_lng <- c("Injection_vs_baseline", "Pill_vs_baseline", "Ring_vs_baseline")
# Cross-sectional comparisons
comparison_order_cross <- c("Injection_vs_Pill", "Ring_vs_Pill", "Ring_vs_Injection")

# Inputs
gsea_dir_itt <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/"
gsea_dir_itt2 <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/"

gsea_dir <- gsea_dir_itt
gsea_dir <- gsea_dir_itt2
# Get lists of pathway names for each collection
#pathway_coll <- c("Hallmark", "Kegg", "BioCarta", "CustomGeneSets", "c7Th17")
pathway_coll <- c("Hallmark", "Kegg", "BioCarta")
pathway_coll <- c("PID")
list_of_pathway_names <- list()
comparison_order <- comparison_order_lng
comparison_order <- comparison_order_cross
comparison_order
for (i in 1:length(pathway_coll)) {
  dir_name <- paste(comparison_order[1], pathway_coll[i], sep = "_")
  dir_path <- paste0(gsea_dir, dir(path = gsea_dir, recursive = FALSE, pattern = gsub("$", "*", dir_name)))
  dir_name
  dir_path
  # Method 1
  # List out all pathway xls files 
  #reg_exp_for_pathways <- paste(toupper(pathway_coll[i]), ".*\\.xls$", sep = "_")
  #g <- list.files(path = dir_path, recursive = FALSE, pattern = reg_exp_for_pathways, full.names = TRUE)
  #pathway_names <- sapply(g, function (k) {str_match(k, paste0("/(", toupper(pathway_coll[i]), ".*).xls"))[,2]})
  #names(pathway_names) <- NULL
  # Method 2
  # get list of pathways from gsea report
  g <- list.files(path = dir_path, recursive = FALSE, pattern = "gsea_report_for.*\\.(xls|tsv)$", full.names = TRUE)
  
  data <- read.table(g[1], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")
  data <- rbind(data, read.table(g[2], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t"))
  pathway_names <- as.character(data$NAME)
  list_of_pathway_names[[pathway_coll[i]]] <- pathway_names
}
list_of_pathway_names

# ----------
# Cross-sectional comparisons

# Merge GSEA stats for different comparisons, same pathway in a collection
get_gsea_stats_n_heatmap_for_a_collection <- function(comparison_order, collection, list_of_pathway_names, 
                                                      analysis_type, analysis_type2, separate_heatmaps, gsea_dir){
  gsea_data_list <- list()
  #collection <- "Kegg"
  #comparison_order <- comparison_order_lng
  #analysis_type
  #analysis_type2
  #separate_heatmaps = "yes"
  # For every pathway in a collection
  for (i in 1:length(list_of_pathway_names[[collection]])) {
    #i <- 1
    gene_set <- list_of_pathway_names[[collection]][i]
    #gene_set
    # Extract gene_set_tables for every comparison
    for (j in 1:length(comparison_order)) {
      #j <- 1
      comparison <- comparison_order[j]
      dir_name <- paste(comparison, collection, sep = "_")
      f <- list.files(path = paste0(gsea_dir, dir(path = gsea_dir, recursive = FALSE, pattern = gsub("$", "*", dir_name))), 
                      recursive = FALSE, pattern = paste0(gene_set,"\\.(xls|tsv)$"), full.names = TRUE)
      #dir_name
      #f
      
      # get dir name
      gene_set_table <- read.table(f, check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")
      # drop the last col, it's all NAs
      
      head(gene_set_table)
      if (colnames(gene_set_table)[2] == "SYMBOL"){
        gene_set_table <- gene_set_table[,-c(8)]
        gene_set_table$`PROBE` <- gene_set_table$SYMBOL
      }
      else {gene_set_table <- gene_set_table[,-c(10)]}
      gene_set_table$`PROBE` <- factor(gene_set_table$`PROBE`, gene_set_table$`PROBE`)
      total_no_genes <- length(unique(gene_set_table$PROBE))
      enriched_genes <- as.character(unique(gene_set_table$PROBE[gene_set_table$`CORE ENRICHMENT` == "Yes"]))
      gene_set_table <- gene_set_table[gene_set_table$`CORE ENRICHMENT` == "Yes",]
      #gene_set_table
      gene_set_table$comparison <- comparison_order[j]
      gene_set_table$pathway <- gene_set
      gene_set_table$analysis <- analysis_type
      
      # Merge GSEA stats from different comparisons
      if (j == 1) { merged_data <- gene_set_table}
      else {  merged_data <- rbind(merged_data, gene_set_table)}
      #merged_data
      # To generate heatmap for every comparison
      
      if (separate_heatmaps == "yes") {
    
        # Get gene list
        gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)
        gene_name_table$gene_id <- rownames(gene_name_table)
        genes_ensembl_list <- gene_name_table[gene_name_table$gene_name %in% enriched_genes, "gene_id"]
        gene_name_table$gene_id <- NULL
        
        # Get sample list
        atable <- superTable_itt
        numerator <- unlist(str_split(comparison_order[j], "_vs_"))[1]
        gp_numerator <- atable$Sample_ID[grepl(numerator, atable$Assigned_treatment)]
        denominator <- unlist(str_split(comparison_order[j], "_vs_"))[2]
        denominator <- gsub("baseline", "Pretreatment", denominator)
        gp_denominator <- atable$Sample_ID[grepl(denominator, atable$Assigned_treatment)]
        samples_of_int <- c(as.character(gp_denominator), as.character(gp_numerator))
        
        # Get rlog expressions for these genes and samples
        rlog_all <- as.data.frame(assay(rlditt))
        rlog_subset <- subsetnedit_rlog2(rlog_all, genes_ensembl_list, samples_of_int)
        #rlog_subset
        # Generate heatmap
        r <- as.numeric(length(rownames(rlog_subset)))
        h <- r/5 
        if (h < 5) {h <- 5}
        w <- as.numeric(length(gp_denominator + gp_denominator))/2 + 1
        #gp_denominator
        #h
        #r
        #w
        
        new_name <- gsub("_", " ", gene_set)
        main_title <- paste0(analysis_type, " GSEA\n", comparison_order[j], "\n", new_name, 
                             "\n Enriched genes: ", length(enriched_genes), ", Total genes: ", total_no_genes)
        if (length(enriched_genes) > 1) {
          plot_heatmap <- heatmap_clust_vline_order(rlog_subset, dendro = "y", title = main_title, limits = c(-1,1), 
                                                    vlines = c(length(gp_denominator) + 0.5), 
                                                    vline_size = 1.5)
          save_file_as =  paste0(par_dir, "GSEA_ITT_figures/", collection, "_", analysis_type2, 
                                 "/Heatmap_", analysis_type, "_", gene_set, "_", comparison, ".pdf")
          #save_file_as
          #collection
          #analysis_type2 <- "cross_sectional"
          ggsave(plot=plot_heatmap, filename = save_file_as, title = main_title, height = h, width = w, limitsize = FALSE)
          #In the next loop add the next pathway results
        }
      }
    }
    # merged data contains gsea stats for all comparisons and the same pathway
    gsea_data_list[[i]] <- merged_data
    
    # Heatmap for union set of leading edge genes
    #gene_set_table <- merged_data
    #total_no_genes <- length(unique(gene_set_table$PROBE))
    #enriched_genes <- as.character(unique(gene_set_table$PROBE[gene_set_table$`CORE ENRICHMENT` == "Yes"]))
  }
  names(gsea_data_list) <- list_of_pathway_names[[collection]]
  return(gsea_data_list)
}

# Inputs
#gsea_dir_itt <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/"
#gsea_dir <- gsea_dir_itt
gsea_dir_itt2 <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/"


h_gsea_data <- list()
b_gsea_data <- list()
k_gsea_data <- list()
p_gsea_data <- list()

h_gsea_data <- get_gsea_stats_n_heatmap_for_a_collection(comparison_order_lng, "Hallmark", list_of_pathway_names, analysis_type = "ITT", analysis_type2 = "post_vs_pre", separate_heatmaps = "yes", gsea_dir_itt)
b_gsea_data <- get_gsea_stats_n_heatmap_for_a_collection(comparison_order_lng, "BioCarta", list_of_pathway_names, analysis_type = "ITT", analysis_type2 = "post_vs_pre", separate_heatmaps = "yes", gsea_dir_itt)
k_gsea_data <- get_gsea_stats_n_heatmap_for_a_collection(comparison_order_lng, "Kegg", list_of_pathway_names, analysis_type = "ITT", analysis_type2 = "post_vs_pre", separate_heatmaps = "yes", gsea_dir_itt)
p_gsea_data <- get_gsea_stats_n_heatmap_for_a_collection(comparison_order_lng, "PID", list_of_pathway_names, analysis_type = "ITT", analysis_type2 = "post_vs_pre", separate_heatmaps = "yes", gsea_dir_itt2)

k_gsea_data2 <- get_gsea_stats_n_heatmap_for_a_collection(comparison_order_lng, "Kegg", list_of_pathway_names, analysis_type = "ITT", analysis_type2 = "post_vs_pre", separate_heatmaps = "yes", gsea_dir_itt)

h2_gsea_data <- list()
b2_gsea_data <- list()
k2_gsea_data <- list()
p2_gsea_data <- list()

h2_gsea_data <- get_gsea_stats_n_heatmap_for_a_collection(comparison_order_cross, "Hallmark", list_of_pathway_names, analysis_type = "ITT", analysis_type2 = "cross_sectional", separate_heatmaps = "yes", gsea_dir_itt)
b2_gsea_data <- get_gsea_stats_n_heatmap_for_a_collection(comparison_order_cross, "BioCarta", list_of_pathway_names, analysis_type = "ITT", analysis_type2 = "cross_sectional", separate_heatmaps = "yes", gsea_dir_itt)
k2_gsea_data <- get_gsea_stats_n_heatmap_for_a_collection(comparison_order_cross, "Kegg", list_of_pathway_names, analysis_type = "ITT", analysis_type2 = "cross_sectional", separate_heatmaps = "yes", gsea_dir_itt)
p2_gsea_data <- get_gsea_stats_n_heatmap_for_a_collection(comparison_order_cross, "PID", list_of_pathway_names, analysis_type = "ITT", analysis_type2 = "cross_sectional", separate_heatmaps = "yes", gsea_dir_itt2)

# ------------- #
# For PID IL6
list_of_pathway_names
a[["PID"]] <- list_of_pathway_names[[1]][66]
p3_gsea_data <- get_gsea_stats_n_heatmap_for_a_collection(comparison_order_cross, "PID", a, analysis_type = "ITT", analysis_type2 = "cross_sectional", separate_heatmaps = "yes", gsea_dir_itt2)

```

* Enrichment line plots, one plot for each pathway

```{r gsea_line_plots_single_ITT, eval=FALSE}
# --------
# Single enrichment plot
# --------
source("/yerkes-cifs/runs/MSigDB/GSEA_enrich_auto_vX1.R")
# tweaked the code to get ylim upto y_outreach
# tweaked the code to set phen1 and phen2 heights

# --------
# Single enrichment plot
# --------#
source("/yerkes-cifs/runs/MSigDB/GSEA_enrich_auto_vX1.R")
# tweaked the code to get ylim upto y_outreach
# tweaked the code to set phen1 and phen2 heights

# ---------------------------- #
#Hallmark - longitudinal
gene_set <- "HALLMARK_IL6_JAK_STAT3_SIGNALING"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.42
y2 <- 0.42
y3 <- 0.42
y4 <- 0.7

gene_set <- "HALLMARK_INFLAMMATORY_RESPONSE"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.48
y2 <- 0.48
y3 <- 0.48
y4 <- 0.7

gene_set <- "HALLMARK_TGF_BETA_SIGNALING"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.43
y2 <- 0.43
y3 <- 0.43
y4 <- 0.6

gene_set <- "HALLMARK_TNFA_SIGNALING_VIA_NFKB"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.43
y2 <- 0.43
y3 <- 0.43
y4 <- 0.7

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Injection_vs_baseline_Hallmark.Gsea.1604125082255/", geneset=gene_set, color1="coral3", phen1="Higher in Net-En", phen2="Higher in control", label1_y_pos = y1, label1="Net-En vs. baseline", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Hallmark_post_vs_pre/ITT_", gene_set, "_Injection_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. control\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Pill_vs_baseline_Hallmark.Gsea.1604125981847/", geneset=gene_set, color1="dodgerblue3", phen1="Higher in COC", phen2="Higher in control", label1_y_pos = y2, label1="COC vs. baseline", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Hallmark_post_vs_pre/ITT_2_", gene_set, "_Pill_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nCOC vs. baseline\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Ring_vs_baseline_Hallmark.Gsea.1604125614270/", geneset=gene_set, color1="seagreen", phen1="Higher in CCVR", phen2="Higher in control", label1_y_pos = y3, label1="CCVR vs. baseline", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Hallmark_post_vs_pre/ITT_", gene_set, "_Ring_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nCCVR vs. baseline\n", gene_set_name), y_outreach = y4)

# -------------------------------------#
#source("/yerkes-cifs/runs/MSigDB/GSEA_enrich_auto_vX.R")
#Hallmark - cross-arm
#gene_set <- "HALLMARK_DNA_REPAIR"
gene_set <- "HALLMARK_TGF_BETA_SIGNALING"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.25
y2 <- 0.25
y3 <- 0.25
y4 <- 0.3

gene_set <- "HALLMARK_TNFA_SIGNALING_VIA_NFKB"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.43
y2 <- 0.43
y3 <- 0.43
y4 <- 0.7

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Injection_vs_Pill_Hallmark.Gsea.1604126813307/", geneset=gene_set, color1="#d8b365",phen1="Higher in Net-En", phen2="Higher in COC", label1_y_pos = y1, label1="Net-En vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Hallmark_cross_sectional/ITT_", gene_set, "_Injection_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Ring_vs_Pill_Hallmark.Gsea.1604127431705/", geneset=gene_set, color1="#5ab4ac", phen1="Higher in CCVR", phen2="Higher in COC", label1_y_pos = y2, label1="CCVR vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Hallmark_cross_sectional/ITT_", gene_set, "_Ring_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA", "\nCCVR vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Ring_vs_Injection_Hallmark.Gsea.1604127813035/", geneset=gene_set, color1="#6a51a3", phen1="Higher in CCVR", phen2="Higher in Net-En", label1_y_pos = y3, label1="CCVR vs. Net-En", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Hallmark_cross_sectional/ITT_", gene_set, "_Ring_vs_Injection_lineplot"), mainTitle = paste0("ITT GSEA", "\nCCVR vs. Net-En\n", gene_set_name), y_outreach = y4)

# -----------------------#
#BioCarta post vs. pre
gene_set <- "BIOCARTA_NFKB_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.35
y2 <- 0.35
y3 <- 0.35
y4 <- 0.7

gene_set <- "BIOCARTA_TNFR2_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.35
y2 <- 0.35
y3 <- 0.35
y4 <- 0.7

gene_set <- "BIOCARTA_CD40_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.3
y2 <- 0.3
y3 <- 0.3
y4 <- 0.7

gene_set <- "BIOCARTA_CXCR4_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.2
y2 <- 0.4
y3 <- 0.4
y4 <- 0.7

gene_set <- "BIOCARTA_CCR5_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.2
y2 <- 0.4
y3 <- 0.4
y4 <- 0.7

gene_set <- "BIOCARTA_CTLA4_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.4
y2 <- 0.4
y3 <- 0.4
y4 <- 0.7


One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Injection_vs_baseline_BioCarta.Gsea.1604125244201/", geneset=gene_set, color1="coral3", phen1="Higher in Net-En", phen2="Higher in non-HC", label1_y_pos = y1, label1="Net-En vs. non-HC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/BioCarta_post_vs_pre/ITT_", gene_set, "_Injection_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. non-HC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Pill_vs_baseline_BioCarta.Gsea.1604126060280/", geneset=gene_set, color1="dodgerblue3", phen1="Higher in COC", phen2="Higher in non-HC", label1_y_pos = y2, label1="COC vs. non-HC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/BioCarta_post_vs_pre/ITT_", gene_set, "_Pill_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nCOC vs. non-HC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Ring_vs_baseline_BioCarta.Gsea.1604125664960/", geneset=gene_set, color1="seagreen", phen1="Higher in CCVR", phen2="Higher in non-HC", label1_y_pos = y3, label1="CCVR vs. non-HC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/BioCarta_post_vs_pre/ITT_", gene_set, "_Ring_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA\n", "\nCCVR vs. non-HC\n", gene_set_name), y_outreach = y4)


# ---------------#
# Biocarta cross-sectional
gene_set <- "BIOCARTA_BCELLSURVIVAL_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.4
y2 <- 0.2
y3 <- 0.4
y4 <- 0.7

gene_set <- "BIOCARTA_NFKB_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.4
y2 <- 0.2
y3 <- 0.2
y4 <- 0.7

gene_set <- "BIOCARTA_IL6_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.5
y2 <- 0.2
y3 <- 0.5
y4 <- 0.7

gene_set <- "BIOCARTA_HIVNEF_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.5
y2 <- 0.5
y3 <- 0.5
y4 <- 0.7

gene_set <- "BIOCARTA_INFLAM_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.5
y2 <- 0.5
y3 <- 0.5
y4 <- 0.7

gene_set <- "BIOCARTA_TNFR2_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.35
y4 <- 0.7

gene_set <- "BIOCARTA_CD40_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.3
y2 <- 0.3
y3 <- 0.3
y4 <- 0.7

gene_set <- "BIOCARTA_RAS_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.3
y4 <- 0.7

gene_set <- "BIOCARTA_CXCR4_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.4
y2 <- 0.4
y3 <- 0.4
y4 <- 0.7

gene_set <- "BIOCARTA_CCR5_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.3
y2 <- 0.4
y3 <- 0.4
y4 <- 0.7

gene_set <- "BIOCARTA_CTLA4_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.3
y2 <- 0.3
y3 <- 0.4
y4 <- 0.7

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Injection_vs_Pill_BioCarta.Gsea.1604126908449/", geneset=gene_set, color1="#d8b365", phen1="Higher in Net-En", phen2="Higher in COC", label1_y_pos = y1, label1="Net-En vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/BioCarta_cross_sectional/ITT_", gene_set, "_Injection_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Ring_vs_Pill_BioCarta.Gsea.1604127524342/", geneset=gene_set, color1="#5ab4ac", phen1="Higher in CCVR", phen2="Higher in COC", label1_y_pos = y2, label1="CCVR vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/BioCarta_cross_sectional/ITT_", gene_set, "_Ring_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA\n", "\nCCVR vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Ring_vs_Injection_BioCarta.Gsea.1604127907912/", geneset=gene_set, color1="#6a51a3", phen1="Higher in CCVR", phen2="Higher in Net-En", label1_y_pos = y3, label1="CCVR vs. Net-En", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/BioCarta_cross_sectional/ITT_", gene_set, "_Ring_vs_Injection_lineplot"), mainTitle = paste0("ITT GSEA", "\nCCVR vs. Net-En\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Injection_vs_Ring_BIOCARTA.Gsea.1610052869015/", geneset=gene_set, color1="#6a51a3", phen1="Higher in Net-En", phen2="Higher in CCVR", label1_y_pos = 0.45, label1="Net-En vs. CCVR", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/BioCarta_cross_sectional/ITT_", gene_set, "_Injection_vs_Ring_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. CCVR\n", gene_set_name), y_outreach = y4)

# ---------------------#
#Kegg post vs. pre
gene_set <- "KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.42
y2 <- 0.42
y3 <- 0.42
y4 <- 0.7

gene_set <- "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.45
y2 <- 0.45
y3 <- 0.45
y4 <- 0.7

gene_set <- "KEGG_GALACTOSE_METABOLISM"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.53
y2 <- 0.53
y3 <- 0.53
y4 <- 0.7

gene_set <- "KEGG_MAPK_SIGNALING_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.3
y2 <- 0.3
y3 <- 0.3
y4 <- 0.7

gene_set <- "KEGG_P53_SIGNALING_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.3
y4 <- 0.7

gene_set <- "KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM"
gene_set_name <- gsub("_"," ", gene_set)
y2 <- 0.3
y4 <- 0.7

gene_set <- "KEGG_O_GLYCAN_BIOSYNTHESIS"
gene_set_name <- gsub("_"," ", gene_set)
y2 <- 0.5
y4 <- 0.7

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Injection_vs_baseline_Kegg.Gsea.1604125348848/", geneset=gene_set, color1="coral3", phen1="Higher in Net-En", phen2="Higher in non-HC", label1_y_pos = y1, label1="Net-En vs. non-HC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Kegg_post_vs_pre/ITT_", gene_set, "_Injection_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. non-HC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Pill_vs_baseline_Kegg.Gsea.1604126168597/", geneset=gene_set, color1="dodgerblue3", phen1="Higher in COC", phen2="Higher in non-HC", label1_y_pos = y2, label1="COC vs. non-HC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Kegg_post_vs_pre/ITT_", gene_set, "_Pill_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nCOC vs. non-HC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Ring_vs_baseline_Kegg.Gsea.1604125744915/", geneset=gene_set, color1="seagreen", phen1="Higher in CCVR", phen2="Higher in non-HC", label1_y_pos = y3, label1="CCVR vs. non-HC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Kegg_post_vs_pre/ITT_", gene_set, "_Ring_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nCCVR vs. non-HC\n", gene_set_name), y_outreach = y4)

# --------------------#
# KEGG cross-sectional
gene_set <- "KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.42
y2 <- 0.42
y3 <- 0.42
y4 <- 0.7

gene_set <- "KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.53
y2 <- 0.53
y3 <- 0.53
y4 <- 0.7

gene_set <- "KEGG_P53_SIGNALING_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.3
y4 <- 0.7

gene_set <- "KEGG_MAPK_SIGNALING_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.3
y4 <- 0.7


One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Injection_vs_Pill_Kegg.Gsea.1604126985714/", geneset=gene_set, color1="#d8b365", phen1="Higher in Net-En", phen2="Higher in COC", label1_y_pos = y1, label1="Net-En vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Kegg_cross_sectional/ITT_", gene_set, "_Injection_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Ring_vs_Pill_Kegg.Gsea.1604127596046/", geneset=gene_set, color1="#5ab4ac", phen1="Higher in CCVR", phen2="Higher in COC", label1_y_pos = y2, label1="CCVR vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Kegg_cross_sectional/ITT_", gene_set, "_Ring_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA\n", "\nCCVR vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT/Ring_vs_Injection_Kegg.Gsea.1604127980445/", geneset=gene_set, color1="#6a51a3", phen1="Higher in CCVR", phen2="Higher in Net-En", label1_y_pos = y3, label1="CCVR vs. Net-En", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/Kegg_cross_sectional/ITT_", gene_set, "_Ring_vs_Injection_lineplot"), mainTitle = paste0("ITT GSEA", "\nCCVR vs. Net-En\n", gene_set_name), y_outreach = y4)

# ---------------------------- #
# PID post vs. pre
gene_set <- "PID_IL6_7_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.53
y2 <- 0.53
y3 <- 0.53
y4 <- 0.7

gene_set <- "PID_WNT_CANONICAL_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.5
y2 <- 0.5
y3 <- 0.5
y4 <- 0.7

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Injection_vs_baseline_PID.Gsea.1606949598945/", geneset=gene_set, color1="coral3", phen1="Higher in Net-En", phen2="Higher in control", label1_y_pos = y1, label1="Net-En vs. control", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/PID_post_vs_pre/ITT_", gene_set, "_Injection_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. control\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Pill_vs_baseline_PID.Gsea.1606953159695/", geneset=gene_set, color1="dodgerblue3", phen1="Higher in COC", phen2="Higher in non-HC", label1_y_pos = y3, label1="COC vs. non-HC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/PID_post_vs_pre/ITT_", gene_set, "_Pill_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nCOC vs. non-HC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Ring_vs_baseline_PID.Gsea.1606952027042/", geneset=gene_set, color1="seagreen", phen1="Higher in CCVR", phen2="Higher in control", label1_y_pos = y2, label1="CCVR vs. control", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/PID_post_vs_pre/ITT_", gene_set, "_Ring_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nCCVR vs. control\n", gene_set_name), y_outreach = y4)

# ------------------------- #
# PID cross-sectional
gene_set <- "PID_IL6_7_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.53
y2 <- 0.53
y3 <- 0.53
y4 <- 0.7

gene_set <- "PID_NCADHERIN_PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.53
y2 <- 0.53
y3 <- 0.53
y4 <- 0.7

gene_set <- "PID_BCR_5PATHWAY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.5
y2 <- 0.25
y3 <- 0.5
y4 <- 0.7

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Injection_vs_Pill_PID.Gsea.1606955796390/", geneset=gene_set, color1="#d8b365", phen1="Higher in Net-En", phen2="Higher in COC", label1_y_pos = y1, label1="Net-En vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/PID_cross_sectional/ITT_", gene_set, "_Injection_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Ring_vs_Pill_PID.Gsea.1606955947845/", geneset=gene_set, color1="#5ab4ac", phen1="Higher in CCVR", phen2="Higher in COC", label1_y_pos = y2, label1="CCVR vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/PID_cross_sectional/ITT_", gene_set, "_Ring_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA\n", "\nCCVR vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Ring_vs_Injection_PID.Gsea.1606958620326/", geneset=gene_set, color1="#6a51a3", phen1="Higher in CCVR", phen2="Higher in Net-En", label1_y_pos = y3, label1="CCVR vs. Net-En", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/PID_cross_sectional/ITT_", gene_set, "_Ring_vs_Injection_lineplot"), mainTitle = paste0("ITT GSEA", "\nCCVR vs. Net-En\n", gene_set_name), y_outreach = y4)

# ------------------------- #
# REACTOME cross-sectional
gene_set <- "REACTOME_REGULATION_OF_TLR_BY_ENDOGENOUS_LIGAND"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.53
y2 <- 0.53
y3 <- 0.53
y4 <- 0.7

gene_set <- "REACTOME_SIGNALING_BY_THE_B_CELL_RECEPTOR_BCR_"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.53
y2 <- 0.53
y3 <- 0.53
y4 <- 0.7

gene_set <- "REACTOME_GAP_JUNCTION_ASSEMBLY"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.2
y2 <- 0.2
y3 <- 0.2
y4 <- 0.7

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Injection_vs_Pill_REACTOME.Gsea.1606955392840/", geneset=gene_set, color1="#d8b365", phen1="Higher in Net-En", phen2="Higher in COC", label1_y_pos = y1, label1="Net-En vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/REACTOME_cross_sectional/ITT_", gene_set, "_Injection_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Ring_vs_Pill_REACTOME.Gsea.1606956510380/", geneset=gene_set, color1="#5ab4ac", phen1="Higher in CCVR", phen2="Higher in COC", label1_y_pos = y2, label1="CCVR vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/REACTOME_cross_sectional/ITT_", gene_set, "_Ring_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA\n", "\nCCVR vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Ring_vs_Injection_REACTOME.Gsea.1606958146155/", geneset=gene_set, color1="#6a51a3", phen1="Higher in CCVR", phen2="Higher in Net-En", label1_y_pos = y3, label1="CCVR vs. Net-En", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/REACTOME_cross_sectional/ITT_", gene_set, "_Ring_vs_Injection_lineplot"), mainTitle = paste0("ITT GSEA", "\nCCVR vs. Net-En\n", gene_set_name), y_outreach = y4)

# ------------------------- #
# GO BP cross-sectional
gene_set <- "GO_POSITIVE_REGULATION_OF_KERATINOCYTE_DIFFERENTIATION"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.53
y2 <- 0.53
y3 <- 0.53
y4 <- 0.7

gene_set <- "GO_REGULATION_OF_ACUTE_INFLAMMATORY_RESPONSE"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.53
y2 <- 0.53
y3 <- 0.53
y4 <- 0.7

gene_set <- "GO_T_HELPER_17_TYPE_IMMUNE_RESPONSE"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.2
y2 <- 0.2
y3 <- 0.2
y4 <- 0.7

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Injection_vs_Pill_GO.Gsea.1606954626546/", geneset=gene_set, color1="#d8b365", phen1="Higher in Net-En", phen2="Higher in COC", label1_y_pos = y1, label1="Net-En vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/GO_cross_sectional/ITT_", gene_set, "_Injection_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Ring_vs_Pill_GO.Gsea.1606956975006/", geneset=gene_set, color1="#5ab4ac", phen1="Higher in CCVR", phen2="Higher in COC", label1_y_pos = y2, label1="CCVR vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/GO_cross_sectional/ITT_", gene_set, "_Ring_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA\n", "\nCCVR vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Ring_vs_Injection_GO.Gsea.1606957612691/", geneset=gene_set, color1="#6a51a3", phen1="Higher in CCVR", phen2="Higher in Net-En", label1_y_pos = y3, label1="CCVR vs. Net-En", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/GO_cross_sectional/ITT_", gene_set, "_Ring_vs_Injection_lineplot"), mainTitle = paste0("ITT GSEA", "\nCCVR vs. Net-En\n", gene_set_name), y_outreach = y4)

source("/yerkes-cifs/runs/MSigDB/GSEA_enrich_auto_vX1.R")

# CSG
gene_set <- "CSG_ISG_SET"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.4
y2 <- 0.4
y3 <- 0.4
y4 <- 0.7

gene_set <- "CSG_IMMUNE_ACTIVATION_SET"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.4
y2 <- 0.4
y3 <- 0.4
y4 <- 0.7

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Injection_vs_baseline_CSG.Gsea.1606966230762/", geneset=gene_set, color1="coral3", phen1="Higher in Net-En", phen2="Higher in non-HC", label1_y_pos = y1, label1="Net-En vs. non-HC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/CSG_plots/ITT_", gene_set, "_Injection_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. non-HC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Pill_vs_baseline_CSG.Gsea.1606966270165/", geneset=gene_set, color1="dodgerblue3", phen1="Higher in COC", phen2="Higher in non-HC", label1_y_pos = y2, label1="COC vs. non-HC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/CSG_plots/ITT_", gene_set, "_Pill_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA", "\nCOC vs. non-HC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Ring_vs_baseline_CSG.Gsea.1606966363952/", geneset=gene_set, color1="seagreen", phen1="Higher in CCVR", phen2="Higher in non-HC", label1_y_pos = y3, label1="CCVR vs. non-HC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/CSG_plots/ITT_", gene_set, "_Ring_vs_baseline_lineplot"), mainTitle = paste0("ITT GSEA\n", "\nCCVR vs. non-HC\n", gene_set_name), y_outreach = y4)

gene_set <- "CSG_ISG_SET"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.3
y2 <- 0.3
y3 <- 0.3
y4 <- 0.7

gene_set <- "CSG_IMMUNE_ACTIVATION_SET"
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.4
y2 <- 0.4
y3 <- 0.4
y4 <- 0.8

gene_set <- "CSG_CHRONIC_INFECTION_SET"
# NETEN VS PILL
gene_set_name <- gsub("_"," ", gene_set)
y1 <- 0.4
y2 <- 0.4
y3 <- 0.4
y4 <- 0.8

# CSG_INFLAMMATION_AND_IMMUNE_ACTIVATION_SET
# RING VS INJECTION

# CSG_INFLAMMATORY_GENES
# RING VS PILL

# CSG_NFKB_TARGET_GENES_SET
# RING VS PILL

IvsP <- One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Injection_vs_Pill_CSG.Gsea.1606966396725/", geneset=gene_set, color1="#d8b365", phen1="Higher in Net-En", phen2="Higher in COC", label1_y_pos = y1, label1="Net-En vs. COC", mainTitle = paste0("ITT GSEA", "\nNet-En vs. COC\n", gene_set_name), y_outreach = y4)
ggdraw(IvsP)
# ---------
One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Injection_vs_Pill_CSG.Gsea.1606966396725/", geneset=gene_set, color1="#d8b365", phen1="Higher in Net-En", phen2="Higher in COC", label1_y_pos = y1, label1="Net-En vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/CSG_plots/ITT_", gene_set, "_Injection_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA", "\nNet-En vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Ring_vs_Pill_CSG.Gsea.1606966431075/", geneset=gene_set, color1="#5ab4ac", phen1="Higher in CCVR", phen2="Higher in COC", label1_y_pos = y2, label1="CCVR vs. COC", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/CSG_plots/ITT_", gene_set, "_Ring_vs_Pill_lineplot"), mainTitle = paste0("ITT GSEA\n", "\nCCVR vs. COC\n", gene_set_name), y_outreach = y4)

One.GSEA.Enrich.Plot(dir1="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_set2/Ring_vs_Injection_CSG.Gsea.1606966459517/", geneset=gene_set, color1="#6a51a3", phen1="Higher in CCVR", phen2="Higher in Net-En", label1_y_pos = y3, label1="CCVR vs. Net-En", filename=paste0("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_ITT_figures/CSG_plots/ITT_", gene_set, "_Ring_vs_Injection_lineplot"), mainTitle = paste0("ITT GSEA", "\nCCVR vs. Net-En\n", gene_set_name), y_outreach = y4)


```


### Strict per-protocol

#### SPP analysis
```{r strict_per_protocol, eval=FALSE}
# Strict per-protocol
# ---------
# Sample table for strict per-protocol
# ---------
#superTable_metadata_filtered$Sample_ID
atable <- superTable_itt
atable <- atable[atable$Assigned_treatment == atable$Treatment,] 

metadata_v2 <- read.table("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/Metadata_v3_withSampledetails_n_InfPAM.txt", header = TRUE, row.names = NULL, check.names = FALSE, sep = "\t")
rownames(metadata_v2) <- metadata_v2$Sample_ID
atable$Infl_category_PAM <- sapply(rownames(atable), function(i) as.character(metadata_v2[i, "Infl_category_PAM"]))

supertable_strictpp <- atable

head(supertable_strictpp)

length(supertable_strictpp$Sample_ID[supertable_strictpp$Treatment == "Pretreatment"])
# 13
length(supertable_strictpp$Sample_ID[supertable_strictpp$Treatment == "Injection_NETEN"])
# 33
length(supertable_strictpp$Sample_ID[supertable_strictpp$Treatment == "Pill"])
# 28
length(supertable_strictpp$Sample_ID[supertable_strictpp$Treatment == "Ring"])
# 25

# ---------
# DESeq
# ---------
ddsspp <- DESeqDataSetFromHTSeqCount(supertable_strictpp, ".", design =~ Treatment)
ddsspp <- ddsspp[row.names(ddsspp) %in% filtered_rcnt_target_genes,]
ddsspp <- DESeq(ddsspp)
vsdspp <- varianceStabilizingTransformation(ddsspp, blind = FALSE, fitType = "parametric")
#rldspp <- rlog(ddsspp, blind = FALSE, fitType = "parametric", betaPriorVar = 10)

# Set factor levels
vsdspp$BV <- factor(vsdspp$BV, levels = c("Neg", "Int", "Pos"))
vsdspp$CST <- factor(vsdspp$CST, levels = c("CST-I", "CST-III", "CST-IV", "<NA>"))
vsdspp$Infl_category_PAM <- factor(vsdspp$Infl_category_PAM, levels = c("Low", "High", "<NA>"))
vsdspp$Infl_category_Kmeans <- factor(vsdspp$Infl_category_Kmeans, levels = c("Low", "High", "<NA>"))
vsdspp$Treatment
vsdspp$Assigned_treatment

rldspp <- rlog(ddsspp, blind = FALSE, fitType = "parametric", betaPriorVar = 10)
write.table(assay(rlditt), file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/UChoose_SPP_DESeq2_rlog.txt", sep="\t", quote=FALSE)

# ---------
# PCA
# ---------
pcaData <- plotPCA(vsdspp, intgroup=c("Treatment", "Assigned_treatment", "BV", "CST", "Infl_category_Kmeans", "Infl_category_PAM"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

# Showing only assigned treatment, all 4 study-arms
pcaData$name
ggplot(pcaData, aes(PC1, PC2, color = Treatment)) + theme_bw() + 
  theme(text = element_text(size = 10), plot.title = element_text(hjust = 0.5)) +
  labs(x=paste0("PC1: ", percentVar[1], "% variance"), y=paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  geom_point(color = 'gray40', size = 4, show_guide = FALSE) +
  geom_point(aes(color = Treatment), shape = 19, size = 3) +
  labs(title = "uCHOOSE study\nIntention-to-treat\nPCA") +
  stat_ellipse() +
  scale_color_manual(name = "Study-arm", 
                     labels = c("Baseline", "Net-En","COC","CCVR"),
                     values = c("dimgrey","coral3","dodgerblue3","seagreen")) 
#geom_text_repel(data=pcaData, aes(label=name), size = 4) 
#ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_StrictPP.pdf", height = 6, width = 8)
ggsave("/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/PCA_StrictPP_ellipse.pdf", height = 6, width = 8)

# ---------
# DESeq results
# ---------
resultsNames(ddsspp)
resspp_InjectionvsPre <- results(ddsspp, contrast = c("Treatment", "Injection_NETEN", "Pretreatment"), 
                                 independentFiltering = FALSE)
resspp_RingvsPre <- results(ddsspp, contrast = c("Treatment", "Ring", "Pretreatment"), 
                            independentFiltering = FALSE)
resspp_PillvsPre <- results(ddsspp, contrast = c("Treatment", "Pill", "Pretreatment"), 
                            independentFiltering = FALSE)
resspp_InjectionvsPill <- results(ddsspp, contrast = c("Treatment", "Injection_NETEN", "Pill"), 
                                  independentFiltering = FALSE)
resspp_RingvsPill <- results(ddsspp, contrast = c("Treatment", "Ring", "Pill"), 
                             independentFiltering = FALSE)
resspp_RingvsInjection <- results(ddsspp, contrast = c("Treatment", "Ring", "Injection_NETEN"), 
                                  independentFiltering = FALSE)


# merge DESeq results into one table

library(lessR)
library(tibble)

# Merge results by gene names
merge_df <- function(resA, resB) {
  merged_results <- Merge(resA, resB, by="row.names")
  row.names(merged_results) <- merged_results$Row.names
  merged_results[,"Row.names"] <- NULL
  return(merged_results)
}

# vector with all results names 
res_vector <- c("resspp_InjectionvsPill", "resspp_RingvsPill", "resspp_RingvsInjection")
# first result
resP <- get(res_vector[1])
res_ext <- unlist(strsplit(res_vector[1], "spp"))[2]
colnames(resP) <- gsub("$", res_ext, colnames(resP))

# merge results 
for (r in 2:length(res_vector)) {
  resQ <- get(res_vector[r])
  res_ext <- unlist(strsplit(res_vector[r], "spp"))[2]
  colnames(resQ) <- gsub("$", res_ext, colnames(resQ))
  merged_resPQ <- merge_df(resP, resQ)
  resP <- merged_resPQ
}
res_merged <- resP

head(res_merged)
res_merged$EnsemblID <- row.names(res_merged)

# Add a column with gene symbols
# gene name table: Ensembl gene id as rownames, gene symbol 1st col
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)
res_merged$GeneSymbol <- sapply(res_merged$EnsemblID, function(i) as.character(gene_name_table[i,]))

write.table(res_merged, file="/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/SPP_DESeq_results_crossarm.txt", sep="\t", quote=FALSE, row.names = FALSE)

# ---------
# Volcano plots
# ---------
gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)

fc <- 0.58
modify_result <- function(res){
  res$gene <- row.names(res)
  res$negLogPval <- -log10(res$pvalue)
  # threshold to define significant DEGs
  res <- res[!(is.na(res$log2FoldChange) | is.na(res$pvalue) | is.na(res$padj)),]
  res$sigGenes <- ifelse((res$padj < 0.05 & abs(res$log2FoldChange) > fc), 1, 0)
  res$deg <- ifelse((res$padj < 0.05 & res$log2FoldChange > fc), 1, ifelse((res$padj < 0.05 & res$log2FoldChange < -fc), -1, 0))
  res <- res[res$lfcSE < 1,]
  #res$gene_symbol <- sapply(res$gene, function(i) as.character(gene_name_table[i,]))
  return(res)
}

volcano_plot <- function(res, pline, main_title){
  ggplot(res, aes(x=log2FoldChange, y=negLogPval)) + 
    scale_color_manual(name="sigGenes", breaks = c(1,0), values = c("1"="red", "0"="black"), labels = c("DEGs", "not significant")) +
    scale_x_continuous(limits = c(-7,7), breaks = c(-6,-4,-2,0,2,4,6)) +
    scale_y_continuous(limits = c(-1,15), breaks = c(0,pline,5,10,15)) +
    geom_point(aes(color=as.factor(sigGenes)), size = 1) +
    theme_bw() +
    theme(axis.text = element_text(size = 9), axis.title = element_text(size = 12), plot.title = element_text(hjust = 0.5, size = 12), 
          panel.grid.minor = element_blank(), 
          legend.title = element_blank(), legend.position = c(1,1), legend.justification = c(1,1), 
          legend.background = element_blank(), legend.key = element_blank()) +
    geom_vline(xintercept = fc, linetype = "dotdash", color = "#636363") +
    geom_vline(xintercept = -fc, linetype = "dotdash", color = "#636363") +
    geom_hline(yintercept = pline, linetype = "dotdash", color = "#636363") +
    #geom_text_repel(data=res[res$sigGenes == 1,], aes(label=gene_symbol), size = 4) +
    labs(x=expression(fold~change~(log[2])), y=expression(p~value~(-log[10])), title=main_title)
}

# For volcano plots
res1 <- modify_result(as.data.frame(resspp_InjectionvsPre))
res2 <- modify_result(as.data.frame(resspp_RingvsPre))
res3 <- modify_result(as.data.frame(resspp_PillvsPre))
res4 <- modify_result(as.data.frame(resspp_InjectionvsPill))
res5 <- modify_result(as.data.frame(resspp_RingvsPill))
res6 <- modify_result(as.data.frame(resspp_RingvsInjection))

res_for_volplot <- c("res1", "res2", "res3", "res4", "res5", "res6")
res_for_volplot_labels <- c("Net-En vs. Baseline", 
                            "CCVR vs. Baseline", 
                            "COC vs. Baseline",
                            "Net-En vs. COC",
                            "CCVR vs. COC", 
                            "CCVR vs. Net-En")

# find the x, y limits for volcano plots, and y-intercepts
min_x <- c()
max_x <- c()
max_y <- c()
yintercept_vector <- c()
genes_up <- c()
genes_down <- c()
for (r in 1:length(res_for_volplot)) {
  resQ <- get(res_for_volplot[r])
  # to find p-value equivalent to 0.05, whichever yields same # of sig genes
  min_x <- append(min_x, min(resQ$log2FoldChange), after = length(min_x))
  max_x <- append(max_x, max(resQ$log2FoldChange), after = length(max_x))
  max_y <- append(max_y, max(resQ$negLogPval), after = length(max_y))
  yintercept <- as.numeric(format(round(min(resQ[resQ$sigGenes == 1 | resQ$sigGenes == -1,]$negLogPval), 2), nsmall = 2))
  yintercept_vector <- append(yintercept_vector, yintercept, after = length(yintercept_vector))
  genes_up <- append(genes_up, sum(resQ$deg[resQ$deg == 1]), after = length(genes_up))
  genes_down <- append(genes_down, abs(sum(resQ$deg[resQ$deg == -1])), after = length(genes_down))
}

# Helps to define axes limits
min(min_x)
max(max_x)
max(max_y)
yintercept_vector

# ACTION: MANUALLY adjust axes limits in volcano_plot function
volplot1 <- volcano_plot(res1, yintercept_vector[1], res_for_volplot_labels[1])
volplot2 <- volcano_plot(res2, yintercept_vector[2], res_for_volplot_labels[2])
volplot3 <- volcano_plot(res3, yintercept_vector[3], res_for_volplot_labels[3])
volplot4 <- volcano_plot(res4, yintercept_vector[4], res_for_volplot_labels[4])
volplot5 <- volcano_plot(res5, yintercept_vector[5], res_for_volplot_labels[5])
volplot6 <- volcano_plot(res6, yintercept_vector[6], res_for_volplot_labels[6])

ggdraw(volplot6)
# Table for up and down-regulated genes
# ---------
res_for_volplot_labels
genes_up
results_details <- matrix(c(res_for_volplot_labels, genes_down, genes_up), 
                          ncol = 3, byrow = FALSE)
rownames(results_details) <- res_for_volplot_labels
colnames(results_details) <- c("comparison", "down-regulated genes", "up-regulated genes")
results_details <- as.data.frame(results_details)

# add DEGs table to group plot
t1 <- tableGrob(results_details, theme=ttheme_minimal(base_size = 10), rows=NULL) 
t1 <- gtable::gtable_add_grob(t1, grobs = rectGrob(gp=gpar(fill=NA, lwd=2)), t = 2, b = nrow(t1), l = 2, r = ncol(t1))

r1 <- rectGrob(gp=gpar(fill="white", col="white"))
grid.arrange(volplot1, volplot2, volplot3, volplot4, volplot5, volplot6, 
             arrangeGrob(r1, r1, t1, ncol = 3), 
             nrow=3, top = "\nStrict per-protocol\nVolcano plots, all comparisons\n") %>%
  ggsave(filename = "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/VolcanoPlots_grouped_StrictPerProtocol.pdf", width = 10, height = 10)
```

#### SPP GSEA
```{r spp_gsea, eval=FALSE, message=FALSE}
# mkdir GSEA_SPP

# [1] Expression dataset file
# rlog files contain col names as 'sample names' and row names are the 'genes'
# For GSEA, we need to add two new cols at the beginning, 'Name' and 'Description'

gene_name_table <- read.table("/yerkes-cifs/runs/STAR/GRCh38_ERCC/GRCh38_ERCC_gene_name_table.txt", header = TRUE, row.names = 1, check.names = FALSE)

rlog_matrix <- as.data.frame(assay(rldspp))
gene_details <- data.frame(row.names(rlog_matrix), 
                           sapply(rownames(rlog_matrix), function(i) as.character(gene_name_table[i,])))
colnames(gene_details) <- c("Name", "Description")
rlog_for_gsea <- cbind(gene_details, rlog_matrix)
write.table(rlog_for_gsea, file = paste0(par_dir, "GSEA_SPP/rlog_for_GSEA_SPP.txt"), sep = "\t", quote = FALSE, row.names = FALSE)

# [2] Phenotype label file
# Function to create Phenotype label file for a specific condition
#Line format:      (number of samples) (space) (number of classes) (space) 1
df_samples <- supertable_strictpp[,c("Sample_ID","Treatment")] 
df_samples$feature <- df_samples$Treatment
length(df_samples$Sample_ID)
line1 <- paste(length(df_samples$Sample_ID), length(unique(df_samples$feature)), 1, sep = " ")
line2 <- paste0("# ", gsub(",", "", toString(unique(df_samples$feature))))
line3 <- gsub(",", "", toString(df_samples$feature))

write.table(line1, file = paste0(par_dir, "GSEA_SPP/itt_phenotype.cls"), 
            quote = FALSE, eol = "\n", row.names = FALSE, col.names = FALSE)
write.table(line2, file = paste0(par_dir, "GSEA_SPP/itt_phenotype.cls"), 
            quote = FALSE, eol = "\n", row.names = FALSE, col.names = FALSE, append = TRUE)
write.table(line3, file = paste0(par_dir, "GSEA_SPP/itt_phenotype.cls"), 
            quote = FALSE, eol = "\n", row.names = FALSE, col.names = FALSE, append = TRUE)


# [3] Run GSEA 
# GSEA results for the following were saved in '/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_SPP'

# Pathway collections
# Hallmark, Kegg, BioCarta, Custom gene-sets (including Serpins, Integrins, epithelial barrier), Th17

# Comparisons
# InjectionvsPill 
# RingvsPill 
# RingvsInjection


# -----------------
# For each pathway collection, combine stats from each cross-treatment comparison
# -----------------

# Merge all results for a pathway collection
merge_df <- function(resA, resB) {
  merged_results <- Merge(resA, resB, by="row.names")
  row.names(merged_results) <- merged_results$Row.names
  merged_results[,"Row.names"] <- NULL
  return(merged_results)
}

modify_gsea_report <- function(resA, res_ext) {
  row.names(resA) <- resA$NAME
  #resA <- resA[,c("NES", "NOM p-val", "FDR q-val")]
  resA <- resA[,c("SIZE", "NES", "NOM p-val", "FDR q-val")]
  resA <- round(resA, digits = 3)
  #resA[,c("NAME","GS<br> follow link to MSigDB", "GS DETAILS")]<- NULL
  colnames(resA) <- gsub("$", paste0("_", res_ext), colnames(resA))
  return(resA)
}

# Compile GSEA stats into one table for each pathway collection
gsea_dir_spp <- "/yerkes-cifs/runs/190419_J00157_0138_AH52H2BBXY/BL33_uCHOOSE_Analysis/uChoose_Analysis_Oct2020/GSEA_SPP/"
pathway_coll <- c("HALLMARK", "KEGG", "BIOCARTA", "CSG", "PID", "REACTOME")
pathway_coll <- c("GOBP")
comparison_order_cross <- c("Injection_vs_Pill", "Ring_vs_Pill", "Ring_vs_Injection")

#data_list <- list()
gsea_dir <- gsea_dir_spp
#pathway_coll <- pathway_coll2
for (i in 1:length(pathway_coll)) {
  #a
  #f
  #data
  #edited_data
  #merged_data
  #i <- 1
  #j <- 1
  #colnames(merged_data)
  
  pathway <- pathway_coll[i]
  # merge results for each pathway coll
  for (j in 1:length(comparison_order)) {
    comparison <- comparison_order[j]
    print(pathway)
    print(comparison)
    
    dir_name <- paste(comparison, pathway, sep = "_")
    a <- dir(path = gsea_dir, recursive = FALSE, pattern = gsub("$", "*", dir_name))
    f <- list.files(path = paste0(gsea_dir, a), recursive = FALSE, pattern = "gsea_report_for.*\\.(xls|tsv)$", full.names = TRUE)
    
    data <- read.table(f[1], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t")
    data <- rbind(data, read.table(f[2], check.names = FALSE, row.names = NULL, header = TRUE, sep = "\t"))
    
    # modify this gsea report
    edited_data <- modify_gsea_report(data, dir_name)
    # merge data
    if (j == 1) { merged_data <- edited_data}
    else {  merged_data <- merge_df(merged_data, edited_data)}
  }
  # push merged data in a list
  merged_data <- data.frame("NAME" = row.names(merged_data), merged_data)
  results_file <- paste0(gsea_dir, "compiled_gsea_SPP_", pathway, ".txt")
  write.table(merged_data, results_file, quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")
  #data_list[[i]] <- merged_data
}
#names(data_list) <- pathway_coll
```